// td-antlers-parser/out/runtime/document/antlersDocument.js
import { ScopeEngine } from '../../antlers/scope/engine.js';
import TagManager from '../../antlers/tagManagerInstance.js';
import { resolveTypedTree } from '../../antlers/tags.js';
import { parseIdeHelper } from '../../idehelper/parser.js';
import { AntlersNode } from '../nodes/abstractNode.js';
import { TailwindClassAnalyzer } from '../analyzers/tailwindClassAnalyzer.js';
import { DocumentParser } from "../parser/documentParser.js";
import { ContextResolver } from "./contexts/contextResolver.js";
import { DocumentCursor } from "./documentCursor.js";
import { DocumentErrors } from "./documentErrors.js";
import { FrontMatterParser } from './frontMatter/frontMatterParser.js';
import { NodeScanner } from "./scanners/nodeScanner.js";
import { RangeScanner } from "./scanners/rangeScanner.js";
import { Transformer } from './transformer.js';
import { YamlDocument } from "./yamlDocument.js";
export class AntlersDocument {
    constructor() {
        this.documentParser = new DocumentParser();
        this.transformer = null;
        this.originalText = '';
        this.ranges = new RangeScanner(this);
        this.nodes = new NodeScanner(this);
        this.cursor = new DocumentCursor(this);
        this.errors = new DocumentErrors(this);
        this.project = null;
        this.documentUri = "";
    }
    static fromText(text, project) {
        const document = new this();
        if (typeof project != "undefined") {
            document.project = project;
        }
        return document.loadString(text);
    }
    static childFromText(text, seedPosition = null) {
        const document = new AntlersDocument();
        if (seedPosition != null) {
            document.getDocumentParser().setSeedPosition(seedPosition);
        }
        document.getDocumentParser().withChildDocuments(true);
        document.loadString(text);
        return {
            renderNodes: document.getDocumentParser().getRenderNodes(),
            content: text,
            document: document
        };
    }
    isValid() {
        if (this.hasInvalidControlFlowStructures() || this.hasUnclosedStructures()) {
            return false;
        }
        return true;
    }
    getDocumentOptions() {
        const nodes = this.getAllNodes();
        if (nodes.length == 0 || (nodes[0] instanceof AntlersNode) == false) {
            return null;
        }
        const checkNode = nodes[0];
        if (checkNode.isComment == false) {
            return null;
        }
        return parseIdeHelper(this.documentUri, checkNode);
    }
    isFormattingEnabled() {
        const options = this.getDocumentOptions();
        if (options == null) {
            return true;
        }
        return options.formatterEnabled;
    }
    hasInvalidControlFlowStructures() {
        return this.documentParser.hasUnclosedIfStructures();
    }
    hasUnclosedStructures() {
        return this.documentParser.hasUnclosedStructures();
    }
    getFrontMatterDoc() {
        if (this.hasFrontMatter()) {
            const yamlDoc = new YamlDocument();
            yamlDoc.loadString(this.getFrontMatter());
            return yamlDoc;
        }
        return null;
    }
    hasFrontMatter() {
        return this.documentParser.getFrontMatter().trim().length > 0;
    }
    getFrontMatter() {
        return this.documentParser.getFrontMatter();
    }
    getFrontMatterScope() {
        if (this.project == null) {
            return null;
        }
        const frontMatterParser = new FrontMatterParser(this.project);
        frontMatterParser.parse(this.getFrontMatter());
        return frontMatterParser.getScope();
    }
    updateProject(project) {
        this.project = project;
        this.updateProjectDetails();
    }
    updateProjectDetails() {
        try {
            if (this.project != null) {
                const scopeEngine = new ScopeEngine(this.project, this.documentUri, this), analysisNodes = this.documentParser.antlersNodes();
                analysisNodes.forEach((node) => {
                    if (this.project != null && TagManager.instance?.canResolveSpecialTypes(node.getTagName())) {
                        const specialResults = TagManager.instance?.resolveSpecialType(node.getTagName(), node, this.project);
                        if (specialResults.context != null) {
                            node.reference = specialResults.context;
                        }
                    }
                });
                scopeEngine.analyzeScope(analysisNodes);
                for (let i = 0; i < analysisNodes.length; i++) {
                    const node = analysisNodes[i];
                    if (node.hasParameters) {
                        for (let j = 0; j < node.parameters.length; j++) {
                            const thisParam = node.parameters[j];
                            if (thisParam.hasInterpolations()) {
                                for (let l = 0; l < thisParam.interpolations.length; l++) {
                                    const thisInterpolation = thisParam.interpolations[l];
                                    if (node.processedInterpolationRegions.has(thisInterpolation)) {
                                        const tInterpolation = node.processedInterpolationRegions.get(thisInterpolation);
                                        tInterpolation.forEach((interpolationNode) => {
                                            if (interpolationNode instanceof AntlersNode) {
                                                interpolationNode.currentScope = node.currentScope;
                                            }
                                        });
                                    }
                                }
                            }
                        }
                    }
                }
                resolveTypedTree(this);
            }
        }
        catch (err) {
            console.log(err);
        }
    }
    loadString(text) {
        this.originalText = text;
        this.documentParser.parse(text);
        this.updateProjectDetails();
        return this;
    }
    getDocumentParser() {
        return this.documentParser;
    }
    getParsedContent() {
        return this.documentParser.getParsedContent();
    }
    getAllNodes() {
        return this.documentParser.getNodes();
    }
    getAllAntlersNodes() {
        return this.documentParser.antlersNodes();
    }
    /**
     * Extracts all literal text from the Antlers document.
     */
    extractText() {
        const parts = [];
        this.nodes.getAllLiteralNodes().forEach((node) => {
            parts.push(node.content);
        });
        return parts.join("");
    }
    getTailwindClassCandidates() {
        return (new TailwindClassAnalyzer()).analyze(this);
    }
    getNameAt(position) {
        const node = this.nodes.getNodeAt(position);
        if (node instanceof AntlersNode) {
            return node.name;
        }
        return null;
    }
    getLinesAround(line) {
        return this.documentParser.getLinesAround(line);
    }
    getContent() {
        return this.documentParser.getContent();
    }
    getOriginalContent() {
        return this.documentParser.getOriginalContent();
    }
    reloadDocument() {
        this.loadString(this.originalText);
    }
    getFeaturesAt(position) {
        const node = this.nodes.getNodeAt(position);
        return ContextResolver.resolveContext(position, node, this, false, node);
    }
    getText(start, end) {
        return this.documentParser.getText(start, end);
    }
    charLeftAt(position) {
        return this.documentParser.charLeftAt(position);
    }
    charAt(position) {
        return this.documentParser.charAt(position);
    }
    charRightAt(position) {
        return this.documentParser.charRightAt(position);
    }
    getLineIndex(lineNumber) {
        return this.documentParser.getLineIndex(lineNumber);
    }
    getLineText(lineNumber) {
        return this.documentParser.getLineText(lineNumber);
    }
    wordAt(position, tabSize = 4) {
        return this.documentParser.wordAt(position, tabSize);
    }
    wordLeftAt(position, tabSize = 4) {
        return this.documentParser.wordLeftAt(position, tabSize);
    }
    wordRightAt(position, tabSize = 4) {
        return this.documentParser.wordRightAt(position, tabSize);
    }
    punctuationLeftAt(position, tabSize = 4) {
        return this.documentParser.punctuationLeftAt(position, tabSize);
    }
    punctuationRightAt(position, tabSize = 4) {
        return this.documentParser.punctuationRightAt(position, tabSize);
    }
    getLineChars(lineNumber) {
        const lineText = this.getLineText(lineNumber);
        if (lineText != null) {
            return lineText.split("");
        }
        return null;
    }
    transform() {
        if (this.transformer == null) {
            this.transformer = new Transformer(this);
        }
        return this.transformer;
    }
}
