// td-antlers-parser/out/runtime/document/documentErrors.js
export class DocumentErrors {
    constructor(doc) {
        this.doc = doc;
    }
    hasStructureErrors() {
        return this.doc.getDocumentParser().getStructureErrors().length > 0;
    }
    getFirstStructureError() {
        return this.doc.getDocumentParser().getStructureErrors()[0];
    }
    hasAny() {
        return this.all().length > 0;
    }
    all() {
        const errorHashes = [], errors = [];
        this.doc.getAllAntlersNodes().forEach((node) => {
            node.getErrors().forEach((error) => {
                if (errorHashes.includes(error.hash()) == false) {
                    errorHashes.push(error.hash());
                    errors.push(error);
                }
            });
        });
        this.doc.getDocumentParser().getAntlersErrors().forEach((error) => {
            if (errorHashes.includes(error.hash()) == false) {
                errorHashes.push(error.hash());
                errors.push(error);
            }
        });
        return errors;
    }
}
