// td-antlers-parser/out/runtime/document/documentCursor.js
import { NodeQueries } from './scanners/nodeQueries.js';
import { ContextResolver } from './contexts/contextResolver.js';
import { AntlersNodeQueries } from './scanners/antlersNodeQueries.js';
import { AntlersNode } from '../nodes/abstractNode.js';
export class DocumentCursor {
    constructor(doc) {
        this.doc = doc;
    }
    getAncestorsAt(line, char) {
        const position = this.position(line, char);
        if (position == null) {
            return [];
        }
        return AntlersNodeQueries.findAncestorsBeforePosition(position, this.doc.getAllAntlersNodes());
    }
    getNameAt(line, char) {
        const node = this.getNodeAt(line, char);
        if (node instanceof AntlersNode) {
            return node.name;
        }
        return null;
    }
    getScopeAt(line, char) {
        const node = this.getNodeAt(line, char);
        if (node instanceof AntlersNode) {
            if (node.currentScope != null) {
                return node.currentScope;
            }
        }
        return null;
    }
    getFeaturesAt(line, char) {
        const position = this.position(line, char), node = this.getNodeAt(line, char);
        return ContextResolver.resolveContext(position, node, this.doc, false, node);
    }
    getNodeAt(line, char) {
        return NodeQueries.findNodeAtPosition(this.position(line, char), this.doc.getDocumentParser().getNodes());
    }
    getNodeBefore(line, char) {
        return NodeQueries.findNodeBeforePosition(this.position(line, char), this.doc.getDocumentParser().getNodes());
    }
    getNodeAfter(line, char) {
        return NodeQueries.findNodeAfterPosition(this.position(line, char), this.doc.getDocumentParser().getNodes());
    }
    getNodesBefore(line, char) {
        return NodeQueries.findNodesBeforePosition(this.position(line, char), this.doc.getDocumentParser().getNodes());
    }
    getAssignmentsBefore(line, char) {
        const assignments = this.doc.getDocumentParser().getLanguageParser().getRuntimeAssignments(), returnNodes = [], pos = this.position(line, char);
        if (pos != null) {
            assignments.forEach((assignment) => {
                if (NodeQueries.isBefore(assignment.target, pos)) {
                    returnNodes.push(assignment);
                }
            });
        }
        return returnNodes;
    }
    getAssignmentsAfter(line, char) {
        const assignments = this.doc.getDocumentParser().getLanguageParser().getRuntimeAssignments(), returnNodes = [], pos = this.position(line, char);
        if (pos != null) {
            assignments.forEach((assignment) => {
                if (NodeQueries.isAfter(assignment.target, pos)) {
                    returnNodes.push(assignment);
                }
            });
        }
        return returnNodes;
    }
    getNodesAfter(line, char) {
        return NodeQueries.findNodesAfterPosition(this.position(line, char), this.doc.getDocumentParser().getNodes());
    }
    getIsWithinPairedNode(line, char) {
        const node = this.getNodeBefore(line, char);
        if (node == null) {
            return false;
        }
        return AntlersNodeQueries.isPairedNode(node);
    }
    position(line, char) {
        return this.doc.getDocumentParser().positionFromCursor(line, char);
    }
    charLeftAt(line, char) {
        return this.doc.getDocumentParser().charLeftAtCursor(line, char);
    }
    charAt(line, char) {
        return this.doc.getDocumentParser().charAtCursor(line, char);
    }
    charRightAt(line, char) {
        return this.doc.getDocumentParser().charRightAtCursor(line, char);
    }
    wordLeftAt(line, char, tabSize = 4) {
        return this.doc.getDocumentParser().wordLeftAtCursor(line, char, tabSize);
    }
    wordRightAt(line, char, tabSize = 4) {
        return this.doc.getDocumentParser().wordRightAtCursor(line, char, tabSize);
    }
    wordAt(line, char, tabSize = 4) {
        return this.doc.getDocumentParser().wordAtCursor(line, char, tabSize);
    }
    punctuationLeftAt(line, char, tabSize = 4) {
        return this.doc.getDocumentParser().punctuationLeftAtCursor(line, char, tabSize);
    }
    punctuationRightAt(line, char, tabSize = 4) {
        return this.doc.getDocumentParser().punctuationRightAtCursor(line, char, tabSize);
    }
}
