// td-antlers-parser/out/runtime/document/documentManager.js
import ProjectManager from '../../projects/projectManager.js';
import { AntlersDocument } from './antlersDocument.js';
class DocumentManager {
    constructor() {
        this.loadedDocuments = new Map();
        this.activeProject = null;
    }
    getDocuments() {
        const docs = [];
        this.loadedDocuments.forEach((doc) => {
            docs.push(doc);
        });
        return docs;
    }
    setProject(project) {
        this.activeProject = project;
        this.loadedDocuments.forEach((document) => {
            if (this.activeProject != null) {
                document.updateProject(this.activeProject);
            }
        });
    }
    refreshDocumentState() {
        if (this.activeProject != null) {
            this.setProject(this.activeProject);
        }
    }
    hasDocument(uri) {
        return this.loadedDocuments.has(uri);
    }
    createDocument(uri) {
        if (!this.loadedDocuments.has(uri)) {
            const doc = new AntlersDocument();
            doc.documentUri = uri;
            this.loadedDocuments.set(uri, doc);
        }
    }
    loadDocument(uri, text) {
        this.loadedDocuments.set(uri, AntlersDocument.fromText(text));
    }
    getDocument(uri) {
        return this.loadedDocuments.get(uri);
    }
    updateDocument(uri, text) {
        if (this.loadedDocuments.has(uri)) {
            const document = this.loadedDocuments.get(uri);
            document.loadString(text);
            const viewRef = ProjectManager.instance?.getStructure()?.findView(uri);
            if (viewRef != null) {
                viewRef.document = document;
            }
        }
    }
    createOrUpdate(uri, text) {
        this.createDocument(uri);
        this.updateDocument(uri, text);
    }
    getDocumentErrors(uri) {
        if (!this.loadedDocuments.has(uri)) {
            return [];
        }
        const doc = this.loadedDocuments.get(uri);
        return doc.errors.all();
    }
}
DocumentManager.instance = null;
if (typeof DocumentManager.instance == 'undefined' || DocumentManager.instance == null) {
    DocumentManager.instance = new DocumentManager();
}
export default DocumentManager;
