// td-antlers-parser/out/antlers/tags/core/form/formSuccess.js
import { makeTagDoc } from '../../../../documentation/utils.js';
import { exclusiveResultList } from '../../../tagManager.js';
import FormHandleParam from './formHandleParam.js';
import { HandleParams } from './parameterCompletions.js';
import { resolveFormSetReference } from './resolveFormSetReference.js';
const FormSuccess = {
    tagName: 'form:success',
    allowsArbitraryParameters: false,
    allowsContentClose: false,
    requiresClose: true,
    hideFromCompletions: false,
    injectParentScope: false,
    introducedIn: null,
    parameters: [
        FormHandleParam
    ],
    resolveSpecialType: resolveFormSetReference,
    resovleParameterCompletionItems: (parameter, params) => {
        if (HandleParams.includes(parameter.name)) {
            return exclusiveResultList(params.project.getUniqueFormNames());
        }
        return null;
    },
    resolveDocumentation: (params) => {
        return makeTagDoc('form:success Tag', 'The `form:success` tag can be used to check if a form submission was successful or not.', 'https://statamic.dev/tags/form-success');
    }
};
export default FormSuccess;
