// td-antlers-parser/out/antlers/tags/core/form/formFields.js
import { makeTagDocWithCodeSample } from '../../../../documentation/utils.js';
const FormFields = {
    tagName: 'form:fields',
    hideFromCompletions: false,
    requiresClose: true,
    injectParentScope: false,
    allowsContentClose: false,
    allowsArbitraryParameters: false,
    introducedIn: null,
    parameters: [
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'get',
            description: 'A list of fields to retrieve.',
            expectsTypes: ['string', 'array'],
            isDynamic: false
        },
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'only',
            description: 'Only the fields listed will be returned.',
            expectsTypes: ['string', 'array'],
            isDynamic: false
        },
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'except',
            description: 'The fields listed will not be returned.',
            expectsTypes: ['string', 'array'],
            isDynamic: false
        }
    ],
    resolveDocumentation: (params) => {
        return makeTagDocWithCodeSample('Form Fields Tag', 'The `form:fields` tag iterates over the fields of a form, allowing you to render them dynamically.', `{{ form:contact }}
    {{ fields }}
        <label>{{ display }}</label>
        {{ field }}
    {{ /fields }}
{{ /form:contact }}`, 'https://statamic.dev/tags/form-fields');
    }
};
export default FormFields;
