// td-antlers-parser/out/antlers/tags/core/form/formSubmission.js
import { makeTagDoc } from '../../../../documentation/utils.js';
import { exclusiveResultList, } from "../../../tagManager.js";
import FormHandleParam from "./formHandleParam.js";
import { HandleParams } from "./parameterCompletions.js";
import { resolveFormSetReference } from "./resolveFormSetReference.js";
import { getFormHandle } from "./utils.js";
const FormSubmission = {
    tagName: "form:submission",
    requiresClose: true,
    allowsContentClose: false,
    allowsArbitraryParameters: false,
    hideFromCompletions: false,
    injectParentScope: false,
    introducedIn: null,
    parameters: [FormHandleParam],
    resolveSpecialType: resolveFormSetReference,
    resovleParameterCompletionItems: (parameter, params) => {
        if (HandleParams.includes(parameter.name)) {
            return exclusiveResultList(params.project.getUniqueFormNames());
        }
        return null;
    },
    augmentScope: (symbol, scope) => {
        const formHandle = getFormHandle(symbol);
        if (formHandle.length > 0) {
            const formFields = scope.statamicProject.getFormBlueprintFields(formHandle);
            scope.addBlueprintFields(symbol, formFields);
        }
        return scope;
    },
    resolveDocumentation: (params) => {
        return makeTagDoc('form:submission Tag', 'The `form:submission` tag can be used to access the form submission submitted by the current user.', 'https://statamic.dev/tags/form-submission');
    }
};
export default FormSubmission;
