// td-antlers-parser/out/antlers/tags/core/form/formSubmissions.js
import { makeTagDoc } from '../../../../documentation/utils.js';
import { exclusiveResultList, } from "../../../tagManager.js";
import FormHandleParam from "./formHandleParam.js";
import { HandleParams } from "./parameterCompletions.js";
import { resolveFormSetReference } from "./resolveFormSetReference.js";
import { getFormHandle } from "./utils.js";
const FormSubmissions = {
    tagName: "form:submissions",
    allowsArbitraryParameters: false,
    allowsContentClose: false,
    hideFromCompletions: false,
    injectParentScope: false,
    requiresClose: true,
    introducedIn: null,
    parameters: [FormHandleParam],
    resolveSpecialType: resolveFormSetReference,
    resovleParameterCompletionItems: (parameter, params) => {
        if (HandleParams.includes(parameter.name)) {
            return exclusiveResultList(params.project.getUniqueFormNames());
        }
        return null;
    },
    augmentScope: (symbol, scope) => {
        if (symbol.isClosingTag) {
            return scope;
        }
        const formHandle = getFormHandle(symbol);
        if (formHandle.length > 0) {
            const formFields = scope.statamicProject.getFormBlueprintFields(formHandle);
            scope.addBlueprintFields(symbol, formFields);
        }
        scope.addVariable({
            name: "date",
            dataType: "date",
            sourceField: null,
            sourceName: "*internal.forms.submissions",
            introducedBy: symbol,
        });
        scope.addVariable({
            name: "no_results",
            dataType: "boolean",
            sourceField: null,
            sourceName: "*internal.forms.submissions",
            introducedBy: symbol,
        });
        scope.addVariable({
            name: "total_results",
            dataType: "number",
            sourceField: null,
            sourceName: "*internal.forms.submissions",
            introducedBy: symbol,
        });
        return scope;
    },
    resolveDocumentation: (params) => {
        return makeTagDoc('form:submissions Tag', 'The `form:submissions` tag can be used to fetch previously saved submissions for a form.', 'https://statamic.dev/tags/form-submissions');
    }
};
export default FormSubmissions;
