// td-antlers-parser/out/antlers/tags/core/form/formErrors.js
import { makeTagDoc } from '../../../../documentation/utils.js';
import { exclusiveResultList } from '../../../tagManager.js';
import FormHandleParam from './formHandleParam.js';
import { HandleParams } from './parameterCompletions.js';
import { resolveFormSetReference } from './resolveFormSetReference.js';
const FormErrors = {
    tagName: 'form:errors',
    allowsArbitraryParameters: false,
    allowsContentClose: true,
    requiresClose: false,
    hideFromCompletions: false,
    injectParentScope: false,
    introducedIn: null,
    parameters: [
        FormHandleParam
    ],
    resolveSpecialType: resolveFormSetReference,
    resovleParameterCompletionItems: (parameter, params) => {
        if (HandleParams.includes(parameter.name)) {
            return exclusiveResultList(params.project.getUniqueFormNames());
        }
        return null;
    },
    resolveDocumentation: (params) => {
        return makeTagDoc('form:errors Tag', 'The `form:errors` tag can be used to retrieve validation errors after a user has submitted a form.', 'https://statamic.dev/tags/form-errors');
    }
};
export default FormErrors;
