// td-antlers-parser/out/antlers/tags/core/taxonomies/taxonomyCount.js
import { makeTagDocWithCodeSample } from '../../../../documentation/utils.js';
const TaxonomyCount = {
    tagName: 'taxonomy:count',
    hideFromCompletions: false,
    requiresClose: false,
    injectParentScope: false,
    allowsContentClose: false,
    allowsArbitraryParameters: true,
    introducedIn: null,
    parameters: [
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: ['from', 'use', 'taxonomy'],
            allowsVariableReference: true,
            name: 'in',
            description: 'The taxonomy (or taxonomies) to count terms from.',
            expectsTypes: ['string', 'array'],
            isDynamic: false
        },
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: ['in', 'use', 'taxonomy'],
            allowsVariableReference: true,
            name: 'from',
            description: 'The taxonomy (or taxonomies) to count terms from.',
            expectsTypes: ['string', 'array'],
            isDynamic: false
        }
    ],
    resolveDocumentation: (params) => {
        return makeTagDocWithCodeSample('Taxonomy Count Tag', 'The `taxonomy:count` tag returns the number of terms in the given taxonomy.', `{{ taxonomy:count from="tags" }}`, 'https://statamic.dev/tags/taxonomy-count');
    }
};
export default TaxonomyCount;
