// td-antlers-parser/out/antlers/tags/core/vite/viteContent.js
import { makeTagDocWithCodeSample } from '../../../../documentation/utils.js';
const ViteContent = {
    tagName: 'vite:content',
    hideFromCompletions: false,
    requiresClose: false,
    injectParentScope: false,
    allowsContentClose: false,
    allowsArbitraryParameters: false,
    introducedIn: null,
    parameters: [
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'src',
            description: 'The path to the asset whose contents should be returned.',
            expectsTypes: ['string'],
            isDynamic: false
        },
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'directory',
            description: 'Override the build directory for this resource.',
            expectsTypes: ['string'],
            isDynamic: false
        }
    ],
    resolveDocumentation: (params) => {
        return makeTagDocWithCodeSample('Vite Content Tag', 'The `vite:content` tag returns the raw contents of a Vite-built asset (useful for inlining SVGs or small stylesheets).', `{{ vite:content src="resources/images/logo.svg" }}`, 'https://statamic.dev/tags/vite-content');
    }
};
export default ViteContent;
