// td-antlers-parser/out/runtime/document/printers/nodeBuffer.js
import { replaceAllInString } from '../../../utils/strings.js';
export class NodeBuffer {
    constructor(node, indent, prepend) {
        this.buffer = '';
        this.closeString = '';
        this.relativeIndentSize = 0;
        this.indentSeed = 0;
        this.baseIndent = indent;
        if (node.isInterpolationNode) {
            this.buffer = '{';
            this.closeString = '}';
        }
        else {
            this.buffer = '{{ ';
            if (node.isSelfClosing) {
                this.closeString = ' /}}';
            }
            else {
                this.closeString = ' }}';
            }
        }
        if (node.pathReference?.isStrictTagReference) {
            this.buffer += '%';
        }
        if (prepend != null && prepend.trim().length > 0) {
            this.buffer += prepend + ' ';
        }
    }
    setIndentSeed(indent) {
        this.indentSeed = indent;
        return this;
    }
    close() {
        if (this.closeString == ' }}') {
            if (this.buffer.endsWith(' ')) {
                this.buffer += '}}';
            }
            else {
                this.buffer += this.closeString;
            }
        }
        else {
            this.buffer += this.closeString;
        }
        return this;
    }
    appendT(text) {
        if (this.buffer.endsWith(' ')) {
            this.buffer = this.buffer.trimEnd();
        }
        this.buffer += text;
        return this;
    }
    appendTS(text, preserveSpace = false) {
        if (this.buffer.endsWith(' ')) {
            this.buffer = this.buffer.trimEnd();
        }
        if (this.buffer.endsWith('{') && !preserveSpace) {
            text = text.trimStart();
        }
        this.buffer += text;
        return this;
    }
    append(text) {
        this.buffer += text;
        return this;
    }
    appendOS(text) {
        if (this.buffer.endsWith(' ') == false
            && this.buffer.endsWith('(') == false
            && this.buffer.endsWith('{') == false
            && this.buffer.endsWith('[') == false
            && this.buffer.endsWith(':') == false) {
            this.buffer += ' ';
        }
        return this.append(text);
    }
    appendS(text) {
        let appendBuffer = '';
        if (this.buffer.endsWith(' ') == false) {
            appendBuffer += ' ';
        }
        appendBuffer += text + ' ';
        this.buffer += appendBuffer;
        return this;
    }
    indent() {
        let repeatCount = this.baseIndent;
        if (repeatCount == 0) {
            repeatCount = 1;
        }
        else {
            repeatCount += 2;
        }
        if (this.relativeIndentSize > 0) {
            repeatCount += this.relativeIndentSize;
        }
        this.buffer += ' '.repeat(repeatCount);
        return this;
    }
    addIndent(number) {
        if (number <= 0) {
            return this;
        }
        this.buffer += ' '.repeat(number);
        return this;
    }
    paramS(param) {
        let bParam = ' ';
        if (param.isVariableReference) {
            bParam += ':';
        }
        bParam += param.name + '=' + param.nameDelimiter + param.value + param.nameDelimiter;
        this.append(bParam);
        return this;
    }
    replace(find, replace) {
        this.buffer = replaceAllInString(this.buffer, find, replace);
        return this;
    }
    relativeIndent(relativeTo) {
        const bufferLines = this.buffer.split("\n");
        if (bufferLines.length == 0) {
            return this;
        }
        let lastLine = bufferLines[bufferLines.length - 1].trimEnd();
        if (lastLine.endsWith('(')) {
            lastLine = lastLine.slice(0, -1);
        }
        if (lastLine.endsWith(relativeTo) == false) {
            return this;
        }
        this.relativeIndentSize = lastLine.lastIndexOf(relativeTo);
        //  this.relativeIndentSize += Math.ceil(relativeTo.length / 2);
        // this.relativeIndentSize += 4;
        return this;
    }
    newlineIndent() {
        this.newLine();
        this.indent();
        return this;
    }
    newlineNDIndent() {
        this.buffer = this.buffer.trimEnd();
        this.newLine();
        this.indent();
        return this;
    }
    newLine() {
        this.buffer += "\n";
        return this;
    }
    getContent() {
        return this.buffer;
    }
    endsWith(value) {
        return this.buffer.endsWith(value);
    }
}
