// td-antlers-parser/out/antlers/tagManager.js
import { CompletionItemKind, } from "vscode-languageserver-types";
import { HtmlClassParameter, HtmlStyleParameter } from './htmlCompat/parameters.js';
const DnyamicParameter = {
    name: "dynamic",
    acceptsVariableInterpolation: false,
    aliases: [],
    allowsVariableReference: true,
    description: "",
    expectsTypes: ["string", "number", "array", "boolean"],
    isRequired: false,
    isDynamic: true,
};
function dynamicParameter(name) {
    if (name == 'class') {
        return HtmlClassParameter;
    }
    else if (name == 'style') {
        return HtmlStyleParameter;
    }
    return {
        name: name,
        acceptsVariableInterpolation: false,
        aliases: [],
        allowsVariableReference: true,
        description: "",
        expectsTypes: ["string", "number", "array"],
        isRequired: false,
        isDynamic: true,
    };
}
function resultList(items) {
    const completionItems = [];
    for (let i = 0; i < items.length; i++) {
        completionItems.push({
            label: items[i],
            kind: CompletionItemKind.Field,
        });
    }
    return exclusiveResult(completionItems);
}
function exclusiveResultList(items) {
    const completionItems = [];
    for (let i = 0; i < items.length; i++) {
        completionItems.push({
            label: items[i],
            kind: CompletionItemKind.Field,
        });
    }
    return exclusiveResult(completionItems);
}
function exclusiveResult(items) {
    return {
        items: items,
        isExclusiveResult: true,
        analyzeDefaults: false,
    };
}
function nonExclusiveResult(items) {
    return {
        items: items,
        isExclusiveResult: false,
        analyzeDefaults: true,
    };
}
const EmptyCompletionResult = {
    items: [],
    isExclusiveResult: false,
    analyzeDefaults: true,
};
export { exclusiveResultList, resultList, dynamicParameter, EmptyCompletionResult, exclusiveResult, nonExclusiveResult };
