// td-antlers-parser/out/projects/blueprints/fields.js
const SetFieldTypes = ["replicator", "bard"];
export { SetFieldTypes };
function blueprintFieldFromScopeVariable(variable) {
    return {
        blueprintName: variable.sourceName,
        displayName: variable.name,
        instructionText: "",
        maxItems: -1,
        name: variable.name,
        refFieldSetField: null,
        type: variable.dataType,
        sets: null,
        import: null,
    };
}
export function variablesToBlueprintFields(variables) {
    const fields = [];
    for (let i = 0; i < variables.length; i++) {
        fields.push(blueprintFieldFromScopeVariable(variables[i]));
    }
    return fields;
}
export function blueprintFieldFromFieldSet(field, prefix = "") {
    const fieldToReturn = {
        blueprintName: "fieldset",
        displayName: prefix + field.name,
        instructionText: field.instructionText,
        maxItems: field.maxItems,
        name: prefix + field.handle,
        type: field.type,
        refFieldSetField: field,
        sets: field.sets,
        import: field.import,
    };
    return fieldToReturn;
}
export function adjustFieldType(fieldType, field, maxItems) {
    if (fieldType == "select") {
        if (typeof field.field.multiple !== "undefined" &&
            field.field.multiple === true) {
            fieldType = "select_multiple";
        }
    }
    if (typeof field.field.multiple === "undefined" || maxItems > 1) {
        if (fieldType == "form") {
            fieldType = "form_multiple";
        }
        else if (fieldType == "sites") {
            fieldType = "sites_multiple";
        }
        else if (fieldType == "taxonomies") {
            fieldType = "taxonomies_multiple";
        }
        else if (fieldType == "user_groups") {
            fieldType = "user_groups_multiple";
        }
        else if (fieldType == "user_roles") {
            fieldType = "user_roles_multiple";
        }
        else if (fieldType == "terms") {
            fieldType = "terms_multiple";
        }
        else if (fieldType == "users") {
            fieldType = "users_multiple";
        }
        else if (fieldType == "structures") {
            fieldType = "structures_multiple";
        }
    }
    return fieldType;
}
export function blueprintFieldsFromFieldset(fields) {
    const blueprintFields = [];
    for (let i = 0; i < fields.length; i++) {
        blueprintFields.push(blueprintFieldFromFieldSet(fields[i]));
    }
    return blueprintFields;
}
