// td-antlers-parser/out/antlers/permissions/permissionManager.js
import { NativePermissions } from "./nativePermissions.js";
class PermissionsManager {
    constructor() {
        this.contextItems = new Map();
        this.triggerNames = [];
        this.collectionTriggers = [];
        this.collectionTriggerCaps = [];
        this.globalTriggers = [];
        this.globalTriggerCaps = [];
        this.formTriggers = [];
        this.formTriggerCaps = [];
        this.structureTriggers = [];
        this.structureTriggerCaps = [];
        this.assetTriggers = [];
        this.assetTriggerCaps = [];
    }
    getCollectionTriggerCaps() {
        return this.collectionTriggerCaps;
    }
    getGlobalTriggerCaps() {
        return this.globalTriggerCaps;
    }
    getFormTriggerCaps() {
        return this.formTriggerCaps;
    }
    getStructureTriggerCaps() {
        return this.structureTriggerCaps;
    }
    getAssetTriggerCaps() {
        return this.assetTriggerCaps;
    }
    loadCorePermissions() {
        this.registerPermissions(NativePermissions);
    }
    getTriggerNames() {
        return this.triggerNames;
    }
    isCollectionTrigger(trigger) {
        return this.collectionTriggers.includes(trigger);
    }
    isGlobalTrigger(trigger) {
        return this.globalTriggers.includes(trigger);
    }
    isFormTrigger(trigger) {
        return this.formTriggers.includes(trigger);
    }
    isStructureTrigger(trigger) {
        return this.structureTriggers.includes(trigger);
    }
    isAssetTrigger(trigger) {
        return this.assetTriggers.includes(trigger);
    }
    getTriggerContextItems(trigger) {
        if (this.contextItems.has(trigger)) {
            return this.contextItems.get(trigger);
        }
        return [];
    }
    registerSpecialTrigger(trigger, specialTrigger, context) {
        if (specialTrigger == "{collection}") {
            if (this.collectionTriggerCaps.includes(context) == false) {
                this.collectionTriggerCaps.push(context);
            }
            if (this.collectionTriggers.includes(trigger) == false) {
                this.collectionTriggers.push(trigger);
            }
        }
        else if (specialTrigger == "{structure}") {
            if (this.structureTriggerCaps.includes(context) == false) {
                this.structureTriggerCaps.push(context);
            }
            if (this.structureTriggers.includes(trigger) == false) {
                this.structureTriggers.push(trigger);
            }
        }
        else if (specialTrigger == "{container}") {
            if (this.assetTriggerCaps.includes(context) == false) {
                this.assetTriggerCaps.push(context);
            }
            if (this.assetTriggers.includes(trigger) == false) {
                this.assetTriggers.push(trigger);
            }
        }
        else if (specialTrigger == "{global}") {
            if (this.globalTriggerCaps.includes(context) == false) {
                this.globalTriggerCaps.push(context);
            }
            if (this.globalTriggers.includes(trigger) == false) {
                this.globalTriggers.push(trigger);
            }
        }
        else if (specialTrigger == "{form}") {
            if (this.formTriggerCaps.includes(context) == false) {
                this.formTriggerCaps.push(context);
            }
            if (this.formTriggers.includes(trigger) == false) {
                this.formTriggers.push(trigger);
            }
        }
    }
    addContextToTrigger(trigger, context) {
        if (this.triggerNames.includes(trigger) == false) {
            this.triggerNames.push(trigger);
        }
        if (this.contextItems.has(trigger) == false) {
            this.contextItems.set(trigger, [context]);
        }
        else {
            this.contextItems.get(trigger)?.push(context);
        }
    }
    registerPermissions(permissions) {
        for (let i = 0; i < permissions.length; i++) {
            const thisPermission = permissions[i];
            let triggers = [], specialTrigger = "", contextItem = "";
            if (thisPermission.handle.includes("[") &&
                thisPermission.handle.includes("]")) {
                const openBracket = thisPermission.handle.indexOf("["), closeBracket = thisPermission.handle.indexOf("]"), len = closeBracket - openBracket, nestedTriggers = thisPermission.handle
                    .substr(openBracket + 1, len - 1)
                    .split(" ")
                    .filter((n) => n.trim().length > 0), additionalPieces = thisPermission.handle
                    .substr(closeBracket + 1)
                    .split(" ")
                    .filter((n) => n.trim().length > 0);
                triggers = nestedTriggers;
                if (additionalPieces.length == 2) {
                    specialTrigger = additionalPieces[0];
                    contextItem = additionalPieces[1];
                }
                else if (additionalPieces.length == 1) {
                    specialTrigger = "";
                    contextItem = additionalPieces[0];
                }
            }
            else {
                const pieces = thisPermission.handle
                    .split(" ")
                    .filter((n) => n.trim().length > 0);
                if (pieces.length == 2) {
                    triggers = [pieces[0]];
                    contextItem = pieces[1];
                }
            }
            for (let j = 0; j < triggers.length; j++) {
                const thisTrigger = triggers[j];
                if (specialTrigger.trim().length > 0) {
                    this.registerSpecialTrigger(thisTrigger, specialTrigger, contextItem);
                }
                else {
                    this.addContextToTrigger(thisTrigger, contextItem);
                }
            }
        }
    }
}
PermissionsManager.instance = null;
if (typeof PermissionsManager.instance == "undefined" ||
    PermissionsManager.instance == null) {
    PermissionsManager.instance = new PermissionsManager();
    PermissionsManager.instance.loadCorePermissions();
}
export default PermissionsManager;
