// td-antlers-parser/out/runtime/analyzers/tailwindClassAnalyzer.js
import { DocumentRange } from '../document/documentRange.js';
import { Position } from '../nodes/position.js';
export class TailwindClassAnalyzer {
    constructor(options = {}) {
        const names = options.attributeNames ?? ['class'];
        this.attributeNames = new Set(names.map((name) => name.toLowerCase()));
    }
    analyze(document) {
        const content = document.getOriginalContent();
        const lineStarts = this.getLineStarts(content);
        const candidates = [];
        this.findClassAttributes(content).forEach((attribute) => {
            this.tokenizeClassValue(content, attribute.valueStart, attribute.valueEnd).forEach((token) => {
                const range = this.rangeFromOffsets(lineStarts, token.start, token.end);
                candidates.push({
                    className: token.value,
                    attributeName: attribute.name,
                    containsAntlers: token.value.includes('{{') || token.value.includes('}}'),
                    range,
                    attributeRange: this.rangeFromOffsets(lineStarts, attribute.nameStart, attribute.valueEnd + 1)
                });
            });
        });
        return candidates;
    }
    findClassAttributes(content) {
        const attributes = [];
        const attributePattern = /([:@A-Za-z0-9_.-]+)\s*=\s*(["'])/g;
        let match = null;
        while ((match = attributePattern.exec(content)) != null) {
            const attributeName = match[1];
            if (!this.attributeNames.has(attributeName.toLowerCase())) {
                continue;
            }
            const quote = match[2], valueStart = (match.index ?? 0) + match[0].length, valueEnd = this.findQuotedValueEnd(content, valueStart, quote);
            if (valueEnd == -1) {
                continue;
            }
            attributes.push({
                name: attributeName,
                nameStart: match.index ?? 0,
                nameEnd: (match.index ?? 0) + attributeName.length,
                valueStart,
                valueEnd
            });
            attributePattern.lastIndex = valueEnd + 1;
        }
        return attributes;
    }
    findQuotedValueEnd(content, valueStart, quote) {
        for (let i = valueStart; i < content.length; i++) {
            if (content[i] == quote) {
                return i;
            }
        }
        return -1;
    }
    tokenizeClassValue(content, valueStart, valueEnd) {
        const tokens = [];
        let tokenStart = null, bracketDepth = 0, parenDepth = 0, antlersDepth = 0, stringQuote = null;
        for (let i = valueStart; i < valueEnd; i++) {
            const char = content[i], next = i + 1 < valueEnd ? content[i + 1] : null, prev = i > valueStart ? content[i - 1] : null;
            if (tokenStart == null && !this.isWhitespace(char)) {
                tokenStart = i;
            }
            if (tokenStart == null) {
                continue;
            }
            if (stringQuote != null) {
                if (char == stringQuote && prev != '\\') {
                    stringQuote = null;
                }
                continue;
            }
            if ((char == '"' || char == "'") && bracketDepth > 0) {
                stringQuote = char;
                continue;
            }
            if (char == '{' && next == '{') {
                antlersDepth += 1;
                i += 1;
                continue;
            }
            if (char == '}' && next == '}' && antlersDepth > 0) {
                antlersDepth -= 1;
                i += 1;
                continue;
            }
            if (antlersDepth == 0) {
                if (char == '[') {
                    bracketDepth += 1;
                }
                else if (char == ']' && bracketDepth > 0) {
                    bracketDepth -= 1;
                }
                else if (char == '(') {
                    parenDepth += 1;
                }
                else if (char == ')' && parenDepth > 0) {
                    parenDepth -= 1;
                }
            }
            if (this.isWhitespace(char) && bracketDepth == 0 && parenDepth == 0 && antlersDepth == 0) {
                this.pushToken(tokens, content, tokenStart, i);
                tokenStart = null;
            }
        }
        if (tokenStart != null) {
            this.pushToken(tokens, content, tokenStart, valueEnd);
        }
        return tokens;
    }
    pushToken(tokens, content, start, end) {
        const value = content.slice(start, end);
        if (value.trim().length == 0) {
            return;
        }
        tokens.push({ value, start, end });
    }
    isWhitespace(char) {
        return /\s/.test(char);
    }
    getLineStarts(content) {
        const lineStarts = [0];
        for (let i = 0; i < content.length; i++) {
            if (content[i] == '\n') {
                lineStarts.push(i + 1);
            }
        }
        return lineStarts;
    }
    rangeFromOffsets(lineStarts, start, end) {
        const range = new DocumentRange();
        range.start = this.positionFromOffset(lineStarts, start);
        range.end = this.positionFromOffset(lineStarts, end);
        return range;
    }
    positionFromOffset(lineStarts, offset) {
        let low = 0, high = lineStarts.length - 1, lineIndex = 0;
        while (low <= high) {
            const mid = Math.floor((low + high) / 2);
            if (lineStarts[mid] <= offset) {
                lineIndex = mid;
                low = mid + 1;
            }
            else {
                high = mid - 1;
            }
        }
        const position = new Position();
        position.index = offset;
        position.offset = offset;
        position.line = lineIndex + 1;
        position.char = offset - lineStarts[lineIndex] + 1;
        return position;
    }
}
