// td-antlers-parser/out/runtime/document/frontMatter/frontMatterParser.js
import YAML from "yaml";
import { Scope } from '../../../antlers/scope/scope.js';
import { Pair, Scalar, YAMLMap, YAMLSeq } from 'yaml/types';
export class FrontMatterParser {
    constructor(projectProvider) {
        this.documentScope = null;
        this.project = projectProvider;
    }
    getKeyValue(node) {
        if (node instanceof Scalar) {
            return node.value;
        }
        return '';
    }
    getNodeValue(node) {
        if (node instanceof Scalar) {
            return node.value;
        }
        else if (node instanceof YAMLMap) {
            return this.analyzeDocument(node.items);
        }
        else if (node instanceof YAMLSeq) {
            return this.analyzeDocument(node.items);
        }
        return '';
    }
    getScalarRuntimeType(value) {
        if (parseFloat(value).toString() == value || parseInt(value).toString() == value) {
            return 'number';
        }
        return 'string';
    }
    analyzeDocument(items) {
        const nestedScope = new Scope(this.project);
        items.forEach((item) => {
            if (item instanceof Pair) {
                const varName = this.getKeyValue(item.key), varValue = this.getNodeValue(item.value);
                if (item.value instanceof Scalar) {
                    nestedScope.addVariable({
                        dataType: this.getScalarRuntimeType(varValue),
                        introducedBy: null,
                        name: varName,
                        sourceField: null,
                        sourceName: '*frontmatter'
                    });
                }
                else if (item.value instanceof YAMLSeq && item.key instanceof Scalar) {
                    const sequenceVariables = new Scope(this.project);
                    item.value.items.forEach((seqValue) => {
                        const sValue = this.getNodeValue(seqValue);
                        sequenceVariables.addVariable({
                            dataType: this.getScalarRuntimeType(sValue),
                            introducedBy: null,
                            name: seqValue.value,
                            sourceField: null,
                            sourceName: '*frontmatter'
                        });
                    });
                    nestedScope.addScopeList(item.key.value, sequenceVariables);
                }
                else {
                    nestedScope.addScopeList(varName, varValue);
                }
            }
        });
        return nestedScope;
    }
    parse(text) {
        try {
            const docs = YAML.parseDocument(text);
            if (docs.contents instanceof YAMLMap) {
                this.documentScope = this.analyzeDocument(docs.contents.items);
            }
        }
        catch (err) {
            // Prevent YAML errors from breaking the extension.
            this.documentScope = null;
        }
    }
    hasScope() {
        if (this.documentScope == null) {
            return false;
        }
        return true;
    }
    getScope() {
        return this.documentScope;
    }
}
