// td-antlers-parser/out/runtime/nodes/abstractNode.js
import { Position } from './position.js';
import { DocumentRange } from '../document/documentRange.js';
import { NodeVirtualModifiers } from './antlersVirtualStructures/nodeVirtualModifiers.js';
import { filterStructuralAntlersNodes } from '../document/scanners/nodeFilters.js';
import { StringUtilities } from '../utilities/stringUtilities.js';
import { NodeVirtualHierarchy } from './antlersVirtualStructures/nodeVirtualHierarchy.js';
import { ConditionPairAnalyzer } from '../analyzers/conditionPairAnalyzer.js';
import { replaceAllInString } from '../../utils/strings.js';
import { getId } from '../../utils/simpleIds.js';
function newRefId() {
    return replaceAllInString(getId(), '-', '_');
}
/**
 * Redefine an already-initialised property as non-enumerable.
 * Call this in the constructor *after* field initialisers have run so that
 * naive traversal code using Object.keys / Object.entries won't encounter
 * back-reference cycles (parent, prev, next, isOpenedBy, etc.).
 */
function defineBackRef(obj, prop) {
    Object.defineProperty(obj, prop, {
        value: obj[prop],
        writable: true,
        enumerable: false,
        configurable: true,
    });
}
export var FragmentPosition;
(function (FragmentPosition) {
    FragmentPosition[FragmentPosition["IsDynamicFragmentName"] = 0] = "IsDynamicFragmentName";
    FragmentPosition[FragmentPosition["InsideFragmentParameter"] = 1] = "InsideFragmentParameter";
    FragmentPosition[FragmentPosition["InsideFragment"] = 2] = "InsideFragment";
    FragmentPosition[FragmentPosition["Unresolved"] = 3] = "Unresolved";
})(FragmentPosition || (FragmentPosition = {}));
export class FragmentNode {
    constructor() {
        this.startPosition = null;
        this.endPosition = null;
        this.index = 0;
        this.embeddedIndex = 0;
        this.refId = null;
        this.parameters = [];
        this.isSelfClosing = false;
        this.isClosingFragment = false;
        this.name = '';
        this.containsStructures = false;
        this.refId = newRefId();
    }
}
export class FragmentParameterNode {
    constructor() {
        this.startPosition = null;
        this.endPosition = null;
        this.content = '';
        this.name = '';
    }
}
function cleanNodes(nodes) {
    const cleaned = [], ids = [];
    for (let i = 0; i < nodes.length; i++) {
        const child = nodes[i], refId = child.refId;
        if (ids.includes(refId)) {
            break;
        }
        cleaned.push(child);
        ids.push(refId);
    }
    return cleaned;
}
export class AbstractNode {
    isEmbedded() {
        return this.isInScriptTag || this.isInStyleTag;
    }
    getContexts() {
        return this.positionContexts;
    }
    addContext(context) {
        if (this.addedContexts.has(context.refId) == false) {
            this.addedContexts.set(context.refId, true);
            this.positionContexts.push(context);
        }
    }
    getRange() {
        if (this.originalAbstractNode != null) {
            return this.originalAbstractNode.getRange();
        }
        if (this.startPosition == null || this.endPosition == null) {
            return DocumentRange.Empty;
        }
        const range = new DocumentRange();
        range.start = this.startPosition;
        range.end = this.endPosition;
        return range;
    }
    withParser(parser) {
        this.parser = parser;
    }
    getParser() {
        return this.parser;
    }
    pushError(error) {
        if (this.parent != null) {
            this.parent.pushError(error);
        }
        this.antlersErrors.push(error);
        if (this.parent == null) {
            if (this.parser != null) {
                this.parser.pushError(error);
            }
        }
    }
    mergeErrors(errors) {
        errors.forEach((error) => {
            if (error.node == null) {
                error.node = this;
            }
            this.antlersErrors.push(error);
        });
    }
    getErrors() {
        return this.antlersErrors;
    }
    constructor() {
        this.refId = null;
        this.index = 0;
        this.parent = null;
        this.content = '';
        this.rawLexContent = '';
        this.sourceContent = '';
        this.startPosition = null;
        this.endPosition = null;
        this.modifierChain = null;
        this.originalAbstractNode = null;
        this.positionContexts = [];
        this.addedContexts = new Map();
        this.fragment = null;
        this.fragmentParameter = null;
        this.fragmentPosition = FragmentPosition.Unresolved;
        this.containsAnyFragments = false;
        this.containsChildStructures = false;
        this.isInScriptTag = false;
        this.isInStyleTag = false;
        /* Start: Internal Parser Variables and APIs. */
        this.isVirtual = true;
        this.isPartOfMethodChain = false;
        this.isVirtualGroupMember = false;
        this.isVirtualGroupOperatorResolve = false;
        this.isSwitchGroupMember = false;
        this.isListGroupMember = false;
        this.producesVirtualStatementTerminator = false;
        this.convertedToOperator = false;
        this.scopeVariable = null;
        this.currentScope = null;
        this.manifestType = '';
        this.scopeName = null;
        this.sourceType = '';
        this.runtimeType = null;
        this.modifiers = new NodeVirtualModifiers(this);
        /* End: Internal Parser Variables and APIs. */
        this.prev = null;
        this.next = null;
        this.methodTarget = null;
        this.libraryTarget = null;
        this.antlersErrors = [];
        this.parser = null;
        this.refId = newRefId();
        // Back-references: non-enumerable to prevent cycles in naive traversal.
        defineBackRef(this, 'parent');
        defineBackRef(this, 'prev');
        defineBackRef(this, 'next');
        defineBackRef(this, 'parser');
        defineBackRef(this, 'originalAbstractNode');
    }
    /**
     * Returns this node's direct structural child nodes in document order.
     *
     * Use node.getChildNodes() for tree traversal.  Properties like parent,
     * isClosedBy, prev, next, etc. are back-references and may form cycles.
     *
     * Each subclass overrides this to return the children it owns.  The base
     * implementation returns an empty array (leaf node).
     */
    getChildNodes() {
        return [];
    }
    innerContent() {
        return this.content;
    }
    rawContent() {
        return this.content;
    }
}
export class ExecutionBranch extends AbstractNode {
    constructor() {
        super(...arguments);
        this.head = null;
        this.tail = null;
        this.nodes = [];
        this.documentText = '';
        this.childDocument = null;
    }
    getImmediateChildren() {
        if (this.head == null) {
            return [];
        }
        return this.head.getImmediateChildren();
    }
    getChildNodes() {
        return this.nodes;
    }
}
export class ConditionNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.logicBranches = [];
        this.chain = [];
        this.nodeContent = '';
        this.fragment = null;
        this.fragmentParameter = null;
        this.fragmentPosition = FragmentPosition.Unresolved;
        this.containsAnyFragments = false;
        this.containsChildStructures = false;
    }
    getChildNodes() {
        return this.logicBranches;
    }
}
export class AntlersNode extends AbstractNode {
    constructor() {
        super();
        this.isNodeAbanonded = false;
        this.isComment = false;
        this.isTagNode = false;
        this.isConditionNode = false;
        /**
         * Indicates whether the tag is required to have a closing pair.
         *
         * Populated from the TagManager metadata for known tags. Linters and
         * downstream consumers can use this to avoid flagging tags like
         * {{ partial:... }} that are allowed to be used standalone.
         */
        this.requiresClose = false;
        /**
         * Indicates whether the tag optionally allows a closing pair.
         *
         * Populated from the TagManager metadata for known tags. Tags like
         * partial, slot, and trans can be used either standalone or as a
         * tag pair.
         */
        this.allowsContentClose = false;
        this.runtimeContent = "";
        this.name = null;
        this.pathReference = null;
        this.isClosingTag = false;
        this.isInterpolationNode = false;
        this.isOpenedBy = null;
        this.isClosedBy = null;
        this.isSelfClosing = false;
        this.children = [];
        this.runtimeNodes = [];
        this.parsedRuntimeNodes = [];
        this.hasParsedRuntimeNodes = false;
        this.parameters = [];
        this.hasParameters = false;
        this.contentOffset = null;
        this.cachedContent = null;
        this.cachedInnerContent = null;
        this.hasRecursiveNode = false;
        this.recursiveReference = null;
        this.rawStart = "";
        this.rawEnd = "";
        this.originalNode = null;
        this.interpolationRegions = new Map();
        this.processedInterpolationRegions = new Map();
        this.hasProcessedInterpolationRegions = false;
        this.ref = 0;
        this.contentStartRelativeIndex = null;
        this.antlersNodeIndex = 0;
        this.nameStartsOn = null;
        this.nameEndsOn = null;
        this.nameMethodPartStartsOn = null;
        this.documentText = '';
        this.nodeContent = '';
        this.childDocument = null;
        this.structure = new NodeVirtualHierarchy(this);
        this.isInlineAntlers = false;
        /** Start: Internal Parser State Variables */
        this.mustClose = false;
        this.reference = null;
        this._conditionParserAbandonPairing = false;
        this._isEndVirtual = false;
        defineBackRef(this, 'isOpenedBy');
        defineBackRef(this, 'isClosedBy');
        defineBackRef(this, 'originalNode');
    }
    getChildNodes() {
        if (this.isClosedBy != null) {
            return this.children.slice(0, -1);
        }
        return this.children;
    }
    getOriginalContent() {
        if (this.startPosition == null || this.endPosition == null || this.parser == null) {
            return '';
        }
        if (this.isClosedBy != null) {
            if (this.isClosedBy.startPosition == null || this.isClosedBy.endPosition == null) {
                return '';
            }
            return this.parser.getText(this.startPosition.index, this.isClosedBy.endPosition.index + 1);
        }
        return this.parser.getText(this.startPosition.index, this.endPosition.index + 1);
    }
    getNodeDocumentText() {
        if (this.startPosition == null || this.endPosition == null || this.parser == null) {
            return '';
        }
        return this.parser.getText(this.startPosition.index, this.endPosition.index + 1);
    }
    getImmediateChildren() {
        const immediateChildren = [];
        for (let i = 0; i < this.children.length; i++) {
            const node = this.children[i];
            if (node.parent == this) {
                if (node == this.isClosedBy) {
                    break;
                }
                immediateChildren.push(node);
            }
        }
        return cleanNodes(immediateChildren);
    }
    getChildren() {
        if (this.isClosedBy == null) {
            return [];
        }
        const newChildren = [];
        for (let i = 0; i < this.children.length; i++) {
            const child = this.children[i];
            if (child instanceof AntlersNode && child.refId == this.isClosedBy.refId) {
                break;
            }
            newChildren.push(child);
        }
        return newChildren;
    }
    getInnerDocumentText() {
        if (this.startPosition == null || this.endPosition == null || this.parser == null) {
            return '';
        }
        if (this.isClosedBy == null) {
            return this.getNodeDocumentText();
        }
        const closeStart = this.isClosedBy.startPosition;
        if (closeStart == null) {
            return '';
        }
        return this.parser.getText(this.endPosition.index + 1, closeStart.index);
    }
    nodeAtIndex(index) {
        if (this.runtimeNodes.length == 0) {
            return null;
        }
        for (let i = 0; i < this.runtimeNodes.length; i++) {
            if (this.runtimeNodes[i].index == index) {
                return this.runtimeNodes[i];
            }
        }
        return null;
    }
    getDepthCount() {
        if (this.parent == null) {
            return 1;
        }
        let parent = this.parent, depth = 1;
        while (parent != null) {
            depth += 1;
            if (parent.parent === parent) {
                parent = null;
                break;
            }
            parent = parent.parent;
        }
        return depth;
    }
    isEmpty() {
        return this.content.trim().length == 0;
    }
    nameMatches(content) {
        if (this.name == null) {
            return false;
        }
        return this.name.content == content;
    }
    copyBasicDetails() {
        return this.copyBasicDetailsTo(new AntlersNode());
    }
    copyBasicDetailsTo(instance) {
        instance.refId = this.refId;
        instance.isComment = this.isComment;
        instance.isTagNode = this.isTagNode;
        instance.children = this.children;
        instance.parent = this.parent;
        instance.parameters = this.parameters;
        instance.isClosingTag = this.isClosingTag;
        instance.rawStart = this.rawStart;
        instance.rawEnd = this.rawEnd;
        instance.startPosition = this.startPosition;
        instance.endPosition = this.endPosition;
        instance.originalAbstractNode = this;
        instance.containsAnyFragments = this.containsAnyFragments;
        instance.containsChildStructures = this.containsChildStructures;
        instance.isInScriptTag = this.isInScriptTag;
        instance.isInStyleTag = this.isInStyleTag;
        return instance;
    }
    id() {
        return this.refId;
    }
    innerContent() {
        if (this.cachedInnerContent == null) {
            this.cachedInnerContent = this.content.replace('"', '\\"');
        }
        return this.cachedInnerContent;
    }
    resetContentCache() {
        this.cachedContent = null;
        this.cachedInnerContent = null;
    }
    hasMethodPart() {
        if (this.name == null) {
            return false;
        }
        const methodPart = this.name.getMethodName();
        if (methodPart != null) {
            return true;
        }
        return false;
    }
    runtimeName() {
        if (this.name == null) {
            return "";
        }
        return this.name.getCompoundTagName();
    }
    getTagName() {
        if (this.name == null) {
            return "";
        }
        return this.name.name;
    }
    getNameContent() {
        if (this.name == null) {
            return "";
        }
        return this.name.content;
    }
    getMethodNameValue() {
        if (this.name == null) {
            return "";
        }
        const methodPart = this.name.getMethodName();
        if (methodPart != null) {
            return methodPart;
        }
        return "";
    }
    methodIsEmptyOrMatches(testValue) {
        return this.getMethodNameValue() == testValue;
    }
    hasParameter(name) {
        const paramMatch = this.findParameter(name);
        return paramMatch != null;
    }
    /**
     * Tests if the AntlersNode contains a parameter with the given value.
     *
     * @param name
     * @param expectedValue
     * @returns
     */
    hasParameterWithValue(name, expectedValue) {
        const paramMatch = this.findParameter(name);
        if (paramMatch == null) {
            return false;
        }
        return paramMatch.value == expectedValue;
    }
    findParameterValue(name, defaultValue) {
        const paramMatch = this.findParameter(name);
        if (paramMatch == null) {
            return defaultValue;
        }
        return paramMatch.value;
    }
    findParameterValueOrNull(name) {
        const paramMatch = this.findParameter(name);
        if (paramMatch == null) {
            return null;
        }
        return paramMatch.value;
    }
    findParameter(name) {
        if (this.parameters.length == 0) {
            return null;
        }
        for (let i = 0; i < this.parameters.length; i++) {
            if (this.parameters[i].name == name) {
                return this.parameters[i];
            }
        }
        return null;
    }
    findAnyParameter(names) {
        for (let i = 0; i < names.length; i++) {
            const tempValue = this.findParameter(names[i]);
            if (tempValue != null) {
                return tempValue;
            }
        }
        return null;
    }
    getContentRelativeStartIndex() {
        if (this.contentStartRelativeIndex == null) {
            const chars = this.content.split("");
            for (let i = 0; i < chars.length; i++) {
                if (!StringUtilities.ctypeSpace(chars[i])) {
                    this.contentStartRelativeIndex = i;
                    break;
                }
            }
            if (this.contentStartRelativeIndex == null) {
                this.contentStartRelativeIndex = chars.length;
            }
            this.contentStartRelativeIndex += this.rawStart.length;
        }
        return this.contentStartRelativeIndex;
    }
    getContentAfterName() {
        let start = 0;
        if (this.nameEndsOn != null) {
            start = this.nameEndsOn.char - this.rawStart.length - 1;
        }
        return this.content.substring(start);
    }
    getContent() {
        if (this.cachedContent == null) {
            if (this.isComment) {
                this.cachedContent = this.content;
                this.contentOffset = this.startPosition;
            }
            else {
                if (this.isTagNode) {
                    let contentToAnalyze = this.content;
                    contentToAnalyze = StringUtilities.trimLeft(contentToAnalyze);
                    let compoundNameLen = 0;
                    if (this.name != null) {
                        compoundNameLen = this.name.compound.length;
                    }
                    if (this.parameters.length > 0) {
                        let startPosIndex = 0;
                        const firstParam = this.parameters[0];
                        if (firstParam.startPosition != null) {
                            startPosIndex = firstParam.startPosition.index - 1;
                        }
                        const leadContent = contentToAnalyze, contentWithoutSpace = leadContent.trimLeft(), leadingWsCount = leadContent.length - contentWithoutSpace.length, leadNameLen = compoundNameLen;
                        // Get the content before the first parameter.
                        contentToAnalyze = contentToAnalyze.substr(0, startPosIndex);
                    }
                    const contentWithoutSpace = contentToAnalyze.trimLeft(), leadingWsCount = contentToAnalyze.length - contentWithoutSpace.length, leadOffset = compoundNameLen + leadingWsCount + 2;
                    if (this.parser != null) {
                        let startPosOffset = 0;
                        if (this.startPosition != null) {
                            startPosOffset = this.startPosition.offset;
                        }
                        this.contentOffset ==
                            this.parser.positionFromOffset(startPosOffset + leadOffset, 0);
                    }
                    if (this.name != null) {
                        if (this.name.name == "if" ||
                            this.name.name == "elseif" ||
                            this.name.name == "unless" ||
                            this.name.name == "elseunless") {
                            contentToAnalyze = " " + contentToAnalyze.trimLeft();
                            this.cachedContent = contentToAnalyze.substr(this.name.compound.length + 1);
                        }
                        else {
                            this.cachedContent = contentToAnalyze;
                        }
                    }
                    else {
                        this.cachedContent = contentToAnalyze;
                    }
                }
                else {
                    if (this.parameters.length > 0) {
                        const firstParam = this.parameters[0];
                        let startPosIndex = 0;
                        if (firstParam.startPosition != null) {
                            startPosIndex = firstParam.startPosition.index - 1;
                        }
                        this.contentOffset = this.startPosition;
                        this.cachedContent = this.content.substr(0, startPosIndex);
                    }
                    else {
                        this.contentOffset = this.startPosition;
                        this.cachedContent = this.content;
                    }
                }
            }
        }
        return this.cachedContent;
    }
    relativePositionFromOffset(offset, index) {
        if (this.parser == null) {
            return null;
        }
        return this.parser.positionFromOffset((this.contentOffset?.offset ?? 0) + offset, index);
    }
    getInterpolationNode(varName) {
        if (this.processedInterpolationRegions.has(varName)) {
            const refRegion = this.processedInterpolationRegions.get(varName);
            if (refRegion.length > 0 && refRegion[0] instanceof AntlersNode) {
                return refRegion[0];
            }
        }
        return null;
    }
    _lexerRelativeOffset(offset, index = null) {
        let indexToUse = offset;
        if (index != null) {
            indexToUse = index;
        }
        if (this.parser == null) {
            const position = new Position();
            position.index = offset;
            position.offset = offset;
            return position;
        }
        let relativeIndex = offset + this.rawStart.length + (this.startPosition?.index ?? 0);
        if (this.isTagNode && this.nameStartsOn != null) {
            relativeIndex = this.nameStartsOn.index + offset;
            if (!ConditionPairAnalyzer.isConditionalStructure(this)) {
                relativeIndex -= this.rawStart.length;
            }
            if (this.name?.name == 'unless') {
                relativeIndex += 6;
            }
        }
        const resolvedOffset = this.parser.positionFromOffset(relativeIndex, relativeIndex, true);
        return resolvedOffset;
    }
    relativeOffset(offset, index = null) {
        let indexToUse = offset;
        if (index != null) {
            indexToUse = index;
        }
        if (this.parser == null) {
            const position = new Position();
            position.index = offset;
            position.offset = offset;
            return position;
        }
        const relativeIndex = offset + this.rawStart.length + (this.startPosition?.index ?? 0);
        const resolvedOffset = this.parser.positionFromOffset(relativeIndex, relativeIndex, true);
        return resolvedOffset;
    }
    isPaired() {
        if (this.isClosingTag && this.isSelfClosing == false && this.isOpenedBy != null) {
            return true;
        }
        if (this.isClosedBy == null) {
            return false;
        }
        if (this.isSelfClosing) {
            return false;
        }
        return true;
    }
    getFinalClosingTag() {
        if (this.isClosedBy == null) {
            return this;
        }
        return this.isClosedBy.getFinalClosingTag();
    }
    rawContent() {
        return this.rawStart + this.content + this.rawEnd;
    }
    getTrueRuntimeNodes() {
        if (this.originalNode != null && this.originalNode != this) {
            return this.originalNode.getTrueRuntimeNodes();
        }
        return this.runtimeNodes;
    }
    getTrueRawContent() {
        if (this.originalNode != null && this.originalNode != this) {
            return this.originalNode.getTrueRawContent();
        }
        return this.rawContent();
    }
    getTrueNode() {
        if (this.originalNode != null && this.originalNode != this) {
            return this.originalNode.getTrueNode();
        }
        return this;
    }
    getStructuralChildren() {
        if (this.isClosedBy != null) {
            // Nodes with children will always have the
            // closing node added as the last child.
            return filterStructuralAntlersNodes(this.children.slice(0, -1));
        }
        return filterStructuralAntlersNodes(this.children);
    }
    findParentOfType(types) {
        if (this.parent == null) {
            return null;
        }
        if (this.parent instanceof AntlersNode) {
            if (this.parent.name != null && types.includes(this.parent.getTagName())) {
                return this.parent;
            }
            return this.parent.findParentOfType(types);
        }
        return null;
    }
    /**
     * Returns the partial view reference for a {{ partial }} tag.
     *
     * Handles both the method-part form ({{ partial:components/image }})
     * and the src parameter form ({{ partial src="components/image" }}).
     * Returns null for non-partial tags or when no reference is present.
     *
     * The returned value preserves any package namespace
     * (e.g. "statamic-peak-seo::errors/redirects").
     */
    getPartialViewReference() {
        if (this.getTagName() != 'partial') {
            return null;
        }
        if (this.hasMethodPart()) {
            const methodValue = this.getMethodNameValue();
            if (methodValue.length > 0) {
                return methodValue;
            }
        }
        const srcParam = this.findParameter('src');
        if (srcParam != null && srcParam.value.length > 0) {
            return srcParam.value;
        }
        return null;
    }
    /**
     * Returns the package namespace for a namespaced partial reference,
     * or null when the partial is project-local.
     *
     * For example, {{ partial:statamic-peak-seo::errors/redirects }}
     * resolves to "statamic-peak-seo".
     */
    getPartialNamespace() {
        const reference = this.getPartialViewReference();
        if (reference == null) {
            return null;
        }
        const separatorIndex = reference.indexOf('::');
        if (separatorIndex <= 0) {
            return null;
        }
        return reference.substring(0, separatorIndex);
    }
    /**
     * Returns the view path for a partial reference with any package
     * namespace stripped away.
     *
     * For example, {{ partial:statamic-peak-seo::errors/redirects }}
     * resolves to "errors/redirects", and
     * {{ partial:components/image }} resolves to "components/image".
     */
    getPartialViewName() {
        const reference = this.getPartialViewReference();
        if (reference == null) {
            return null;
        }
        const separatorIndex = reference.indexOf('::');
        if (separatorIndex < 0) {
            return reference;
        }
        return reference.substring(separatorIndex + 2);
    }
    /**
     * Indicates whether the partial reference targets a package namespace
     * (e.g. a tag like {{ partial:statamic-peak-seo::errors/redirects }}).
     *
     * Returns false for non-partial tags and project-local partials.
     */
    isNamespacedPartial() {
        return this.getPartialNamespace() != null;
    }
}
export class ParserFailNode extends AntlersNode {
    static makeWithStartPosition(startPosition) {
        const newNode = new ParserFailNode();
        newNode.startPosition = startPosition;
        newNode.endPosition = startPosition;
        return newNode;
    }
}
export class AntlersParserFailNode extends ParserFailNode {
}
export class RecursiveNode extends AntlersNode {
    constructor() {
        super();
        this.recursiveParent = null;
        this.isNestedRecursive = false;
        defineBackRef(this, 'recursiveParent');
    }
}
export class PhpExecutionNode extends AntlersNode {
    constructor() {
        super(...arguments);
        this.isEchoNode = false;
    }
}
export class CommentParserFailNode extends ParserFailNode {
}
export class PhpParserFailNode extends ParserFailNode {
}
export class EscapedContentNode extends AntlersNode {
    constructor() {
        super(...arguments);
        this.isNoParseRegion = false;
    }
}
export class ParameterNode extends AbstractNode {
    constructor() {
        super();
        this.modifier = null;
        this.isModifierParameter = false;
        this.nameDelimiter = '"';
        this.isVariableReference = false;
        this.name = "";
        this.value = "";
        this.containsEscapedContent = false;
        this.interpolations = [];
        this.parent = null;
        this.blockPosition = null;
        this.namePosition = null;
        this.valuePosition = null;
        this.hasValidValueDelimiter = true;
        // Re-declared parent field overwrites the non-enumerable definition
        // from AbstractNode, so we need to redefine it here.
        defineBackRef(this, 'parent');
    }
    containsSimpleValue() {
        return (this.hasInterpolations() == false && this.isVariableReference == false);
    }
    hasInterpolations() {
        return this.interpolations.length > 0;
    }
    getArrayValue() {
        return this.value.split("|");
    }
    getValue() {
        return this.value;
    }
}
export class ModifierChainNode {
    constructor() {
        this.modifierTarget = null;
        this.modifierChain = [];
        this.startPosition = null;
        this.endPosition = null;
        this.lastManifestedModifier = null;
        this.lastModifier = null;
    }
    updateValues() {
        if (this.modifierChain.length == 0) {
            this.startPosition = null;
            this.endPosition = null;
        }
        else {
            const firstModifier = this.modifierChain[0], lastModifier = this.modifierChain[this.modifierChain.length - 1];
            this.startPosition = firstModifier.startPosition;
            this.endPosition = lastModifier.endPosition;
        }
    }
}
export class ModifierNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.modifier = null;
        this.methodStyleArguments = null;
        this.nameNode = null;
        this.name = '';
        this.valueNodes = [];
        this.parameters = [];
    }
    getChildNodes() {
        const children = [];
        if (this.nameNode != null)
            children.push(this.nameNode);
        children.push(...this.valueNodes);
        children.push(...this.parameters);
        if (this.methodStyleArguments != null)
            children.push(this.methodStyleArguments);
        return children;
    }
    getParameterValues() {
        const values = [];
        this.parameters.forEach((param) => {
            values.push(param.value);
        });
        return values;
    }
}
export class ValueDirectionNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.order = 0;
        this.name = null;
        this.directionNode = null;
    }
    getChildNodes() {
        const children = [];
        if (this.name != null)
            children.push(this.name);
        if (this.directionNode != null)
            children.push(this.directionNode);
        return children;
    }
}
export class ModifierParameterNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.value = '';
    }
}
export class FalseConstant extends AbstractNode {
}
export class NullConstant extends AbstractNode {
}
export class TrueConstant extends AbstractNode {
}
export class AdditionOperator extends AbstractNode {
}
export class DivisionOperator extends AbstractNode {
}
export class ExponentiationOperator extends AbstractNode {
}
export class FactorialOperator extends AbstractNode {
    constructor() {
        super(...arguments);
        this.repeat = 1;
    }
}
export class ModulusOperator extends AbstractNode {
}
export class MultiplicationOperator extends AbstractNode {
}
export class SubtractionOperator extends AbstractNode {
}
export class AdditionAssignmentOperator extends AbstractNode {
}
export class DivisionAssignmentOperator extends AbstractNode {
}
export class LeftAssignmentOperator extends AbstractNode {
}
export class ModulusAssignmentOperator extends AbstractNode {
}
export class MultiplicationAssignmentOperator extends AbstractNode {
}
export class SubtractionAssignmentOperator extends AbstractNode {
}
export class EqualCompOperator extends AbstractNode {
}
export class GreaterThanCompOperator extends AbstractNode {
}
export class GreaterThanEqualCompOperator extends AbstractNode {
}
export class LessThanCompOperator extends AbstractNode {
}
export class LessThanEqualCompOperator extends AbstractNode {
}
export class NotEqualCompOperator extends AbstractNode {
}
export class NotStrictEqualCompOperator extends AbstractNode {
}
export class SpaceshipCompOperator extends AbstractNode {
}
export class StrictEqualCompOperator extends AbstractNode {
}
export class ConditionalVariableFallbackOperator extends AbstractNode {
}
export class LanguageOperatorConstruct extends AbstractNode {
}
export class StatementSeparatorNode extends AbstractNode {
}
export class DirectionGroup extends AbstractNode {
    constructor() {
        super(...arguments);
        this.orderClauses = [];
    }
    getChildNodes() {
        return this.orderClauses;
    }
}
export class InlineBranchSeparator extends AbstractNode {
}
export class ModifierSeparator extends AbstractNode {
}
export class NullCoalescenceGroup extends AbstractNode {
    constructor() {
        super(...arguments);
        this.left = null;
        this.right = null;
    }
    getChildNodes() {
        const children = [];
        if (this.left != null)
            children.push(this.left);
        if (this.right != null)
            children.push(this.right);
        return children;
    }
}
export class ModifierValueSeparator extends AbstractNode {
    constructor() {
        super(...arguments);
        this.isTenaryBranchSeparator = false;
    }
}
export class InlineTernarySeparator extends AbstractNode {
}
export class LogicalAndOperator extends AbstractNode {
}
export class LogicalNegationOperator extends AbstractNode {
}
export class LogicalOrOperator extends AbstractNode {
}
export class LogicalXorOperator extends AbstractNode {
}
export class NullCoalesceOperator extends AbstractNode {
}
export class ScopeAssignmentOperator extends AbstractNode {
}
export class StringConcatenationOperator extends AbstractNode {
}
export class LogicGroupEnd extends AbstractNode {
}
export class LogicGroupBegin extends AbstractNode {
}
export class ListValueNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.values = [];
        this.isNamedNode = false;
        this.parsedName = null;
    }
    getChildNodes() {
        return this.values;
    }
}
export class LogicGroup extends AbstractNode {
    constructor() {
        super(...arguments);
        this.start = null;
        this.end = null;
        this.nodes = [];
        this.scopeOperator = null;
    }
    getChildNodes() {
        const children = [];
        if (this.start != null)
            children.push(this.start);
        children.push(...this.nodes);
        if (this.scopeOperator != null)
            children.push(this.scopeOperator);
        if (this.end != null)
            children.push(this.end);
        return children;
    }
}
export class SwitchCase extends AbstractNode {
    constructor() {
        super(...arguments);
        this.condition = null;
        this.expression = null;
    }
    getChildNodes() {
        const children = [];
        if (this.condition != null)
            children.push(this.condition);
        if (this.expression != null)
            children.push(this.expression);
        return children;
    }
}
export class SwitchGroup extends AbstractNode {
    constructor() {
        super(...arguments);
        this.cases = [];
    }
    getChildNodes() {
        return this.cases;
    }
}
export class ConditionalFallbackGroup extends AbstractNode {
    constructor() {
        super(...arguments);
        this.left = null;
        this.right = null;
    }
    getChildNodes() {
        const children = [];
        if (this.left != null)
            children.push(this.left);
        if (this.right != null)
            children.push(this.right);
        return children;
    }
}
export class ArrayNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.nodes = [];
    }
    getChildNodes() {
        return this.nodes;
    }
}
export class TernaryCondition extends AbstractNode {
    constructor() {
        super(...arguments);
        this.head = null;
        this.truthBranch = null;
        this.falseBranch = null;
    }
    getChildNodes() {
        const children = [];
        if (this.head != null)
            children.push(this.head);
        if (this.truthBranch != null)
            children.push(this.truthBranch);
        if (this.falseBranch != null)
            children.push(this.falseBranch);
        return children;
    }
}
export class ArgSeparator extends AbstractNode {
}
export class SemanticGroup extends AbstractNode {
    constructor() {
        super(...arguments);
        this.nodes = [];
        this.separatorToken = null;
    }
    getChildNodes() {
        if (this.separatorToken != null) {
            return [...this.nodes, this.separatorToken];
        }
        return this.nodes;
    }
}
export class ScopedLogicGroup extends LogicGroup {
    constructor() {
        super(...arguments);
        this.scope = null;
    }
    getChildNodes() {
        const children = [];
        if (this.scope != null)
            children.push(this.scope);
        children.push(...super.getChildNodes());
        return children;
    }
    extract() {
        let scopeName = null;
        if (this.scope != null) {
            scopeName = this.scope.name;
        }
        const sematicWrapper = this.nodes[0];
        return [
            scopeName,
            sematicWrapper.nodes
        ];
    }
}
export class AliasedScopeLogicGroup extends ScopedLogicGroup {
    constructor() {
        super(...arguments);
        this.alias = null;
    }
    getChildNodes() {
        const children = super.getChildNodes();
        if (this.alias != null)
            children.push(this.alias);
        return children;
    }
}
export class ArgumentGroup extends AbstractNode {
    constructor() {
        super(...arguments);
        this.args = [];
        this.hasNamedArguments = false;
        this.numberOfNamedArguments = 0;
    }
    getChildNodes() {
        return this.args;
    }
}
export class TupleListStart extends AbstractNode {
}
export class TupleList extends AbstractNode {
}
export class TupleScopedLogicGroup extends LogicGroup {
    constructor() {
        super(...arguments);
        this.target = null;
        this.name = null;
        this.item1 = null;
        this.item2 = null;
        this.isDynamicNames = false;
    }
    getChildNodes() {
        const children = [];
        if (this.target instanceof AbstractNode)
            children.push(this.target);
        if (this.name instanceof AbstractNode)
            children.push(this.name);
        if (this.item1 instanceof AbstractNode)
            children.push(this.item1);
        if (this.item2 instanceof AbstractNode)
            children.push(this.item2);
        children.push(...super.getChildNodes());
        return children;
    }
}
export class LibraryInvocationConstruct extends AbstractNode {
    constructor() {
        super(...arguments);
        this.libraryName = '';
        this.methodName = '';
        this.arguments = null;
    }
    getChildNodes() {
        if (this.arguments != null)
            return [this.arguments];
        return [];
    }
}
export class LiteralNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.sourceContent = '';
    }
}
export class MethodInvocationNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.method = null;
        this.args = null;
    }
    getChildNodes() {
        const children = [];
        if (this.method != null)
            children.push(this.method);
        if (this.args != null)
            children.push(this.args);
        return children;
    }
}
export class ModifierNameNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.name = '';
    }
}
export class ModifierValueNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.value = '';
        this.name = '';
    }
}
export class NamedArgumentNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.name = null;
        this.value = null;
    }
    getChildNodes() {
        const children = [];
        if (this.name != null)
            children.push(this.name);
        if (this.value != null)
            children.push(this.value);
        return children;
    }
}
export class NameValueNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.name = null;
        this.value = null;
    }
    getChildNodes() {
        const children = [];
        if (this.name != null)
            children.push(this.name);
        if (this.value != null)
            children.push(this.value);
        return children;
    }
}
export class NumberNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.value = null;
    }
}
export class StringValueNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.value = '';
        this.sourceTerminator = '';
        this.sourceContent = '';
    }
}
export class VariableNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.name = '';
        this.mergeRefName = '';
        this.variableReference = null;
        this.interpolationNodes = [];
        this.isInterpolationReference = false;
    }
    getChildNodes() {
        return this.interpolationNodes;
    }
}
export class AccessorNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.name = '';
    }
}
export class PathNode extends AbstractNode {
    constructor() {
        super(...arguments);
        this.delimiter = '';
        this.name = '';
        this.isStringVar = false;
        this.isFinal = false;
    }
}
export class VariableReference {
    constructor() {
        this.originalContent = '';
        this.normalizedReference = '';
        this.isStrictTagReference = false;
        this.isStrictVariableReference = false;
        this.isExplicitVariableReference = false;
        this.isVariableVariable = false;
        this.pathParts = [];
        this.isFinal = false;
        /** Start: Internal Parser State Variables */
        this._isFromEmptyFailState = false;
    }
    /** End: Internal Parser State Variables */
    clone() {
        const reference = new VariableReference();
        reference.originalContent = this.originalContent;
        reference.normalizedReference = this.normalizedReference;
        reference.isStrictVariableReference = this.isStrictVariableReference;
        reference.isExplicitVariableReference = this.isExplicitVariableReference;
        reference.pathParts = this.pathParts;
        reference.isFinal = this.isFinal;
        return reference;
    }
    implodePaths() {
        const stringParts = [];
        this.pathParts.forEach((part) => {
            if (part instanceof PathNode) {
                stringParts.push(part.name);
            }
            else if (part instanceof VariableReference) {
                stringParts.push(part.implodePaths());
            }
        });
        return stringParts.join('.');
    }
}
export class StaticTracedAssignment {
    constructor(target, value, operator) {
        this.target = target;
        this.value = value;
        this.operator = operator;
    }
}
