// td-antlers-parser/out/runtime/parser/fragmentsParser.js
import { FragmentNode, FragmentParameterNode } from '../nodes/abstractNode.js';
import { Position } from '../nodes/position.js';
import { StringUtilities } from '../utilities/stringUtilities.js';
import { DocumentParser } from './documentParser.js';
import { isStartOfString } from './scanners/isStartOfString.js';
import { skipToEndOfLine } from './scanners/skipToEndOfLine.js';
import { skipToEndOfMultilineComment } from './scanners/skipToEndOfMultilineComment.js';
import { skipToEndOfString } from './scanners/skipToEndOfString.js';
export class FragmentsParser {
    constructor() {
        this.content = '';
        this.nodeIndex = [];
        this.fragmentStartIndex = [];
        this.charLen = 0;
        this.documentOffsets = new Map();
        this.inputLen = 0;
        this.seedOffset = 0;
        this.shiftLine = 0;
        this.chars = [];
        this.currentIndex = 0;
        this.lastDocumentOffsetKey = null;
        this.cur = null;
        this.next = null;
        this.prev = null;
        this.chunkSize = 5;
        this.currentChunkOffset = 0;
        this.fragments = [];
        this.isScript = false;
        this.fragmentOpeningIndex = new Map();
        this.indexedFragments = new Map();
        this.embeddedIndexedFragments = new Map();
    }
    updateIndex(index) {
        this.currentIndex = index;
    }
    inputLength() {
        return this.inputLen;
    }
    incrementIndex() {
        this.currentIndex += 1;
    }
    getCurrentIndex() {
        return this.currentIndex;
    }
    getCurrent() {
        return this.cur;
    }
    getNext() {
        return this.next;
    }
    getPrev() {
        return this.prev;
    }
    pushChar(value) {
        return;
    }
    getChar(index) {
        return this.chars[index];
    }
    getSeedOffset() {
        return this.seedOffset;
    }
    setDocumentOffsets(offsets, lastOffsetIndex) {
        this.documentOffsets = offsets;
        this.lastDocumentOffsetKey = lastOffsetIndex;
        return this;
    }
    getClosingFragmentAfter(fragment) {
        if (typeof fragment === 'undefined') {
            return null;
        }
        const lowerName = fragment.name.toLowerCase();
        for (let i = 0; i < this.fragments.length; i++) {
            const thisFragment = this.fragments[i];
            if (thisFragment.isClosingFragment && !thisFragment.isSelfClosing && thisFragment.name.toLowerCase() == lowerName && thisFragment.index > fragment.index) {
                return thisFragment;
            }
        }
        return null;
    }
    getContentSubstring(from, length) {
        return this.content.substr(from, length);
    }
    resetState() {
        this.fragments = [];
        this.inputLen = 0;
        this.fragmentStartIndex = [];
        this.fragmentOpeningIndex.clear();
        return this;
    }
    resetIntermediateState() {
        this.chars = [];
        this.charLen = 0;
        this.currentIndex = 0;
        this.seedOffset = 0;
        this.cur = null;
        this.next = null;
        this.prev = null;
    }
    positionFromOffset(offset, index, isRelativeOffset = false) {
        let lineToUse = 0, charToUse = 0;
        if (!this.documentOffsets.has(offset)) {
            if (this.documentOffsets.size == 0) {
                lineToUse = 1;
                charToUse = offset + 1;
            }
            else {
                let nearestOffset = null, nearestOffsetIndex = null, lastOffset = null, lastOffsetIndex = null;
                for (const documentOffset of this.documentOffsets.keys()) {
                    if (documentOffset >= offset) {
                        if (lastOffsetIndex != null && offset > lastOffsetIndex) {
                            nearestOffset = lastOffset;
                            nearestOffsetIndex = lastOffsetIndex;
                        }
                        else {
                            nearestOffset = this.documentOffsets.get(documentOffset);
                            nearestOffsetIndex = documentOffset;
                        }
                        break;
                    }
                    lastOffset = this.documentOffsets.get(documentOffset);
                    lastOffsetIndex = documentOffset;
                }
                if (nearestOffset == null) {
                    nearestOffset = lastOffset;
                    nearestOffsetIndex = lastOffsetIndex;
                }
                if (nearestOffset != null) {
                    if (isRelativeOffset) {
                        const tChar = offset - (nearestOffsetIndex ?? 0);
                        charToUse = tChar;
                        lineToUse = nearestOffset.line;
                        if (offset <= (nearestOffsetIndex ?? 0)) {
                            lineToUse = nearestOffset.line;
                            charToUse = offset + 1;
                        }
                        else {
                            lineToUse = nearestOffset.line + 1;
                        }
                    }
                    else {
                        const tChar = offset - (nearestOffsetIndex ?? 0);
                        charToUse = tChar;
                        lineToUse = nearestOffset.line;
                        if (offset <= (nearestOffsetIndex ?? 0)) {
                            lineToUse = nearestOffset.line;
                            charToUse = offset + 1;
                        }
                        else {
                            lineToUse = nearestOffset.line + 1;
                        }
                    }
                }
                else {
                    if (this.lastDocumentOffsetKey != null) {
                        const lastOffset = this.documentOffsets.get(this.lastDocumentOffsetKey);
                        lineToUse = lastOffset.line + 1;
                        charToUse = offset + this.lastDocumentOffsetKey;
                    }
                }
            }
        }
        else {
            const offsetDetails = this.documentOffsets.get(offset);
            lineToUse = offsetDetails.line;
            charToUse = offsetDetails.char;
        }
        const position = new Position();
        position.index = index;
        position.offset = offset;
        position.line = lineToUse + this.shiftLine;
        position.char = charToUse;
        return position;
    }
    getFragments() {
        return this.fragments;
    }
    getFragmentsBetween(startIndex, endIndex) {
        const returnFragments = [];
        this.fragments.forEach((fragment) => {
            const start = fragment.startPosition?.index ?? 0, end = fragment.endPosition?.index ?? 0;
            if (start > startIndex && end < endIndex) {
                returnFragments.push(fragment);
            }
        });
        return returnFragments;
    }
    getFragmentContaining(position) {
        for (let i = 0; i < this.fragments.length; i++) {
            const thisFragment = this.fragments[i];
            if ((thisFragment.startPosition?.index ?? 0) < position.index && (thisFragment.endPosition?.index ?? 0) > position.index) {
                return thisFragment;
            }
        }
        return null;
    }
    checkCurrentOffsets() {
        if (this.currentIndex > this.chars.length) {
            this.cur = null;
            this.prev = null;
            this.next = null;
            return;
        }
        this.cur = this.chars[this.currentIndex];
        this.prev = null;
        this.next = null;
        if (this.currentIndex > 0) {
            this.prev = this.chars[this.currentIndex - 1];
        }
        if ((this.currentIndex + 1) < this.inputLen) {
            let doPeek = true;
            if (this.currentIndex == this.charLen - 1) {
                const nextChunk = StringUtilities.split(StringUtilities.substring(this.content, this.currentChunkOffset + this.chunkSize, this.chunkSize));
                this.currentChunkOffset += this.chunkSize;
                if (this.currentChunkOffset == this.inputLen) {
                    doPeek = false;
                }
                nextChunk.forEach((nextChar) => {
                    this.chars.push(nextChar);
                    this.charLen += 1;
                });
            }
            if (doPeek && (this.currentIndex + 1) < this.chars.length) {
                this.next = this.chars[this.currentIndex + 1];
            }
        }
    }
    processInputText(input) {
        this.content = input;
        this.inputLen = this.content.length;
        const fragmentCandidates = [...this.content.matchAll(/(<(?!\?[php|?=]))/gm)], newCandidates = [];
        fragmentCandidates.forEach((candidate) => {
            const candidateIndex = candidate.index ?? 0;
            for (let i = 0; i < this.nodeIndex.length; i++) {
                const thisIndex = this.nodeIndex[i];
                if (candidateIndex >= thisIndex.start && candidateIndex <= thisIndex.end) {
                    return;
                }
            }
            newCandidates.push(candidateIndex);
        });
        this.fragmentStartIndex = newCandidates;
    }
    setIndexRanges(index) {
        this.nodeIndex = index;
        return this;
    }
    fetchAt(start, length) {
        return this.content.substr(start, length);
    }
    getFragmentsContainingStructures() {
        const structures = [];
        let anyContainsStructures = false;
        this.fragments.forEach((fragment) => {
            if (fragment.containsStructures) {
                anyContainsStructures = true;
            }
        });
        this.fragments.forEach((fragment) => {
            const lowerName = fragment.name.toLowerCase();
            if (lowerName != 'script' && lowerName != 'style') {
                return;
            }
            if (!fragment.isClosingFragment && !fragment.isSelfClosing && (anyContainsStructures || fragment.containsStructures)) {
                const close = this.getClosingFragmentAfter(fragment);
                if (close != null) {
                    const structureContent = this.content.substring(((fragment.endPosition?.offset) ?? 0) + 1, close?.startPosition?.offset ?? 0), outerContent = this.content.substring(((fragment.startPosition?.offset) ?? 0), (close?.endPosition?.offset ?? 0) + 1);
                    structures.push({
                        start: fragment,
                        end: close,
                        content: structureContent,
                        outerContent: outerContent
                    });
                }
            }
        });
        return structures;
    }
    parse(text) {
        this.resetState().processInputText(text);
        const indexCount = this.fragmentStartIndex.length;
        if (indexCount == 0) {
            return [];
        }
        for (let i = 0; i < this.fragmentStartIndex.length; i++) {
            const offset = this.fragmentStartIndex[i];
            this.resetIntermediateState();
            this.seedOffset = offset;
            this.currentChunkOffset = offset;
            this.parseIntermediateText();
            if (this.isScript) {
                let didFind = false;
                for (let j = i + 1; j < this.fragmentStartIndex.length; j++) {
                    const checkIndex = this.fragmentStartIndex[j], thisChunk = this.fetchAt(checkIndex, 8).toLowerCase();
                    if (thisChunk == '</script') {
                        i = j - 1;
                        this.isScript = false;
                        didFind = true;
                        break;
                    }
                }
                if (!didFind) {
                    // Break if we get to this point since the script is not closed.
                    break;
                }
            }
        }
        let fragmentIndex = 0, embeddedIndex = 0;
        this.fragments.forEach((fragment) => {
            const lowerName = fragment.name.toLowerCase();
            if (lowerName == 'script' || lowerName == 'style') {
                fragment.embeddedIndex = embeddedIndex;
                this.embeddedIndexedFragments.set(embeddedIndex, fragment);
                embeddedIndex += 1;
            }
            fragment.index = fragmentIndex;
            this.indexedFragments.set(fragmentIndex, fragment);
            fragmentIndex += 1;
        });
        return this.fragments;
    }
    getFragment(index) {
        return this.indexedFragments.get(index);
    }
    getEmbeddedFragment(index) {
        if (!this.embeddedIndexedFragments.has(index)) {
            return null;
        }
        return this.embeddedIndexedFragments.get(index);
    }
    hasFragments() {
        return this.fragments.length > 0;
    }
    skipToEndOfPhp() {
        for (this.currentIndex; this.currentIndex < this.inputLen; this.currentIndex += 1) {
            this.checkCurrentOffsets();
            if (isStartOfString(this.cur)) {
                skipToEndOfString(this);
            }
            if (this.cur == DocumentParser.Punctuation_ForwardSlash && this.next == DocumentParser.Punctuation_ForwardSlash) {
                skipToEndOfLine(this, true);
                continue;
            }
            if (this.cur == DocumentParser.Punctuation_ForwardSlash && this.next == DocumentParser.Punctuation_Asterisk) {
                skipToEndOfMultilineComment(this, true);
                continue;
            }
            if ((this.cur == DocumentParser.Punctuation_Question && this.next == DocumentParser.Punctuation_GreaterThan) || this.next == null) {
                break;
            }
        }
    }
    parseIntermediateText() {
        this.chars = this.content.substr(this.currentChunkOffset, this.chunkSize).split('');
        this.charLen = this.chars.length;
        const fragmentStartedAt = this.currentIndex, potentialParameterRanges = [], nameChars = [];
        let hasFoundName = false, isClosingFragment = false;
        for (this.currentIndex = 0; this.currentIndex < this.inputLen; this.currentIndex += 1) {
            this.checkCurrentOffsets();
            if (StringUtilities.ctypeSpace(this.cur)) {
                hasFoundName = true;
            }
            if (this.next == DocumentParser.Punctuation_LessThan) {
                let shift = 1;
                if (this.cur == DocumentParser.Punctuation_ForwardSlash) {
                    shift = 2;
                }
                const curFetch = this.fetchAt(this.currentChunkOffset + shift, 5);
                if (curFetch.toLowerCase() == '<?php' || curFetch.startsWith('<?=')) {
                    this.skipToEndOfPhp();
                    continue;
                }
            }
            if (this.currentIndex > 0 && this.cur == DocumentParser.Punctuation_LessThan) {
                break;
            }
            if (this.next == null && this.cur != DocumentParser.Punctuation_GreaterThan) {
                break;
            }
            if (!hasFoundName && this.cur == DocumentParser.Punctuation_LessThan && this.next == DocumentParser.Punctuation_ForwardSlash) {
                isClosingFragment = true;
            }
            if (!hasFoundName && this.cur != DocumentParser.Punctuation_LessThan &&
                this.cur != DocumentParser.Punctuation_ForwardSlash &&
                this.cur != DocumentParser.Punctuation_GreaterThan) {
                nameChars.push(this.cur);
            }
            if (isStartOfString(this.cur)) {
                const stringStartedOn = this.currentIndex;
                skipToEndOfString(this);
                const fragmentParameter = new FragmentParameterNode();
                fragmentParameter.startPosition = this.positionFromOffset(stringStartedOn + this.seedOffset, stringStartedOn + this.seedOffset);
                fragmentParameter.endPosition = this.positionFromOffset(this.currentIndex + this.seedOffset, this.currentIndex + this.seedOffset);
                const startedOn = stringStartedOn + this.seedOffset, endedOn = this.currentIndex + this.seedOffset + 1;
                fragmentParameter.content = this.content.substr(startedOn, endedOn - startedOn);
                const nameChars = [];
                for (let j = stringStartedOn - 2; j >= 0; j--) {
                    const cur = this.chars[j];
                    if (StringUtilities.ctypeSpace(cur)) {
                        break;
                    }
                    nameChars.push(cur);
                }
                fragmentParameter.name = nameChars.reverse().join('');
                potentialParameterRanges.push(fragmentParameter);
                continue;
            }
            if (this.cur == DocumentParser.Punctuation_ForwardSlash && this.next == DocumentParser.Punctuation_GreaterThan) {
                const fragment = new FragmentNode();
                fragment.startPosition = this.positionFromOffset(fragmentStartedAt + this.seedOffset, fragmentStartedAt + this.seedOffset);
                fragment.endPosition = this.positionFromOffset(this.currentIndex + this.seedOffset + 1, this.currentIndex + this.seedOffset + 1);
                fragment.parameters = potentialParameterRanges;
                fragment.isSelfClosing = true;
                fragment.name = nameChars.join('');
                fragment.isClosingFragment = isClosingFragment;
                this.fragments.push(fragment);
                if (isClosingFragment) {
                    this.fragmentOpeningIndex.set(fragment.startPosition.index + 1, 1);
                }
                else {
                    this.fragmentOpeningIndex.set(fragment.startPosition.index, 1);
                }
                break;
            }
            if (this.cur == DocumentParser.Punctuation_GreaterThan) {
                const fragment = new FragmentNode();
                fragment.startPosition = this.positionFromOffset(fragmentStartedAt + this.seedOffset, fragmentStartedAt + this.seedOffset);
                fragment.endPosition = this.positionFromOffset(this.currentIndex + this.seedOffset, this.currentIndex + this.seedOffset);
                fragment.parameters = potentialParameterRanges;
                fragment.name = nameChars.join('');
                fragment.isClosingFragment = isClosingFragment;
                this.fragments.push(fragment);
                if (isClosingFragment) {
                    this.fragmentOpeningIndex.set(fragment.startPosition.index + 1, 1);
                }
                else {
                    this.fragmentOpeningIndex.set(fragment.startPosition.index, 1);
                }
                if (fragment.name.toLowerCase() == 'script' && fragment.isClosingFragment == false) {
                    this.isScript = true;
                }
                break;
            }
        }
    }
}
