// td-antlers-parser/out/runtime/parser/languageParser.js
import { AntlersError } from '../errors/antlersError.js';
import { AntlersErrorCodes } from '../errors/antlersErrorCodes.js';
import { LineRetriever } from '../errors/lineRetriever.js';
import { TypeLabeler } from '../errors/typeLabeler.js';
import { AbstractNode, AdditionAssignmentOperator, AdditionOperator, AliasedScopeLogicGroup, ArgSeparator, ArgumentGroup, ArrayNode, ConditionalVariableFallbackOperator, DirectionGroup, DivisionAssignmentOperator, DivisionOperator, EqualCompOperator, ExponentiationOperator, FactorialOperator, FalseConstant, GreaterThanCompOperator, GreaterThanEqualCompOperator, InlineBranchSeparator, InlineTernarySeparator, LanguageOperatorConstruct, LeftAssignmentOperator, LessThanCompOperator, LessThanEqualCompOperator, LibraryInvocationConstruct, ListValueNode, LogicalAndOperator, LogicalNegationOperator, LogicalOrOperator, LogicalXorOperator, LogicGroup, LogicGroupBegin, LogicGroupEnd, MethodInvocationNode, ModifierChainNode, ModifierNameNode, ModifierNode, ModifierSeparator, ModifierValueNode, ModifierValueSeparator, ModulusAssignmentOperator, ModulusOperator, MultiplicationAssignmentOperator, MultiplicationOperator, NamedArgumentNode, NameValueNode, NotEqualCompOperator, NotStrictEqualCompOperator, NullCoalescenceGroup, NullCoalesceOperator, NullConstant, NumberNode, PathNode, ScopeAssignmentOperator, ScopedLogicGroup, SemanticGroup, SpaceshipCompOperator, StatementSeparatorNode, StaticTracedAssignment, StrictEqualCompOperator, StringConcatenationOperator, StringValueNode, SubtractionAssignmentOperator, SubtractionOperator, SwitchCase, SwitchGroup, TernaryCondition, TrueConstant, TupleListStart, ValueDirectionNode, VariableNode } from '../nodes/abstractNode.js';
import { LibraryManager } from '../runtime/libraries/libraryManager.js';
import { LanguageOperatorRegistry } from '../runtime/sandbox/languageOperatorRegistry.js';
import { NodeHelpers } from '../utilities/nodeHelpers.js';
import { StringUtilities } from '../utilities/stringUtilities.js';
import { DocumentParser } from './documentParser.js';
import { LanguageKeywords } from './languageKeywords.js';
import { PathParser } from './pathParser.js';
export class LanguageParser {
    constructor() {
        this.pathParser = new PathParser();
        this.tokens = [];
        this.positionHydrationSkippedKeys = new Set([
            'parent',
            'prev',
            'next',
            'parser',
            'originalAbstractNode',
            'methodTarget',
            'libraryTarget',
            'originalNode',
            'recursiveParent',
            'isOpenedBy',
            'isClosedBy',
            'modifier',
            'scopeVariable',
            'currentScope',
            'reference',
            'runtimeType',
            'fragment',
            'fragmentParameter'
        ]);
        this.createdVariables = [];
        this.createdModifierChains = [];
        this.createdLanguageOperators = [];
        this.createdArrays = [];
        this.tracedAssignments = [];
        this.mergedVariablePaths = [];
        this.mergedComponents = new Map();
        this.modifierNameMapping = new Map();
        this.isRealModifierSeparator = new Map();
        this.isRoot = true;
        this.retriggerNonVirtualGroupMembers = [];
    }
    hasModifierConstruct(name) {
        return this.modifierNameMapping.has(name);
    }
    getModifierConstruct(name) {
        return this.modifierNameMapping.get(name);
    }
    isActualModifierSeparator(node) {
        return this.isRealModifierSeparator.has(node);
    }
    isMergedVariableComponent(node) {
        return this.mergedComponents.has(node);
    }
    getMergedVariable(node) {
        return this.mergedComponents.get(node);
    }
    getMergedVariablePaths() {
        return this.mergedVariablePaths;
    }
    getCreatedModifierChains() {
        return this.createdModifierChains;
    }
    getCreatedLangaugeOperators() {
        return this.createdLanguageOperators;
    }
    getCreatedArrays() {
        return this.createdArrays;
    }
    getCreatedVariables() {
        return this.createdVariables;
    }
    getRuntimeAssignments() {
        return this.tracedAssignments;
    }
    getAssignedVariables() {
        const variables = [], addedVariableNames = [];
        for (let i = 0; i < this.tracedAssignments.length; i++) {
            const thisAssignment = this.tracedAssignments[i];
            if (thisAssignment.target instanceof VariableNode) {
                if (addedVariableNames.includes(thisAssignment.target.name) == false) {
                    addedVariableNames.push(thisAssignment.target.name);
                    variables.push(thisAssignment.target);
                }
            }
        }
        return variables;
    }
    getAssignedVariableNames() {
        const addedVariableNames = [];
        for (let i = 0; i < this.tracedAssignments.length; i++) {
            const thisAssignment = this.tracedAssignments[i];
            if (thisAssignment.target instanceof VariableNode) {
                if (addedVariableNames.includes(thisAssignment.target.name) == false) {
                    addedVariableNames.push(thisAssignment.target.name);
                }
            }
        }
        return addedVariableNames;
    }
    reset() {
        this.tokens = [];
        this.createdModifierChains = [];
        this.createdLanguageOperators = [];
        this.createdArrays = [];
        this.createdModifierChains = [];
        this.createdLanguageOperators = [];
        this.tracedAssignments = [];
        this.createdArrays = [];
        this.mergedVariablePaths = [];
        this.mergedComponents.clear();
        this.modifierNameMapping.clear();
    }
    parse(tokens) {
        this.tokens = tokens;
        this.tokens = this.combineVariablePaths(this.tokens);
        this.tokens = this.createLanguageOperators(this.tokens);
        this.tokens = this.createLogicalGroups(this.tokens);
        this.tokens = this.associateMethodCalls(this.tokens);
        this.tokens = this.createTupleLists(this.tokens);
        this.tokens = this.createLibraryInvocations(this.tokens);
        this.tokens = this.associateModifiers(this.tokens);
        this.tokens = this.createNullCoalescenceGroups(this.tokens);
        // With modifiers and variables accounted for, this will
        // rewrite some tokens to a more "correct" type based
        // on the now more well-known context.
        this.tokens = this.correctTypes(this.tokens);
        this.tokens = this.createTernaryGroups(this.tokens);
        this.tokens = this.applyOperationOrder(this.tokens);
        this.validateNeighboringOperators(this.tokens);
        this.tokens = this.createLogicGroupsToRemoveMethodInvocationAmbiguity(this.tokens);
        this.tokens = this.createLogicGroupsToResolveOperatorAmbiguity(this.tokens);
        this.validateNoDanglingLogicGroupEnds(this.tokens);
        if (this.isRoot) {
            this.tokens = this.insertAutomaticStatementSeparators(this.tokens);
            if (this.retriggerNonVirtualGroupMembers.length > 0) {
                this.retriggerNonVirtualGroupMembers.forEach((grouping) => {
                    this.markAllNonVirtualAsListGroupMember(grouping);
                });
            }
        }
        const semanticGroups = this.createSemanticGroups(this.tokens);
        this.hydrateMissingNodePositions(semanticGroups);
        return semanticGroups;
    }
    setIsRoot(isRoot) {
        this.isRoot = isRoot;
        return this;
    }
    validateNeighboringOperators(tokens) {
        const tokenCount = tokens.length;
        for (let i = 0; i < tokenCount; i++) {
            const thisToken = tokens[i];
            if (LanguageParser.isOperatorType(thisToken)) {
                if (i + 1 >= tokenCount) {
                    thisToken.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_OPERATOR, thisToken, 'Unexpected operator while parsing input text.'));
                    return;
                }
                const peek = tokens[i + 1];
                if (LanguageParser.isOperatorType(peek)) {
                    peek.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_OPERATOR_INVALID_ON_RIGHT, peek, 'Unexpected operator while parsing input text.'));
                    continue;
                }
            }
        }
    }
    static isAssignmentOperator(token) {
        if (token instanceof AdditionAssignmentOperator ||
            token instanceof DivisionAssignmentOperator || token instanceof LeftAssignmentOperator ||
            token instanceof ModulusAssignmentOperator || token instanceof MultiplicationAssignmentOperator ||
            token instanceof SubtractionAssignmentOperator) {
            return true;
        }
        return false;
    }
    static isOperatorType(token) {
        if (token instanceof AdditionAssignmentOperator ||
            token instanceof DivisionAssignmentOperator ||
            token instanceof LeftAssignmentOperator ||
            token instanceof ModulusAssignmentOperator ||
            token instanceof MultiplicationAssignmentOperator ||
            token instanceof SubtractionAssignmentOperator ||
            token instanceof ConditionalVariableFallbackOperator ||
            token instanceof LanguageOperatorConstruct ||
            token instanceof LogicalAndOperator ||
            token instanceof LogicalNegationOperator ||
            token instanceof LogicalOrOperator ||
            token instanceof LogicalXorOperator ||
            token instanceof NullCoalesceOperator ||
            token instanceof StringConcatenationOperator ||
            token instanceof AdditionOperator ||
            token instanceof DivisionOperator ||
            token instanceof ExponentiationOperator ||
            token instanceof ModulusOperator ||
            token instanceof MultiplicationOperator ||
            token instanceof SubtractionOperator ||
            token instanceof EqualCompOperator ||
            token instanceof GreaterThanCompOperator ||
            token instanceof GreaterThanEqualCompOperator ||
            token instanceof LessThanCompOperator ||
            token instanceof LessThanEqualCompOperator ||
            token instanceof NotEqualCompOperator ||
            token instanceof NotStrictEqualCompOperator ||
            token instanceof SpaceshipCompOperator ||
            token instanceof StrictEqualCompOperator) {
            return true;
        }
        return false;
    }
    markAllNonVirtualAsListGroupMember(nodes) {
        nodes.forEach((node) => {
            if (node instanceof LogicGroup || node instanceof ScopedLogicGroup || node instanceof AliasedScopeLogicGroup) {
                if (node.start != null && node.start.isVirtual == false) {
                    node.start.isListGroupMember = true;
                }
                if (node.end != null && node.end.isVirtual == false) {
                    node.end.isListGroupMember = true;
                }
                if (node.scopeOperator != null) {
                    node.scopeOperator.isListGroupMember = true;
                }
                node.nodes.forEach((lgNode) => {
                    this.markAllNonVirtualAsListGroupMember([lgNode]);
                });
                return;
            }
            else if (node instanceof SemanticGroup) {
                if (node.separatorToken != null && node.separatorToken.isVirtual == false) {
                    node.separatorToken.isListGroupMember = true;
                }
                this.markAllNonVirtualAsListGroupMember(node.nodes);
            }
            if (node.isVirtual == false) {
                node.isListGroupMember = true;
            }
            if (node.originalAbstractNode != null && node.originalAbstractNode != node) {
                this.markAllNonVirtualAsListGroupMember([node.originalAbstractNode]);
            }
        });
    }
    markAllNonVirtualAsSwitch(nodes) {
        nodes.forEach((node) => {
            if (node instanceof LogicGroup || node instanceof ScopedLogicGroup || node instanceof AliasedScopeLogicGroup) {
                if (node.start != null && node.start.isVirtual == false) {
                    node.start.isSwitchGroupMember = true;
                }
                if (node.end != null && node.end.isVirtual == false) {
                    node.end.isSwitchGroupMember = true;
                }
                if (node.scopeOperator != null) {
                    node.scopeOperator.isSwitchGroupMember = true;
                }
                node.nodes.forEach((lgNode) => {
                    this.markAllNonVirtualAsSwitch([lgNode]);
                });
                return;
            }
            else if (node instanceof SemanticGroup) {
                if (node.separatorToken != null && node.separatorToken.isVirtual == false) {
                    node.separatorToken.isSwitchGroupMember = true;
                }
                this.markAllNonVirtualAsSwitch(node.nodes);
            }
            if (node.isVirtual == false) {
                node.isSwitchGroupMember = true;
            }
            if (node.originalAbstractNode != null && node.originalAbstractNode != node) {
                this.markAllNonVirtualAsSwitch([node.originalAbstractNode]);
            }
        });
    }
    validateNoDanglingLogicGroupEnds(tokens) {
        tokens.forEach((token) => {
            if (token instanceof LogicGroupEnd) {
                token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_LOGIC_GROUP_NO_START, token, 'Unexpected [T_LOGIC_END] while parsing input text.'));
            }
        });
    }
    isValidModifierValue(node) {
        if (node instanceof ModifierValueNode ||
            node instanceof VariableNode ||
            node instanceof TrueConstant ||
            node instanceof FalseConstant ||
            node instanceof NullConstant ||
            node instanceof NumberNode ||
            node instanceof StringValueNode) {
            return true;
        }
        return false;
    }
    wrapNumberInVariable(node) {
        const variableNode = new VariableNode();
        variableNode.startPosition = node.startPosition;
        variableNode.endPosition = node.endPosition;
        variableNode.name = node.value?.toString() ?? '';
        variableNode.content = node.value?.toString() ?? '';
        variableNode.originalAbstractNode = node;
        variableNode.refId = node.refId;
        variableNode.modifierChain = node.modifierChain;
        variableNode.index = node.index;
        this.createdVariables.push(variableNode);
        return variableNode;
    }
    wrapStringInVariable(node) {
        const variableNode = new VariableNode();
        variableNode.startPosition = node.startPosition;
        variableNode.endPosition = node.endPosition;
        variableNode.name = node.value?.toString() ?? '';
        variableNode.content = node.value?.toString() ?? '';
        variableNode.originalAbstractNode = node;
        variableNode.refId = node.refId;
        variableNode.modifierChain = node.modifierChain;
        variableNode.index = node.index;
        this.createdVariables.push(variableNode);
        return variableNode;
    }
    wrapArithmeticModifier(operator, modifierName) {
        const node = new ModifierNameNode();
        node.startPosition = operator.startPosition;
        node.endPosition = operator.endPosition;
        node.originalAbstractNode = operator;
        node.content = modifierName;
        node.name = modifierName;
        node.index = operator.index;
        return node;
    }
    combineVariablePaths(tokens) {
        const newNodes = [], tokenCount = tokens.length;
        for (let i = 0; i < tokenCount; i++) {
            const node = tokens[i], newNodeCount = newNodes.length;
            if (node instanceof InlineBranchSeparator) {
                if (newNodeCount == 0) {
                    node.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_BRANCH_SEPARATOR_FOR_VARCOMBINE, node, 'Unexpected [T_BRANCH_SEPARATOR] while parsing input text.'));
                    continue;
                }
                if (i + 1 > tokenCount) {
                    let lastNodeText = '';
                    if (i > 0) {
                        const lastNode = tokens[i - 1];
                        lastNodeText = LineRetriever.getNearText(lastNode);
                    }
                    node.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_PARSING_BRANCH_GROUP, node, 'Unexpected end of input while parsing input text near "' + lastNodeText + LineRetriever.getNearText(node) + '".'));
                    continue;
                }
                const left = newNodes[newNodeCount - 1];
                let right = tokens[i + 1];
                if (right instanceof NumberNode && NodeHelpers.distance(left, right) == 1) {
                    right = this.wrapNumberInVariable(right);
                }
                if (left instanceof VariableNode && right instanceof VariableNode && NodeHelpers.distance(left, right) == 1) {
                    // Note: It is important when we do this merge
                    // that we start from the right, and merge
                    // the closest left node and work back.
                    // The right-most node may have a
                    // valid modifier chain on it.
                    // Remove the var off the stack.
                    newNodes.pop();
                    NodeHelpers.mergeVarContentRight(':', node, right);
                    NodeHelpers.mergeVarRight(left, right);
                    newNodes.push(right);
                    this.createdVariables.push(right);
                    this.mergedVariablePaths.push(right);
                    this.mergedComponents.set(node, right);
                    this.mergedComponents.set(left, right);
                    this.mergedComponents.set(right, right);
                    // Skip over the adjusted right.
                    i += 1;
                    continue;
                }
                else {
                    newNodes.push(node);
                    continue;
                }
            }
            else if (node instanceof StringValueNode && newNodeCount > 0) {
                if ((i + 1) >= tokenCount) {
                    newNodes.push(node);
                    continue;
                }
                const left = newNodes[newNodeCount - 1], right = tokens[i + 1];
                if (left instanceof VariableNode && right instanceof VariableNode && NodeHelpers.distance(left, node) <= 1 && NodeHelpers.distance(node, right) <= 1) {
                    newNodes.push(this.wrapStringInVariable(node));
                    continue;
                    /*
                    Note: We don't need to do this just for static analysis/formatting.
                    NodeHelpers.mergeVarContentLeft(node.sourceTerminator + node.value + node.sourceTerminator, node, left);
                    NodeHelpers.mergeVarContentLeft(right.name, right, left);
                    this.createdVariables.push(left);
                    newNodes.push(left);
                    this.mergedVariablePaths.push(left);
                    this.mergedComponents.set(node, left);
                    this.mergedComponents.set(left, left);
                    this.mergedComponents.set(right, left);
                    */
                    i += 1;
                    continue;
                }
                else {
                    newNodes.push(node);
                }
            }
            else if (node instanceof SubtractionOperator && newNodeCount > 0) {
                const left = newNodes[newNodeCount - 1], right = tokens[i + 1];
                if (left instanceof VariableNode && (right instanceof VariableNode || right instanceof LogicalOrOperator)) {
                    const lDistance = NodeHelpers.distance(left, node), rDistance = NodeHelpers.distance(node, right);
                    if (lDistance <= 1 && rDistance <= 1) {
                        //NodeHelpers.mergeVarContentLeft('-', node, left);
                        //NodeHelpers.mergeVarContentLeft(right.name, right, left);
                        left.endPosition = right.endPosition;
                        this.mergedComponents.set(node, left);
                        this.mergedComponents.set(left, left);
                        this.mergedComponents.set(right, left);
                        i += 1;
                    }
                    else {
                        newNodes.push(node);
                    }
                }
                else {
                    newNodes.push(node);
                }
            }
            else {
                newNodes.push(node);
            }
        }
        newNodes.forEach((node) => {
            if (node instanceof VariableNode && !node.name.trim().startsWith('[')) {
                node.variableReference = this.pathParser.parse(node.name, node.startPosition);
                node.mergeErrors(this.pathParser.getAntlersErrors());
            }
        });
        return newNodes;
    }
    convertVarNodeToOperator(variable) {
        const operator = new LanguageOperatorConstruct();
        operator.content = variable.name;
        operator.startPosition = variable.startPosition;
        operator.endPosition = variable.endPosition;
        operator.originalAbstractNode = variable;
        variable.convertedToOperator = true;
        this.createdLanguageOperators.push(operator);
        return operator;
    }
    convertOperatorToVarNode(operator) {
        const varNodeWrap = new VariableNode();
        varNodeWrap.startPosition = operator.startPosition;
        varNodeWrap.endPosition = operator.endPosition;
        varNodeWrap.content = operator.content;
        varNodeWrap.name = operator.content;
        varNodeWrap.originalAbstractNode = operator;
        this.createdVariables.push(varNodeWrap);
        return varNodeWrap;
    }
    createLanguageOperators(tokens) {
        const nodeCount = tokens.length;
        for (let i = 0; i < nodeCount; i++) {
            const thisNode = tokens[i];
            if (thisNode instanceof VariableNode) {
                if (thisNode.name.includes(DocumentParser.Punctuation_FullStop)) {
                    const checkParts = thisNode.name.split(DocumentParser.Punctuation_FullStop);
                    if (LanguageOperatorRegistry.operators.includes(checkParts[0]) ||
                        LibraryManager.deferredCoreLibraries.includes(checkParts[0])) {
                        tokens[i] = this.convertVarNodeToOperator(thisNode);
                        continue;
                    }
                }
                if (LanguageOperatorRegistry.operators.includes(thisNode.name)) {
                    tokens[i] = this.convertVarNodeToOperator(thisNode);
                    continue;
                }
            }
            if (thisNode instanceof LanguageOperatorConstruct) {
                if (i + 1 >= nodeCount) {
                    tokens[i] = this.convertOperatorToVarNode(thisNode);
                    continue;
                }
                const next = tokens[i + 1];
                if (next instanceof StatementSeparatorNode) {
                    tokens[i] = this.convertOperatorToVarNode(thisNode);
                    continue;
                }
            }
        }
        return tokens;
    }
    createLogicalGroups(tokens) {
        const groupedTokens = [], tokenCount = tokens.length;
        const negatedGroupedTokens = [];
        for (let i = 0; i < tokenCount; i++) {
            const token = tokens[i];
            if (token instanceof LogicalNegationOperator) {
                const negationCount = this.countTypeRight(tokens, i, LogicalNegationOperator);
                if (negatedGroupedTokens.length == 0 && tokens.length == negationCount) {
                    token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_LOGIC_NEGATION_OPERATOR, token, 'Unexpected [T_LOGIC_INVERSE] while parsing input text.'));
                    continue;
                }
                if (i > 0 && negatedGroupedTokens.length > 0) {
                    const prev = negatedGroupedTokens[negatedGroupedTokens.length - 1];
                    if (prev instanceof NumberNode || prev instanceof LogicGroupEnd || prev instanceof LogicGroup) {
                        if (prev instanceof LogicGroup) {
                            if (prev.start instanceof LogicalNegationOperator) {
                                prev.start.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_FACTORIAL_MATERIALIZED_BOOL_DETECTED, prev.start, '[T_AOP_FACTORIAL] operand will always materialize boolean type.'));
                                continue;
                            }
                        }
                        const factorialOperator = new FactorialOperator();
                        factorialOperator.startPosition = token.startPosition;
                        factorialOperator.endPosition = token.endPosition;
                        factorialOperator.content = '!'.repeat(negationCount);
                        factorialOperator.originalAbstractNode = token;
                        factorialOperator.repeat = negationCount;
                        negatedGroupedTokens.push(factorialOperator);
                        if (negationCount > 0) {
                            i += negationCount - 1;
                        }
                        continue;
                    }
                }
                // Just ignore these at the parser level.
                // An even number of negation operators are the same has having no negation operators.
                if (negationCount % 2 == 0) {
                    i += negationCount - 1;
                    continue;
                }
                // We want to peek to the one after the last negation operator.
                const peek = tokens[i + negationCount];
                if (peek instanceof LogicGroupBegin) {
                    // Scan right to count the negations.
                    const targetSliceOffset = i + negationCount + 1;
                    if (targetSliceOffset >= tokenCount) {
                        peek.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_WHILE_PARSING_LOGIC_GROUP_NEGATION_OFFSET, peek, 'Unexpected end of input while parsing input text.'));
                        continue;
                    }
                    const groupResults = this.findLogicalGroupEnd(tokens[i + negationCount], tokens.slice(targetSliceOffset));
                    if (groupResults.found) {
                        const subGroup = groupResults.logicalGroup;
                        const wrapperGroup = new LogicGroup();
                        wrapperGroup.start = token;
                        wrapperGroup.end = subGroup;
                        wrapperGroup.nodes.push(token);
                        wrapperGroup.nodes.push(subGroup);
                        wrapperGroup.startPosition = token.startPosition;
                        wrapperGroup.endPosition = subGroup.endPosition;
                        negatedGroupedTokens.push(wrapperGroup);
                        i += groupResults.skipCount + negationCount;
                    }
                    continue;
                }
                const results = this.resolveValueRight(tokens, i);
                if (results.value != null) {
                    negatedGroupedTokens.push(results.value);
                }
                i += results.negationCount;
            }
            else {
                negatedGroupedTokens.push(token);
            }
        }
        const negatedTokenCount = negatedGroupedTokens.length;
        for (let i = 0; i < negatedTokenCount; i++) {
            const token = negatedGroupedTokens[i];
            if (token instanceof LogicGroupBegin) {
                if (i + 1 >= negatedTokenCount) {
                    token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_WHILE_PARSING_LOGIC_GROUP_END_DUE_TO_NEGATION, token, 'Unexpected end of input while parsing input text.'));
                    continue;
                }
                const group = this.findLogicalGroupEnd(token, negatedGroupedTokens.slice(i + 1));
                if (group.found && group.logicalGroup != null) {
                    groupedTokens.push(group.logicalGroup);
                }
                i += group.skipCount;
            }
            else {
                if (token instanceof FactorialOperator) {
                    if (groupedTokens.length == 0) {
                        token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_FACTORIAL_WHILE_CREATING_GROUPS, token, 'Unexpected [T_AOP_FACTORIAL] while parsing input text.'));
                        continue;
                    }
                    let prev = groupedTokens[groupedTokens.length - 1];
                    if (prev instanceof LogicGroup == false &&
                        prev instanceof NumberNode == false) {
                        token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_FACTORIAL_OPERAND, token, 'Unexpected left operand encountered for [T_AOP_FACTORIAL] while parsing input text.'));
                        continue;
                    }
                    prev = groupedTokens.pop();
                    const wrapperGroup = new LogicGroup();
                    wrapperGroup.startPosition = prev.startPosition;
                    wrapperGroup.endPosition = token.endPosition;
                    wrapperGroup.originalAbstractNode = token;
                    wrapperGroup.nodes.push(prev);
                    wrapperGroup.nodes.push(token);
                    groupedTokens.push(wrapperGroup);
                }
                else {
                    groupedTokens.push(token);
                }
            }
        }
        return groupedTokens;
    }
    makeArgGroup(nodes) {
        const argGroup = new ArgumentGroup(), nodeCount = nodes.length;
        for (let i = 0; i < nodeCount; i++) {
            const thisNode = nodes[i];
            if (this.isOperand(thisNode) == false) {
                thisNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_TOKEN_WHILE_PARSING_METHOD, thisNode, 'Unexpected [' + TypeLabeler.getPrettyTypeName(thisNode) + '] while parsing argument group.'));
                continue;
            }
            let next = null;
            if (i + 1 < nodeCount) {
                next = nodes[i + 1];
            }
            if (next == null || next instanceof ArgSeparator) {
                argGroup.args.push(nodes[i]);
                i += 1;
                continue;
            }
            else if (next instanceof InlineBranchSeparator) {
                if (i + 2 >= nodeCount) {
                    thisNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_ARG_UNEXPECTED_NAMED_ARGUMENT, thisNode, 'Unexpected end of input while parsing named argument.'));
                    continue;
                }
                const valueNode = nodes[i + 2];
                const namedArgument = new NamedArgumentNode();
                namedArgument.startPosition = thisNode.startPosition;
                namedArgument.endPosition = valueNode.endPosition;
                namedArgument.content = thisNode.content + valueNode.content;
                namedArgument.name = thisNode;
                namedArgument.value = valueNode;
                if (thisNode instanceof VariableNode == false) {
                    thisNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_INVALID_NAMED_ARG_IDENTIFIER, thisNode, 'Invalid type [' + TypeLabeler.getPrettyTypeName(thisNode) + '] supplied for named argument name.'));
                }
                argGroup.hasNamedArguments = true;
                argGroup.numberOfNamedArguments += 1;
                argGroup.args.push(namedArgument);
                if (i + 3 < nodeCount && nodes[i + 3] instanceof ArgSeparator) {
                    i += 3;
                }
                else {
                    i += 2;
                }
                continue;
            }
        }
        let remainderMustBeNamed = false;
        argGroup.args.forEach((arg) => {
            if (arg instanceof NamedArgumentNode) {
                remainderMustBeNamed = true;
            }
            else {
                if (remainderMustBeNamed) {
                    arg.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_UNNAMED_METHOD_ARGUMENT, arg, 'Unnamed arguments are not allowed to appear after a named argument.'));
                }
            }
        });
        if (argGroup.args.length > 0) {
            argGroup.startPosition = argGroup.args[0].startPosition;
            argGroup.endPosition = argGroup.args[argGroup.args.length - 1].endPosition;
        }
        return argGroup;
    }
    cleanVariableForMethodInvocation(node) {
        if (node.name.startsWith(':') || node.name.startsWith('.')) {
            node.name = node.name.substr(1);
            if (node.variableReference != null) {
                node.variableReference.originalContent = node.variableReference.originalContent.substr(1);
                node.variableReference.normalizedReference = node.variableReference.normalizedReference.substr(1);
            }
            if (node.startPosition != null) {
                node.startPosition.offset += 1;
                node.startPosition.char += 1;
            }
        }
    }
    reduceVariableReferenceForObjectAccessor(node) {
        const lastPath = node.variableReference?.pathParts.pop(), accessor = lastPath.name, accessorLen = accessor.length + 1;
        node.name = node.name.substr(0, -accessorLen);
        if (node.variableReference != null) {
            node.variableReference.originalContent = node.variableReference.originalContent.substr(0, -accessorLen);
            node.variableReference.normalizedReference = node.variableReference.normalizedReference.substr(0, -accessorLen);
            node.variableReference.pathParts[node.variableReference.pathParts.length - 1].isFinal = true;
        }
        if (node.endPosition != null) {
            node.endPosition.offset -= accessorLen;
            node.endPosition.char -= accessorLen;
        }
        return {
            accessor: accessor,
            accessorLen: accessorLen
        };
    }
    associateMethodCalls(tokens) {
        const newTokens = [], nodeCount = tokens.length;
        for (let i = 0; i < nodeCount; i++) {
            const thisNode = tokens[i];
            let prevNode = null, next = null;
            if (newTokens.length > 0) {
                prevNode = newTokens[newTokens.length - 1];
            }
            if (i + 1 < nodeCount) {
                next = tokens[i + 1];
            }
            if (thisNode instanceof VariableNode &&
                next instanceof LogicGroup && prevNode != null &&
                (this.isProperMethodChainTargetStrict(prevNode) ||
                    prevNode instanceof InlineBranchSeparator)) {
                const argGroup = next;
                let argNodes = [];
                if (argGroup.nodes.length > 0) {
                    if (argGroup.nodes[0] instanceof SemanticGroup) {
                        const tSg = argGroup.nodes[0];
                        argNodes = tSg.nodes;
                    }
                    else {
                        argNodes = argGroup.nodes;
                    }
                }
                const methodInvocation = new MethodInvocationNode();
                methodInvocation.startPosition = thisNode.startPosition;
                if (prevNode instanceof InlineBranchSeparator) {
                    const branchSeparator = newTokens.pop();
                    methodInvocation.startPosition = branchSeparator.startPosition;
                }
                this.cleanVariableForMethodInvocation(thisNode);
                methodInvocation.endPosition = argGroup.endPosition;
                methodInvocation.args = this.makeArgGroup(argNodes);
                methodInvocation.method = thisNode;
                prevNode.methodTarget = methodInvocation;
                if (prevNode instanceof VariableNode && prevNode.mergeRefName.length > 0) {
                    prevNode.name = NodeHelpers.getUnrefName(prevNode.mergeRefName);
                }
                newTokens.push(methodInvocation);
                i += 1;
                continue;
            }
            else if (thisNode instanceof LogicGroup &&
                prevNode instanceof VariableNode &&
                prevNode.variableReference != null && prevNode.variableReference.pathParts.length >= 2) {
                let argNodes = [];
                if (thisNode.nodes.length > 0) {
                    if (thisNode.nodes[0] instanceof SemanticGroup) {
                        const tSg = thisNode.nodes[0];
                        argNodes = tSg.nodes;
                    }
                    else {
                        argNodes = thisNode.nodes;
                    }
                }
                if (prevNode.variableReference.pathParts[0] instanceof PathNode) {
                    if (LibraryManager.deferredCoreLibraries.includes(prevNode.variableReference.pathParts[0].name)) {
                        const varRef = newTokens.pop();
                        const libraryInvocation = new LibraryInvocationConstruct();
                        libraryInvocation.startPosition = varRef.startPosition;
                        libraryInvocation.endPosition = thisNode.endPosition;
                        libraryInvocation.content = varRef.content;
                        libraryInvocation.originalAbstractNode = thisNode;
                        const libraryReference = varRef.variableReference?.pathParts.shift();
                        const libraryName = libraryReference.name, methodName = varRef.variableReference?.implodePaths();
                        libraryInvocation.libraryName = libraryName;
                        libraryInvocation.methodName = methodName;
                        libraryInvocation.arguments = this.makeArgGroup(argNodes);
                        newTokens.push(libraryInvocation);
                        i += 1;
                        continue;
                    }
                }
                const reductionResult = this.reduceVariableReferenceForObjectAccessor(prevNode);
                const methodInvocation = new MethodInvocationNode();
                methodInvocation.startPosition = thisNode.startPosition;
                methodInvocation.endPosition = thisNode.endPosition;
                prevNode.methodTarget = methodInvocation;
                if (prevNode instanceof VariableNode && prevNode.mergeRefName.length > 0) {
                    prevNode.name = NodeHelpers.getUnrefName(prevNode.mergeRefName);
                }
                if (thisNode.start != null) {
                    thisNode.start.isPartOfMethodChain = true;
                }
                if (thisNode.end != null) {
                    thisNode.end.isPartOfMethodChain = true;
                }
                const wrappedMethod = new VariableNode();
                wrappedMethod.name = reductionResult.accessor;
                wrappedMethod.content = reductionResult.accessor;
                wrappedMethod.isPartOfMethodChain = true;
                const args = this.makeArgGroup(argNodes);
                if (methodInvocation.startPosition != null) {
                    methodInvocation.startPosition.offset -= reductionResult.accessorLen;
                    methodInvocation.startPosition.char -= reductionResult.accessorLen;
                    wrappedMethod.startPosition = methodInvocation.startPosition.shiftRight(1);
                    wrappedMethod.endPosition = methodInvocation.startPosition.shiftRight(Math.max(reductionResult.accessorLen - 1, 1));
                }
                methodInvocation.method = wrappedMethod;
                methodInvocation.args = args;
                if (StringUtilities.isNumeric(prevNode.name)) {
                    const varNode = newTokens.pop();
                    newTokens.push(this.convertOperatorToVarNode(varNode));
                }
                newTokens.push(methodInvocation);
                continue;
            }
            else if (thisNode instanceof VariableNode && prevNode instanceof MethodInvocationNode) {
                this.cleanVariableForMethodInvocation(thisNode);
                if (i + 1 > nodeCount) {
                    thisNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_METHOD_CALL_MISSING_ARG_GROUP, thisNode, 'Unexpected end of input while parsing method call.'));
                    continue;
                }
                const argGroupT = tokens[i + 1];
                if (argGroupT instanceof LogicGroup == false) {
                    thisNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_INVALID_METHOD_CALL_ARG_GROUP, thisNode, 'Unexpected [' + TypeLabeler.getPrettyTypeName(argGroupT) + '] while parsing [T_METHOD_CALL]; expecting [T_ARG_GROUP].'));
                    continue;
                }
                const argGroup = argGroupT;
                let argNodes = [];
                if (argGroup.nodes.length > 0) {
                    if (argGroup.nodes[0] instanceof SemanticGroup) {
                        const tSg = argGroup.nodes[0];
                        argNodes = tSg.nodes;
                    }
                    else {
                        argNodes = argGroup.nodes;
                    }
                }
                const methodInvocation = new MethodInvocationNode();
                methodInvocation.startPosition = thisNode.startPosition;
                methodInvocation.endPosition = argGroup.endPosition;
                methodInvocation.args = this.makeArgGroup(argNodes);
                methodInvocation.method = thisNode;
                prevNode.methodTarget = methodInvocation;
                newTokens.push(methodInvocation);
                i += 1;
                continue;
            }
            else if (thisNode instanceof InlineBranchSeparator && prevNode instanceof MethodInvocationNode) {
                if (i + 1 > nodeCount) {
                    thisNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_METHOD_CALL_MISSING_METHOD, thisNode, 'Unexpected end of input while parsing method call.'));
                    continue;
                }
                const next = tokens[i + 1];
                if (i + 2 > nodeCount) {
                    thisNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_METHOD_CALL_MISSING_ARG_GROUP, thisNode, 'Unexpected end of input while parsing method call.'));
                    continue;
                }
                const argGroupT = tokens[i + 2];
                if (argGroupT instanceof LogicGroup == false) {
                    thisNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_INVALID_METHOD_CALL_ARG_GROUP, thisNode, 'Unexpected [' + TypeLabeler.getPrettyTypeName(argGroupT) + '] while parsing [T_METHOD_CALL]; expecting [T_ARG_GROUP].'));
                    continue;
                }
                const argGroup = argGroupT;
                let argNodes = [];
                if (argGroup.nodes.length > 0) {
                    if (argGroup.nodes[0] instanceof SemanticGroup) {
                        const tSg = argGroup.nodes[0];
                        argNodes = tSg.nodes;
                    }
                    else {
                        argNodes = argGroup.nodes;
                    }
                }
                const methodInvocation = new MethodInvocationNode();
                methodInvocation.startPosition = thisNode.startPosition;
                methodInvocation.endPosition = argGroupT.endPosition;
                methodInvocation.args = this.makeArgGroup(argNodes);
                methodInvocation.method = next;
                prevNode.methodTarget = methodInvocation;
                newTokens.push(methodInvocation);
                i += 2;
                continue;
            }
            else if (thisNode instanceof MethodInvocationNode) {
                if (i + 1 > nodeCount) {
                    thisNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_METHOD_CALL_MISSING_METHOD, thisNode, 'Unexpected end of input while parsing method call.'));
                    continue;
                }
                const methodNode = tokens[i + 1];
                if (i + 2 > nodeCount) {
                    thisNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_INVALID_METHOD_CALL_ARG_GROUP, thisNode, 'Unexpected end of input while parsing method call.'));
                    continue;
                }
                const nextT = tokens[i + 2];
                if (nextT instanceof LogicGroup == false) {
                    thisNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_INVALID_METHOD_CALL_ARG_GROUP, thisNode, 'Unexpected [' + TypeLabeler.getPrettyTypeName(nextT) + '] while parsing [T_METHOD_CALL]; expecting [T_ARG_GROUP].'));
                    continue;
                }
                const next = nextT;
                let argNodes = [];
                if (next.nodes.length > 0) {
                    if (next.nodes[0] instanceof SemanticGroup) {
                        const tSg = next.nodes[0];
                        argNodes = tSg.nodes;
                    }
                    else {
                        argNodes = next.nodes;
                    }
                }
                const args = this.makeArgGroup(argNodes);
                if (prevNode instanceof VariableNode && LibraryManager.deferredCoreLibraries.includes(prevNode.name)) {
                    const varRef = newTokens.pop();
                    const libraryInvocation = new LibraryInvocationConstruct();
                    libraryInvocation.startPosition = varRef.startPosition;
                    libraryInvocation.endPosition = next.endPosition;
                    libraryInvocation.libraryName = varRef.name;
                    libraryInvocation.methodName = methodNode.name;
                    libraryInvocation.content = libraryInvocation.libraryName + '->' + libraryInvocation.methodName;
                    libraryInvocation.arguments = args;
                    prevNode.libraryTarget = libraryInvocation;
                    newTokens.push(libraryInvocation);
                    i += 2;
                    continue;
                }
                thisNode.method = methodNode;
                thisNode.args = args;
                if (newTokens.length > 0) {
                    newTokens[newTokens.length - 1].methodTarget = thisNode;
                }
                i += 2;
                newTokens.push(thisNode);
                continue;
            }
            else {
                newTokens.push(thisNode);
            }
        }
        return newTokens;
    }
    convertVariableToStringNode(node) {
        const wrappedNode = new StringValueNode();
        wrappedNode.content = node.name;
        wrappedNode.value = node.name;
        wrappedNode.originalAbstractNode = node;
        wrappedNode.startPosition = node.startPosition;
        wrappedNode.endPosition = node.endPosition;
        wrappedNode.index = node.index;
        return wrappedNode;
    }
    createTupleLists(tokens) {
        const tokenCount = tokens.length, newTokens = [];
        for (let i = 0; i < tokenCount; i++) {
            const thisToken = tokens[i];
            if (thisToken instanceof TupleListStart) {
                thisToken.isListGroupMember = true;
                if (i + 1 >= tokenCount) {
                    thisToken.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_PARSING_TUPLE_LIST, thisToken, 'Unexpected end of input while parsing tuple list.'));
                    continue;
                }
                const peekT = tokens[i + 1];
                if (peekT instanceof LogicGroup == false) {
                    peekT.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_TYPE_FOR_TUPLE_LIST, peekT, 'Unexpected [' + TypeLabeler.getPrettyTypeName(peekT) + '] while parsing tuple list.'));
                    continue;
                }
                const peek = peekT;
                this.markAllNonVirtualAsListGroupMember([peek]);
                this.retriggerNonVirtualGroupMembers.push([peek]);
                // Each value in the next T_LOGIC_GROUP
                // should be a semantic group instance.
                // The first semantic group will have
                // the tuple list's variable names.
                const listNodeLength = peek.nodes.length;
                if (listNodeLength == 0) {
                    peek.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_MISSING_BODY_TUPLE_LIST, peek, 'Missing tuple list body while parsing tuple list.'));
                    continue;
                }
                const nameGroup = peek.nodes.shift();
                const targetGroupLength = nameGroup.nodes.length;
                const listNames = [];
                for (let j = 0; j < targetGroupLength; j++) {
                    const subNodeT = nameGroup.nodes[j];
                    if (subNodeT instanceof ArgSeparator == false) {
                        let breakFromVarRef = false;
                        if (subNodeT instanceof VariableNode) {
                            if (subNodeT.variableReference == null) {
                                breakFromVarRef = true;
                            }
                            if (subNodeT.variableReference != null) {
                                if (subNodeT.variableReference.pathParts.length > 1) {
                                    breakFromVarRef = true;
                                }
                            }
                            if (subNodeT.name.trim().length == 0) {
                                breakFromVarRef = true;
                            }
                        }
                        if (breakFromVarRef || subNodeT instanceof VariableNode == false) {
                            peek.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_INVALID_TUPLE_LIST_NAME_TYPE, peek, 'Invalid [' + TypeLabeler.getPrettyTypeName(subNodeT) + '] name type found while parsing tuple list.'));
                            continue;
                        }
                        const subNode = subNodeT;
                        listNames.push(this.convertVariableToStringNode(subNode));
                    }
                }
                const listValueLength = listNames.length;
                if (listValueLength == 0) {
                    peek.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_INVALID_MANIFESTED_NAME_GROUP, peek, 'Invalid Name expression produced an invalid name group while parsing tuple list.'));
                    continue;
                }
                // We will simply convert the tuple list syntax into an array node under the hood.
                const arrayNode = new ArrayNode();
                this.createdArrays.push(arrayNode);
                arrayNode.startPosition = thisToken.startPosition;
                arrayNode.endPosition = thisToken.endPosition;
                peek.nodes.forEach((valueNodeCandidateT) => {
                    if (valueNodeCandidateT instanceof SemanticGroup == false) {
                        valueNodeCandidateT.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_INVALID_TUPLE_LIST_VALUE_TYPE_GROUP, valueNodeCandidateT, 'Invalid [' + TypeLabeler.getPrettyTypeName(valueNodeCandidateT) + '] name type found while parsing tuple list value expression.'));
                        return;
                    }
                    const valueNodeCandidate = valueNodeCandidateT;
                    const nestedArrayNode = new ArrayNode();
                    this.createdArrays.push(nestedArrayNode);
                    nestedArrayNode.startPosition = valueNodeCandidate.startPosition;
                    const valueCandidates = [];
                    for (let j = 0; j < targetGroupLength; j++) {
                        const valueToken = valueNodeCandidate.nodes[j];
                        if (typeof valueToken === 'undefined' || valueToken == null) {
                            break;
                        }
                        if (valueToken instanceof ArgSeparator == false) {
                            if (this.isOperand(valueToken) == false) {
                                valueToken.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_INVALID_TUPLE_LIST_VALUE_TYPE, valueToken, 'Unexpected [' + TypeLabeler.getPrettyTypeName(valueToken) + ']  while parsing tuple list value.'));
                                continue;
                            }
                            valueCandidates.push(valueToken);
                        }
                    }
                    for (let j = 0; j < listValueLength; j++) {
                        if (j > valueCandidates.length) {
                            break;
                        }
                        const valueToken = valueCandidates[j];
                        const namedValueNode = new NameValueNode();
                        namedValueNode.startPosition = valueToken.startPosition;
                        namedValueNode.endPosition = valueToken.endPosition;
                        namedValueNode.name = listNames[j];
                        namedValueNode.value = valueToken;
                        nestedArrayNode.nodes.push(namedValueNode);
                        arrayNode.endPosition = valueToken.endPosition;
                        nestedArrayNode.endPosition = valueToken.endPosition;
                    }
                    arrayNode.nodes.push(nestedArrayNode);
                });
                newTokens.push(arrayNode);
                i += 1;
                continue;
            }
            else {
                newTokens.push(thisToken);
            }
        }
        return tokens;
    }
    makeOrderGroup(nodes) {
        const orders = [], nodeCount = nodes.length;
        let orderCount = 0;
        for (let i = 0; i < nodeCount; i++) {
            const thisNode = nodes[i];
            let next = null;
            if (i + 1 < nodeCount) {
                next = nodes[i + 1];
            }
            if (i > 0) {
                if (next == null || next instanceof ArgSeparator) {
                    orderCount += 1;
                    const orderNode = new ValueDirectionNode();
                    orderNode.order = orderCount;
                    orderNode.name = nodes[i - 1];
                    orderNode.directionNode = thisNode;
                    if (this.isOperand(orderNode.name) == false) {
                        orderNode.name.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_INVALID_ORDER_BY_NAME_VALUE, orderNode.name, 'Invalid value or expression supplied for order by name.'));
                    }
                    if (this.isOperand(orderNode.directionNode) == false) {
                        orderNode.directionNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_INVALID_ORDER_BY_SORT_VALUE, orderNode.directionNode, 'Invalid value or expression supplied for order by direction.'));
                    }
                    orderNode.startPosition = orderNode.name.startPosition;
                    orderNode.endPosition = orderNode.directionNode.endPosition;
                    orders.push(orderNode);
                    i += 1;
                    continue;
                }
            }
        }
        return orders;
    }
    getValues(nodes) {
        const valueNode = new ListValueNode(), nodeCount = nodes.length, values = [];
        for (let i = 0; i < nodeCount; i++) {
            const thisNode = nodes[i];
            let next = null;
            if (i + 1 < nodeCount) {
                next = nodes[i + 1];
            }
            if (next == null || next instanceof ArgSeparator) {
                values.push(thisNode);
                i += 1;
                continue;
            }
        }
        valueNode.values = values;
        return valueNode;
    }
    createLibraryInvocations(tokens) {
        const newTokens = [], tokenCount = tokens.length;
        for (let i = 0; i < tokenCount; i++) {
            const token = tokens[i];
            if (token instanceof LanguageOperatorConstruct) {
                if (token.content.includes('.')) {
                    const content = token.content.split('.'), libraryName = content.shift(), methodName = content.join('.');
                    const libraryInvocationConstruct = new LibraryInvocationConstruct();
                    libraryInvocationConstruct.startPosition = token.startPosition;
                    libraryInvocationConstruct.endPosition = token.endPosition;
                    libraryInvocationConstruct.content = token.content;
                    libraryInvocationConstruct.libraryName = libraryName;
                    libraryInvocationConstruct.methodName = methodName;
                    libraryInvocationConstruct.originalAbstractNode = token;
                    if (i + 1 >= tokenCount) {
                        token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_WHILE_PARSING_ARG_GROUP, token, 'Unexpected end of input while parsing argument group.'));
                        continue;
                    }
                    if (tokens[i + 1] instanceof LogicGroup == false) {
                        token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_EXPECTING_ARGUMENT_GROUP, token, 'Unexpected token while parsing method call. Expecting [T_ARG_SEPARATOR] got [' + TypeLabeler.getPrettyTypeName(token) + '].'));
                        continue;
                    }
                    const logicGroup = tokens[i + 1];
                    let argGroup = null;
                    if (logicGroup.nodes.length > 0) {
                        const semanticGroup = logicGroup.nodes[0];
                        argGroup = this.makeArgGroup(semanticGroup.nodes);
                    }
                    if (argGroup == null) {
                        // Wrap everything in an empty argument group.
                        argGroup = new ArgumentGroup();
                        argGroup.startPosition = logicGroup.startPosition;
                        argGroup.endPosition = logicGroup.endPosition;
                        argGroup.originalAbstractNode = logicGroup;
                    }
                    libraryInvocationConstruct.arguments = argGroup;
                    newTokens.push(libraryInvocationConstruct);
                    i += 1;
                    continue;
                }
                else if (token.content == LanguageOperatorRegistry.ARR_ORDERBY) {
                    if (i + 1 >= tokenCount) {
                        token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_PARSING_ORDER_GROUP, token, 'Unexpected end of input while parsing order group.'));
                        continue;
                    }
                    const nextTokenT = tokens[i + 1];
                    if (nextTokenT instanceof LogicGroup == false) {
                        nextTokenT.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_EXPECTING_ORDER_GROUP_FOR_ORDER_BY_OPERAND, nextTokenT, 'Unexpected [' + TypeLabeler.getPrettyTypeName(nextTokenT) + '] while parsing order group.'));
                        continue;
                    }
                    const nextToken = nextTokenT;
                    let subNodes = nextToken.nodes;
                    if (subNodes[0] instanceof SemanticGroup) {
                        const tSg = subNodes[0];
                        subNodes = tSg.nodes;
                    }
                    const orderClauses = this.makeOrderGroup(subNodes);
                    if (orderClauses.length == 0) {
                        token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EMPTY_DIRECTION_GROUP, token, 'Unexpected empty [T_DIRECTION_GROUP]. Must have at least one order clause, and each property must have a direction specified.'));
                    }
                    const orderGroup = new DirectionGroup();
                    orderGroup.orderClauses = orderClauses;
                    if (orderClauses.length > 0) {
                        if (orderClauses[0].directionNode != null) {
                            orderGroup.startPosition = orderClauses[0].directionNode.startPosition;
                        }
                        const ocLastDn = orderClauses[orderClauses.length - 1].directionNode;
                        if (ocLastDn != null) {
                            orderGroup.endPosition = ocLastDn.endPosition;
                        }
                    }
                    newTokens.push(token);
                    newTokens.push(orderGroup);
                    i += 1;
                    continue;
                }
                else if (token.content == LanguageOperatorRegistry.ARR_GROUPBY) {
                    if (i + 1 >= tokenCount) {
                        token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_WHILE_PARSING_GROUP_BY, token, 'Unexpected end of input while parsing group by clause.'));
                        continue;
                    }
                    const nextTokenT = tokens[i + 1];
                    if (nextTokenT instanceof AliasedScopeLogicGroup || nextTokenT instanceof ScopedLogicGroup) {
                        nextTokenT.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_GROUP_BY_SCOPED_GROUP_MUST_BE_ENCLOSED, nextTokenT, 'Type [' + TypeLabeler.getPrettyTypeName(nextTokenT) + '] must be enclosed with parenthesis to be used with groupby.'));
                        continue;
                    }
                    if (nextTokenT instanceof LogicGroup == false) {
                        nextTokenT.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_TOKEN_FOR_GROUP_BY, nextTokenT, 'Unexpected [' + TypeLabeler.getPrettyTypeName(nextTokenT) + '] while parsing group by.'));
                        continue;
                    }
                    const nextToken = nextTokenT;
                    let subNodes = nextToken.nodes;
                    if (subNodes[0] instanceof SemanticGroup) {
                        const tSg = subNodes[0];
                        subNodes = tSg.nodes;
                    }
                    const listValues = this.getValues(subNodes);
                    newTokens.push(token);
                    newTokens.push(listValues);
                    if (i + 2 < tokenCount && i + 3 < tokenCount) {
                        const peekOne = tokens[i + 2], peekTwo = tokens[i + 3];
                        if (NodeHelpers.isVariableMatching(peekOne, LanguageKeywords.ScopeAs)) {
                            if (peekTwo instanceof StringValueNode) {
                                listValues.isNamedNode = true;
                                listValues.parsedName = peekTwo;
                                i += 3;
                                continue;
                            }
                            else {
                                peekTwo.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_GROUP_BY_AS_ALIAS_TYPE, peekTwo, 'Expecting [T_STRING] for group by collection alias; got [' + TypeLabeler.getPrettyTypeName(peekTwo) + '].'));
                            }
                        }
                    }
                    i += 1;
                    continue;
                }
                else if (token.content == LanguageOperatorRegistry.STRUCT_SWITCH) {
                    if (i + 1 >= tokenCount) {
                        token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_WHILE_PARSING_SWITCH_GROUP, token, 'Unexpected end of input while parsing [T_SWITCH_GROUP].'));
                        continue;
                    }
                    const nextT = tokens[i + 1];
                    if (nextT instanceof ScopedLogicGroup == false) {
                        nextT.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_TOKEN_FOR_SWITCH_GROUP, nextT, 'Unexpected [' + TypeLabeler.getPrettyTypeName(nextT) + '] while parsing [T_SWITCH_GROUP].'));
                        continue;
                    }
                    const next = nextT;
                    this.markAllNonVirtualAsSwitch([nextT]);
                    if (next.scope == null || next.scope instanceof LogicGroup == false) {
                        next.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_SWITCH_START_VALUE, next, 'Unexpected input while parsing [T_SWITCH_GROUP].'));
                        continue;
                    }
                    const tSwitchScope = next.scope;
                    if (tSwitchScope.nodes.length == 0 || tSwitchScope.nodes[0] instanceof SemanticGroup == false) {
                        next.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_SWITCH_START_VALUE_NO_VALUE, next, 'Unexpected input while parsing [T_SWITCH_GROUP].'));
                        continue;
                    }
                    const wrapperSemanticGroup = tSwitchScope.nodes[0];
                    this.markAllNonVirtualAsSwitch([tSwitchScope]);
                    if (wrapperSemanticGroup.nodes.length == 0 || wrapperSemanticGroup.nodes[0] instanceof LogicGroup == false) {
                        token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_SWITCH_START_VALUE_NO_VALUE, token, 'Unexpected input while parsing [T_SWITCH_GROUP].'));
                        continue;
                    }
                    this.markAllNonVirtualAsSwitch([token]);
                    this.markAllNonVirtualAsSwitch(wrapperSemanticGroup.nodes);
                    const firstCondition = wrapperSemanticGroup.nodes;
                    if (next.nodes.length == 0) {
                        token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_SWITCH_START_VALUE_NO_VALUE, token, 'Unexpected input while parsing [T_SWITCH_GROUP].'));
                    }
                    const expressionNodeT = tSwitchScope.nodes[0];
                    if (expressionNodeT instanceof SemanticGroup == false) {
                        token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_SWITCH_START_VALUE_NO_VALUE, token, 'Unexpected input while parsing [T_SWITCH_GROUP].'));
                        continue;
                    }
                    const expressionNode = tSwitchScope.nodes[0];
                    this.markAllNonVirtualAsSwitch(expressionNode.nodes);
                    const subTokens = expressionNode.nodes;
                    const switchGroup = new SwitchGroup();
                    const wrapperGroup = new LogicGroup();
                    wrapperGroup.nodes = firstCondition;
                    wrapperGroup.startPosition = wrapperGroup.nodes[0].startPosition;
                    wrapperGroup.endPosition = wrapperGroup.nodes[wrapperGroup.nodes.length - 1].endPosition;
                    const firstCase = new SwitchCase();
                    firstCase.condition = wrapperGroup;
                    const expWrapper = new LogicGroup();
                    expWrapper.nodes.push(subTokens.shift());
                    expWrapper.startPosition = expWrapper.nodes[0].startPosition;
                    expWrapper.endPosition = expWrapper.nodes[expWrapper.nodes.length - 1].endPosition;
                    firstCase.expression = expWrapper;
                    firstCase.startPosition = firstCase.condition.startPosition;
                    firstCase.endPosition = firstCase.expression.nodes[firstCase.expression.nodes.length - 1].endPosition;
                    subTokens.shift();
                    switchGroup.cases.push(firstCase);
                    const subTokenCount = subTokens.length;
                    if (subTokenCount > 0) {
                        if (subTokens[0] instanceof LogicGroup == false) {
                            token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_PARSER_INVALID_SWITCH_TOKEN, token, 'Invalid [' + TypeLabeler.getPrettyTypeName(subTokens[0]) + '] while parsing case statement.'));
                            continue;
                        }
                        for (let c = 0; c < subTokenCount; c++) {
                            const thisToken = subTokens[c];
                            let next = null;
                            if (thisToken instanceof ArgSeparator) {
                                thisToken.isSwitchGroupMember = true;
                                continue;
                            }
                            if (c + 1 < subTokenCount) {
                                next = subTokens[c + 1];
                            }
                            if (next instanceof ScopeAssignmentOperator) {
                                next.isSwitchGroupMember = true;
                                const newCase = new SwitchCase();
                                newCase.condition = thisToken;
                                this.markAllNonVirtualAsSwitch([newCase.condition]);
                                const expWrapper = new LogicGroup();
                                expWrapper.nodes.push(subTokens[c + 2]);
                                expWrapper.startPosition = expWrapper.nodes[0].startPosition;
                                expWrapper.endPosition = expWrapper.nodes[expWrapper.nodes.length - 1].endPosition;
                                newCase.expression = expWrapper;
                                this.markAllNonVirtualAsSwitch([expWrapper]);
                                newCase.startPosition = newCase.condition.startPosition;
                                newCase.endPosition = newCase.expression.nodes[newCase.expression.nodes.length - 1].endPosition;
                                switchGroup.cases.push(newCase);
                                if (c + 3 < subTokenCount) {
                                    if (subTokens[c + 3] instanceof ArgSeparator == false) {
                                        subTokens[c + 3].pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_PARSER_INVALID_SWITCH_TOKEN, subTokens[c + 3], 'Invalid [' + TypeLabeler.getPrettyTypeName(subTokens[c + 3]) + '] while parsing case statement; expecting [T_ARG_SEPARATOR].'));
                                        continue;
                                    }
                                }
                                subTokens[c + 3].isSwitchGroupMember = true;
                                c += 2;
                                continue;
                            }
                        }
                    }
                    newTokens.push(new NullConstant());
                    newTokens.push(token);
                    newTokens.push(switchGroup);
                    i += 1;
                    continue;
                }
                else if (token.content == LanguageOperatorRegistry.ARR_MAKE) {
                    if (i + 1 >= tokenCount) {
                        token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_ARR_MAKE_MISSING_TARGET, token, 'Missing target variable for arr operator.'));
                        continue;
                    }
                    const nextTokenT = tokens[i + 1];
                    if (nextTokenT instanceof LogicGroup == false) {
                        nextTokenT.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_ARR_MAKE_UNEXPECTED_TYPE, nextTokenT, 'Unexpected [' + TypeLabeler.getPrettyTypeName(nextTokenT) + '] while parsing array.'));
                        continue;
                    }
                    const isNextScopedLogicGroup = nextTokenT instanceof ScopedLogicGroup;
                    const nextToken = nextTokenT;
                    let subNodes = nextToken.nodes;
                    if (subNodes[0] instanceof SemanticGroup) {
                        const tSg = subNodes[0];
                        subNodes = tSg.nodes;
                    }
                    if (isNextScopedLogicGroup) {
                        const nextTScoped = nextTokenT;
                        subNodes.unshift(new ScopeAssignmentOperator());
                        subNodes.unshift(nextTScoped.scope);
                    }
                    const values = this.getArrayValues(subNodes);
                    const arrayNode = new ArrayNode();
                    arrayNode.startPosition = token.startPosition;
                    const valueNodeCount = values.values.length;
                    if (valueNodeCount > 0) {
                        arrayNode.endPosition = values.values[valueNodeCount - 1].endPosition;
                    }
                    else {
                        arrayNode.endPosition = token.endPosition;
                    }
                    if (nextToken.end != null) {
                        arrayNode.endPosition = nextToken.end.endPosition;
                    }
                    arrayNode.nodes = values.values;
                    this.createdArrays.push(arrayNode);
                    newTokens.push(arrayNode);
                    i += 1;
                }
                else {
                    newTokens.push(token);
                }
            }
            else {
                newTokens.push(token);
            }
        }
        return newTokens;
    }
    getArrayValues(nodes) {
        const valueNode = new ListValueNode(), nodeCount = nodes.length, values = [];
        for (let i = 0; i < nodeCount; i++) {
            const thisNode = nodes[i];
            if (thisNode instanceof ArgSeparator) {
                thisNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_ARR_UNEXPECT_ARG_SEPARATOR, thisNode, 'Unexpected [T_ARG_SEPARATOR] while parsing array.'));
                continue;
            }
            if (thisNode instanceof ScopeAssignmentOperator) {
                if (i == 0) {
                    thisNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_ARR_KEY_PAIR_MISSING_KEY, thisNode, 'Missing key for key/value pair while parsing array.'));
                    continue;
                }
                else {
                    if (nodes[i - 1] instanceof ArgSeparator) {
                        thisNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_ARR_KEY_PAIR_MISSING_KEY, thisNode, 'Missing key for key/value pair while parsing array.'));
                        continue;
                    }
                }
            }
            let next = null;
            if (i + 1 < nodeCount) {
                next = nodes[i + 1];
            }
            if (next instanceof ScopeAssignmentOperator) {
                if (i + 2 >= nodeCount) {
                    next.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_ARR_MAKE_MISSING_ARR_KEY_PAIR_VALUE, next, 'Missing key/pair value while parsing array.'));
                    continue;
                }
                const keyValue = nodes[i + 2];
                const namedValueNode = new NameValueNode();
                if (this.isValidArrayKeyNode(thisNode) == false) {
                    thisNode.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_ARR_KEY_PAIR_INVALID_KEY_TYPE, thisNode, 'Invalid [' + TypeLabeler.getPrettyTypeName(thisNode) + '] type for key/value key.'));
                    continue;
                }
                namedValueNode.name = thisNode;
                namedValueNode.value = keyValue;
                namedValueNode.startPosition = thisNode.startPosition;
                namedValueNode.endPosition = keyValue.endPosition;
                values.push(namedValueNode);
                i += 3;
                continue;
            }
            if (next == null || next instanceof ArgSeparator) {
                const namedValueNode = new NameValueNode();
                namedValueNode.value = thisNode;
                namedValueNode.startPosition = thisNode.startPosition;
                namedValueNode.endPosition = thisNode.endPosition;
                values.push(namedValueNode);
                i += 1;
                continue;
            }
        }
        valueNode.values = values;
        return valueNode;
    }
    isValidArrayKeyNode(node) {
        if (node instanceof NumberNode || node instanceof StringValueNode) {
            return true;
        }
        return false;
    }
    associateModifiers(tokens) {
        const newNodes = [], tokenCount = tokens.length;
        let applyModifiersToNode = null;
        for (let i = 0; i < tokenCount; i++) {
            const node = tokens[i];
            if (node instanceof ModifierSeparator) {
                const newNodeCount = newNodes.length;
                if (newNodeCount == 0) {
                    node.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_MODIFIER_SEPARATOR, node, 'Unexpected [T_MODIFIER_SEPARATOR] while parsing input text.'));
                    continue;
                }
                this.isRealModifierSeparator.set(node, true);
                applyModifiersToNode = newNodes[newNodeCount - 1];
                if (applyModifiersToNode.modifierChain == null) {
                    applyModifiersToNode.modifierChain = new ModifierChainNode();
                    applyModifiersToNode.modifierChain.modifierTarget = applyModifiersToNode;
                    this.createdModifierChains.push(applyModifiersToNode.modifierChain);
                }
                if (i + 1 >= tokenCount) {
                    node.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_WHILE_PARSING_MODIFIER_DETAILS, node, 'Unexpected end of input while preparing to gather modifier details.'));
                    continue;
                }
                const results = this.scanToEndOfModifier(tokens.slice(i + 1));
                const resultCount = results.length;
                if (resultCount > 1) {
                    if (results[1] instanceof InlineBranchSeparator) {
                        this.isRealModifierSeparator.set(results[1], true);
                        const firstToken = results[0];
                        if (firstToken instanceof AdditionOperator) {
                            results[0] = this.wrapArithmeticModifier(firstToken, 'add');
                        }
                        else if (firstToken instanceof SubtractionOperator) {
                            results[0] = this.wrapArithmeticModifier(firstToken, 'subtract');
                        }
                        else if (firstToken instanceof DivisionOperator) {
                            results[0] = this.wrapArithmeticModifier(firstToken, 'divide');
                        }
                        else if (firstToken instanceof MultiplicationOperator) {
                            results[0] = this.wrapArithmeticModifier(firstToken, 'multiply');
                        }
                        else if (firstToken instanceof ModulusOperator) {
                            results[0] = this.wrapArithmeticModifier(firstToken, 'mod');
                        }
                    }
                }
                if (resultCount == 0) {
                    node.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNSET_MODIFIER_DETAILS, node, 'Invalid or missing modifier details.'));
                    continue;
                }
                if (results[0] instanceof ModifierNameNode == false) {
                    node.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_MODIFIER_NAME_NOT_START_OF_DETAILS, node, 'Invalid [' + TypeLabeler.getPrettyTypeName(results[0]) + ']; expecting [T_MODIFIER_NAME]'));
                    continue;
                }
                const modifier = this.createModifier(results);
                if (modifier.nameNode != null) {
                    this.modifierNameMapping.set(modifier.nameNode, modifier);
                }
                i += resultCount;
                applyModifiersToNode.modifierChain.modifierChain.push(modifier);
                applyModifiersToNode.modifierChain.updateValues();
                continue;
            }
            else {
                newNodes.push(node);
            }
        }
        return newNodes;
    }
    createNullCoalescenceGroups(tokens) {
        const newTokens = [], tokenCount = tokens.length;
        for (let i = 0; i < tokenCount; i++) {
            const node = tokens[i];
            if (node instanceof NullCoalesceOperator) {
                const left = newTokens.pop();
                if (i + 1 >= tokenCount) {
                    node.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_WHILE_PARSING_NULL_COALESCENCE_GROUP, node, 'Unexpected end of input while parsing input text.'));
                    continue;
                }
                const right = tokens[i + 1];
                const nullCoalescenceGroup = new NullCoalescenceGroup();
                nullCoalescenceGroup.left = left;
                nullCoalescenceGroup.right = right;
                newTokens.push(nullCoalescenceGroup);
                i += 1;
                continue;
            }
            else {
                newTokens.push(node);
            }
        }
        return newTokens;
    }
    isOperand(token) {
        return token instanceof VariableNode || token instanceof LogicGroup ||
            token instanceof StringValueNode || token instanceof NumberNode ||
            token instanceof FalseConstant || token instanceof NullConstant ||
            token instanceof TrueConstant || token instanceof LibraryInvocationConstruct ||
            token instanceof DirectionGroup || token instanceof ListValueNode ||
            token instanceof SwitchGroup || token instanceof ArrayNode || token instanceof TupleListStart ||
            token instanceof ScopeAssignmentOperator;
    }
    isProperMethodChainTargetStrict(token) {
        return token instanceof LogicGroup ||
            token instanceof StringValueNode || token instanceof NumberNode ||
            token instanceof FalseConstant || token instanceof NullConstant ||
            token instanceof TrueConstant || token instanceof LibraryInvocationConstruct ||
            token instanceof DirectionGroup || token instanceof ListValueNode ||
            token instanceof SwitchGroup || token instanceof ArrayNode;
    }
    assertOperandRight(tokens, i) {
        if (i + 1 > tokens.length - 1) {
            tokens[i].pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_END_OF_INPUT, tokens[i], 'Unexpected end of input; expecting operand for operator ' + TypeLabeler.getPrettyTypeName(tokens[i]) + ' near "' + LineRetriever.getNearText(tokens[i]) + '".'));
            return false;
        }
        const token = tokens[i + 1];
        if (!this.isOperand(token)) {
            if (token instanceof LogicalAndOperator) {
                const nDistance = NodeHelpers.distance(tokens[i], token);
                if (nDistance < 0) {
                    return true;
                }
            }
            token.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_EXPECTING_OPERAND, tokens[i], 'Expecting operand, found ' + TypeLabeler.getPrettyTypeName(token) + ' near "' + LineRetriever.getNearText(tokens[i]) + '".'));
            return false;
        }
        return true;
    }
    resolveValueRight(nodes, index) {
        let value = null, negationCount = 0, lastNegation = null;
        if (nodes.length > 0) {
            let doContinue = true;
            while (doContinue) {
                const curNode = nodes[index];
                if (curNode instanceof LogicalNegationOperator) {
                    lastNegation = curNode;
                    negationCount += 1;
                    index += 1;
                    doContinue = false;
                    continue;
                }
                else if (this.isOperand(curNode)) {
                    value = curNode;
                    doContinue = false;
                    break;
                }
                else {
                    doContinue = false;
                    break;
                }
            }
        }
        if (negationCount % 2 != 0) {
            const logicGroup = new LogicGroup();
            logicGroup.nodes = [];
            if (lastNegation != null && value != null) {
                logicGroup.nodes.push(lastNegation);
                logicGroup.nodes.push(value);
            }
            return {
                value: logicGroup,
                negationCount
            };
        }
        return {
            value: value,
            negationCount: negationCount
        };
    }
    correctTypes(tokens) {
        const newNodes = [];
        tokens.forEach((node) => {
            if (node instanceof ModifierValueSeparator) {
                const branchSeparator = new InlineBranchSeparator();
                branchSeparator.startPosition = node.startPosition;
                branchSeparator.endPosition = node.endPosition;
                branchSeparator.originalAbstractNode = node;
                newNodes.push(branchSeparator);
            }
            else if (node instanceof ModifierValueNode) {
                const varNode = new VariableNode();
                varNode.name = node.value;
                varNode.startPosition = node.startPosition;
                varNode.endPosition = node.endPosition;
                varNode.modifierChain = node.modifierChain;
                varNode.originalAbstractNode = node;
                this.createdVariables.push(varNode);
                newNodes.push(varNode);
            }
            else {
                newNodes.push(node);
            }
        });
        return newNodes;
    }
    createTernaryGroups(tokens) {
        let newTokens = [];
        const tokenCount = tokens.length;
        for (let i = 0; i < tokenCount; i++) {
            const node = tokens[i];
            if (node instanceof InlineTernarySeparator) {
                const separator = this.seek(InlineBranchSeparator, i + 1);
                if (separator.node?.originalAbstractNode != null) {
                    if (separator.node.originalAbstractNode instanceof ModifierValueSeparator) {
                        separator.node.originalAbstractNode.isTenaryBranchSeparator = true;
                    }
                }
                if (separator.found == false) {
                    node.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_TERNARY_EXPECTING_BRANCH_SEPARATOR, node, 'Unexpected end of input; expecting [T_BRANCH_SEPARATOR].'));
                    continue;
                }
                const result = this.collectUntil(newTokens);
                newTokens = result.tokens;
                const condition = result.collectedTokens[0];
                // const targetTokenIndex = (separator.foundAt ?? 0) - i  -1;
                const targetTokenIndex = (separator.foundAt ?? 0);
                if (targetTokenIndex >= tokenCount) {
                    node.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_WHILE_PARSING_TERNARY_GROUP, node, 'Unexpected end of input while parsing ternary group.'));
                    continue;
                }
                const truthBranch = tokens.slice(i + 1, targetTokenIndex);
                if (truthBranch.length > 1 || truthBranch.length == 0) {
                    node.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_TERNARY_UNEXPECTED_EXPRESSION_LENGTH, node, 'Unexpected number of operations within ternary truth branch.'));
                    continue;
                }
                const truthBranchNodes = truthBranch[0];
                const falseBranchStart = (separator.foundAt ?? 0) + 1;
                if (falseBranchStart >= tokenCount) {
                    node.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_WHILE_PARSING_TERNARY_GROUP_FALSE_BRANCH, node, 'Unexpected end of input while parsing ternary false execution branch.'));
                    continue;
                }
                const falseBranch = tokens[falseBranchStart];
                const ternaryStructure = new TernaryCondition();
                ternaryStructure.head = condition;
                ternaryStructure.truthBranch = truthBranchNodes;
                ternaryStructure.falseBranch = falseBranch;
                newTokens.push(ternaryStructure);
                const targetJumpIndex = (separator.foundAt ?? 0) + 1;
                i = targetJumpIndex;
                continue;
            }
            else {
                newTokens.push(node);
            }
        }
        return newTokens;
    }
    groupNodesByType(nodes, type) {
        const newNodes = [], nodeCount = nodes.length;
        for (let i = 0; i < nodeCount; i++) {
            const node = nodes[i];
            if (i > 0 && node instanceof type) {
                const left = [];
                const poppedLeft = newNodes.pop();
                left.push(poppedLeft);
                if (!this.assertOperandRight(nodes, i)) {
                    continue;
                }
                if (i + 1 >= nodeCount) {
                    node.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_WHILE_REDUCING_NEGATION_OPERATORS, node, 'Unexpected end of input while parsing input text.'));
                    continue;
                }
                const right = nodes[i + 1];
                if (typeof right !== 'undefined') {
                    right.isVirtualGroupMember = true;
                }
                if (typeof poppedLeft !== 'undefined') {
                    poppedLeft.isVirtualGroupMember = true;
                }
                node.isVirtualGroupMember = true;
                const logicGroup = new LogicGroup();
                logicGroup.nodes = left;
                logicGroup.nodes.push(node);
                logicGroup.nodes.push(right);
                i += 1;
                newNodes.push(logicGroup);
                continue;
            }
            else {
                newNodes.push(node);
            }
        }
        return newNodes;
    }
    flagNodeAsOperatorResolve(node) {
        if (node instanceof LogicGroup) {
            if (node.nodes.length == 0) {
                return;
            }
            this.flagNodeAsOperatorResolve(node.nodes[node.nodes.length - 1]);
        }
        node.isVirtualGroupOperatorResolve = true;
        if (node.next instanceof StatementSeparatorNode == false) {
            node.producesVirtualStatementTerminator = true;
        }
    }
    groupNodesByOperatorType(nodes) {
        const newNodes = [], nodeCount = nodes.length;
        for (let i = 0; i < nodeCount; i++) {
            const node = nodes[i];
            if (i > 0 && LanguageParser.isOperatorType(node)) {
                const left = [];
                const poppedLeft = newNodes.pop();
                left.push(poppedLeft);
                if (!this.assertOperandRight(nodes, i)) {
                    continue;
                }
                if (i + 1 >= nodeCount) {
                    node.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_WHILE_REDUCING_NEGATION_OPERATORS, node, 'Unexpected end of input while parsing input text.'));
                    continue;
                }
                const right = nodes[i + 1];
                this.flagNodeAsOperatorResolve(right);
                if (LanguageParser.isAssignmentOperator(node)) {
                    this.tracedAssignments.push(new StaticTracedAssignment(poppedLeft, right, node));
                }
                if (typeof poppedLeft !== 'undefined') {
                    poppedLeft.isVirtualGroupMember = true;
                }
                if (typeof right !== 'undefined') {
                    right.isVirtualGroupMember = true;
                }
                node.isVirtualGroupMember = true;
                const logicGroup = new LogicGroup();
                logicGroup.nodes = left;
                logicGroup.nodes.push(node);
                logicGroup.nodes.push(right);
                i += 1;
                newNodes.push(logicGroup);
                continue;
            }
            else {
                newNodes.push(node);
            }
        }
        return newNodes;
    }
    applyOperationOrder(nodes) {
        nodes = this.groupNodesByType(nodes, ExponentiationOperator);
        nodes = this.groupNodesByType(nodes, MultiplicationOperator);
        nodes = this.groupNodesByType(nodes, DivisionOperator);
        nodes = this.groupNodesByType(nodes, AdditionOperator);
        nodes = this.groupNodesByType(nodes, SubtractionOperator);
        return nodes;
    }
    createLogicGroupsToRemoveMethodInvocationAmbiguity(nodes) {
        const newNodes = [], nodeLen = nodes.length, lastNodeIndex = nodeLen - 1;
        for (let i = 0; i < nodeLen; i++) {
            const thisNode = nodes[i];
            if (thisNode instanceof MethodInvocationNode) {
                const lastNode = newNodes.pop();
                const wrapperGroup = new LogicGroup();
                wrapperGroup.startPosition = lastNode.startPosition;
                if (typeof lastNode !== 'undefined') {
                    lastNode.isVirtualGroupMember = true;
                }
                if (typeof thisNode !== 'undefined') {
                    thisNode.isVirtualGroupMember = true;
                }
                wrapperGroup.nodes.push(lastNode);
                wrapperGroup.nodes.push(thisNode);
                if (i != lastNodeIndex) {
                    for (let j = i + 1; j < nodeLen; j++) {
                        if (nodes[j] instanceof MethodInvocationNode) {
                            if (typeof nodes[j] === 'undefined') {
                                continue;
                            }
                            nodes[j].isVirtualGroupMember = true;
                            wrapperGroup.nodes.push(nodes[j]);
                            if (j == lastNodeIndex) {
                                j += 1;
                                break;
                            }
                        }
                        else {
                            if (j == lastNodeIndex) {
                                j += 1;
                                break;
                            }
                            i = j - 1;
                            break;
                        }
                    }
                }
                wrapperGroup.endPosition = wrapperGroup.nodes[wrapperGroup.nodes.length - 1].endPosition;
                newNodes.push(wrapperGroup);
            }
            else {
                newNodes.push(thisNode);
            }
        }
        return newNodes;
    }
    countTypeRight(tokens, start, type) {
        let count = 0;
        for (let i = start; i < tokens.length; i++) {
            if (tokens[i] instanceof type) {
                count += 1;
            }
            else {
                break;
            }
        }
        return count;
    }
    createLogicGroupsToResolveOperatorAmbiguity(nodes) {
        nodes = this.groupNodesByType(nodes, GreaterThanEqualCompOperator);
        nodes = this.groupNodesByType(nodes, GreaterThanCompOperator);
        nodes = this.groupNodesByType(nodes, LessThanEqualCompOperator);
        nodes = this.groupNodesByType(nodes, LessThanCompOperator);
        nodes = this.groupNodesByType(nodes, StrictEqualCompOperator);
        nodes = this.groupNodesByType(nodes, EqualCompOperator);
        nodes = this.groupNodesByType(nodes, NotStrictEqualCompOperator);
        nodes = this.groupNodesByType(nodes, NotEqualCompOperator);
        nodes = this.groupNodesByType(nodes, SpaceshipCompOperator);
        nodes = this.groupNodesByOperatorType(nodes);
        return nodes;
    }
    createModifier(tokens) {
        const modifierName = tokens.shift(), values = [], tokenCount = tokens.length;
        if (tokens.length > 0 && tokens[0] instanceof LogicGroup) {
            const unwrapped = this.unpack(tokens[0].nodes), tArgGroup = this.makeArgGroup(unwrapped);
            // Reassign the end position.
            tArgGroup.endPosition = tokens[0].endPosition;
            const modifier = new ModifierNode();
            modifier.nameNode = modifierName;
            modifier.methodStyleArguments = tArgGroup;
            if (tokens.length > 1) {
                for (let i = 1; i < tokenCount; i++) {
                    tokens[i].pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_MODIFIER_UNEXPECTED_TOKEN_METHOD_SYNTAX, tokens[i], 'Unexpected [' + TypeLabeler.getPrettyTypeName(tokens[i]) + '] while parsing modifier argument group. Expecting [T_MODIFIER_SEPARATOR] or end of current expression.'));
                }
            }
            modifier.startPosition = modifierName.startPosition;
            modifier.endPosition = tArgGroup.endPosition;
            return modifier;
        }
        for (let i = 0; i < tokenCount; i++) {
            if (tokens[i] instanceof ModifierValueSeparator || tokens[i] instanceof InlineBranchSeparator) {
                if (i + 1 >= tokenCount) {
                    tokens[i].pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_MODIFIER_UNEXPECTED_END_OF_VALUE_LIST, tokens[i], 'Unexpected end of modifier value list while parsing modifier.'));
                    return tokens;
                }
                const next = tokens[i + 1];
                if (this.isValidModifierValue(next)) {
                    if (next instanceof VariableNode) {
                        if (next.variableReference != null && next.variableReference.pathParts.length > 0) {
                            next.variableReference.pathParts.forEach((combinedPart) => {
                                if (combinedPart instanceof PathNode) {
                                    const modifierValue = new ModifierValueNode();
                                    modifierValue.startPosition = combinedPart.startPosition;
                                    modifierValue.endPosition = combinedPart.endPosition;
                                    modifierValue.value = combinedPart.name;
                                    modifierValue.name = combinedPart.name;
                                    modifierValue.originalAbstractNode = combinedPart;
                                    if (modifierValue.endPosition == null && modifierValue.startPosition == null) {
                                        modifierValue.startPosition = next.startPosition;
                                        modifierValue.endPosition = next.endPosition;
                                    }
                                    values.push(modifierValue);
                                }
                            });
                        }
                        else {
                            const modifierValue = new ModifierValueNode();
                            modifierValue.startPosition = next.startPosition;
                            modifierValue.endPosition = next.endPosition;
                            modifierValue.value = next.name;
                            modifierValue.name = next.name;
                            modifierValue.originalAbstractNode = next;
                            values.push(modifierValue);
                        }
                    }
                    else {
                        values.push(next);
                    }
                    i += 1;
                    continue;
                }
                else {
                    next.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_MODIFIER_UNEXPECTED_VALUE, next, 'Unexpected [' + TypeLabeler.getPrettyTypeName(next) + '] while parsing modifier value.'));
                    return tokens;
                }
            }
        }
        const modifierNode = new ModifierNode();
        modifierNode.nameNode = modifierName;
        modifierNode.valueNodes = values;
        modifierNode.startPosition = modifierNode.nameNode.startPosition;
        if (modifierNode.valueNodes.length > 0) {
            modifierNode.endPosition = modifierNode.valueNodes[modifierNode.valueNodes.length - 1].endPosition;
        }
        else {
            modifierNode.endPosition = modifierNode.nameNode.endPosition;
        }
        return modifierNode;
    }
    scanToEndOfModifier(tokens) {
        const subTokens = [], tokenCount = tokens.length;
        for (let i = 0; i < tokenCount; i++) {
            const subToken = tokens[i];
            if (subToken instanceof ModifierValueSeparator) {
                const subTokenCount = subTokens.length;
                if (subTokenCount == 0) {
                    subToken.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_WHILE_PARSING_MODIFIER_VALUE, subToken, 'Unexpected end of input while parsing modifier value.'));
                    continue;
                }
                this.isRealModifierSeparator.set(subToken, true);
                const last = subTokens[subTokenCount - 1];
                if (NodeHelpers.distance(last, subToken) > 1) {
                    break;
                }
            }
            if (subToken instanceof StringValueNode || subToken instanceof VariableNode || subToken instanceof NumberNode) {
                const subTokenCount = subTokens.length;
                if (subTokenCount > 0) {
                    const last = subTokens[subTokenCount - 1];
                    if (last instanceof LogicGroup) {
                        break;
                    }
                }
            }
            if (subToken instanceof ModifierSeparator ||
                subToken instanceof LogicGroupEnd ||
                subToken instanceof LogicGroupBegin ||
                subToken instanceof EqualCompOperator ||
                subToken instanceof GreaterThanCompOperator ||
                subToken instanceof GreaterThanEqualCompOperator ||
                subToken instanceof LessThanCompOperator ||
                subToken instanceof LessThanEqualCompOperator ||
                subToken instanceof NotEqualCompOperator ||
                subToken instanceof NotStrictEqualCompOperator ||
                subToken instanceof SpaceshipCompOperator ||
                subToken instanceof StrictEqualCompOperator ||
                subToken instanceof LogicalAndOperator ||
                subToken instanceof LogicalOrOperator ||
                subToken instanceof LogicalXorOperator ||
                subToken instanceof NullCoalesceOperator ||
                subToken instanceof StringConcatenationOperator ||
                subToken instanceof LanguageOperatorConstruct ||
                subToken instanceof LibraryInvocationConstruct ||
                subToken instanceof MethodInvocationNode ||
                subToken instanceof LogicalNegationOperator) {
                break;
            }
            else {
                subTokens.push(subToken);
            }
        }
        return subTokens;
    }
    static isAssignmentOperatorNode(node) {
        return node instanceof AdditionAssignmentOperator || node instanceof DivisionAssignmentOperator ||
            node instanceof LeftAssignmentOperator || node instanceof ModulusAssignmentOperator ||
            node instanceof MultiplicationAssignmentOperator || node instanceof SubtractionOperator;
    }
    insertAutomaticStatementSeparators(tokens) {
        const tokenCount = tokens.length, adjustedTokens = [];
        for (let i = 0; i < tokenCount; i++) {
            const thisToken = tokens[i];
            if (LanguageParser.isAssignmentOperatorNode(thisToken)) {
                if (i + 2 < tokenCount) {
                    const peek = tokens[i + 2];
                    if (peek instanceof StatementSeparatorNode == false) {
                        adjustedTokens.push(thisToken);
                        adjustedTokens.push(tokens[i + 1]);
                        adjustedTokens.push(new StatementSeparatorNode());
                        i += 1;
                        continue;
                    }
                    else {
                        adjustedTokens.push(thisToken);
                        adjustedTokens.push(tokens[i + 1]);
                        adjustedTokens.push(tokens[i + 2]);
                        i += 2;
                        continue;
                    }
                }
                else {
                    adjustedTokens.push(thisToken);
                    adjustedTokens.push(tokens[i + 1]);
                    adjustedTokens.push(new StatementSeparatorNode());
                    break;
                }
            }
            else {
                adjustedTokens.push(thisToken);
            }
        }
        return tokens;
    }
    createSemanticGroups(tokens) {
        const groups = [], tokenCount = tokens.length;
        let groupNodes = [];
        for (let i = 0; i < tokenCount; i++) {
            if (tokens[i] instanceof StatementSeparatorNode) {
                const semanticGroup = new SemanticGroup();
                semanticGroup.nodes = groupNodes;
                semanticGroup.separatorToken = tokens[i];
                groups.push(semanticGroup);
                groupNodes = [];
                continue;
            }
            else {
                groupNodes.push(tokens[i]);
                if (i + 1 >= tokenCount) {
                    const semanticGroup = new SemanticGroup();
                    semanticGroup.nodes = groupNodes;
                    groups.push(semanticGroup);
                    break;
                }
            }
        }
        return groups;
    }
    hydrateMissingNodePositions(tokens) {
        const visited = new Set();
        tokens.forEach((token) => {
            this.hydrateNodePositions(token, visited);
        });
    }
    hydrateNodePositions(value, visited) {
        if (value == null) {
            return;
        }
        if (Array.isArray(value)) {
            value.forEach((item) => this.hydrateNodePositions(item, visited));
            return;
        }
        if (value instanceof Map) {
            value.forEach((item) => this.hydrateNodePositions(item, visited));
            return;
        }
        if (typeof value !== 'object') {
            return;
        }
        const node = value instanceof AbstractNode ? value : null;
        if (node != null) {
            if (visited.has(node)) {
                return;
            }
            visited.add(node);
        }
        Object.entries(value).forEach(([key, child]) => {
            if (this.positionHydrationSkippedKeys.has(key)) {
                return;
            }
            this.hydrateNodePositions(child, visited);
        });
        if (node != null && (node.startPosition == null || node.endPosition == null)) {
            const inferredRange = this.inferNodeRange(node, new Set());
            if (node.startPosition == null) {
                node.startPosition = inferredRange.start;
            }
            if (node.endPosition == null) {
                node.endPosition = inferredRange.end;
            }
        }
    }
    inferNodeRange(value, seen) {
        if (value == null || seen.has(value)) {
            return {
                start: null,
                end: null
            };
        }
        if (Array.isArray(value)) {
            let start = null, end = null;
            seen.add(value);
            value.forEach((item) => {
                const range = this.inferNodeRange(item, seen);
                start = this.earlierPosition(start, range.start);
                end = this.laterPosition(end, range.end);
            });
            return { start, end };
        }
        if (value instanceof Map) {
            let start = null, end = null;
            seen.add(value);
            value.forEach((item) => {
                const range = this.inferNodeRange(item, seen);
                start = this.earlierPosition(start, range.start);
                end = this.laterPosition(end, range.end);
            });
            return { start, end };
        }
        if (typeof value !== 'object') {
            return {
                start: null,
                end: null
            };
        }
        seen.add(value);
        let start = null, end = null;
        if (value instanceof AbstractNode) {
            start = this.earlierPosition(start, value.startPosition);
            end = this.laterPosition(end, value.endPosition);
        }
        Object.entries(value).forEach(([key, child]) => {
            if (this.positionHydrationSkippedKeys.has(key)) {
                return;
            }
            const range = this.inferNodeRange(child, seen);
            start = this.earlierPosition(start, range.start);
            end = this.laterPosition(end, range.end);
        });
        return { start, end };
    }
    earlierPosition(current, candidate) {
        if (candidate == null) {
            return current;
        }
        if (current == null || candidate.offset < current.offset) {
            return candidate;
        }
        return current;
    }
    laterPosition(current, candidate) {
        if (candidate == null) {
            return current;
        }
        if (current == null || candidate.offset > current.offset) {
            return candidate;
        }
        return current;
    }
    findLogicalGroupEnd(root, nodes) {
        const subNodes = [], nodeCount = nodes.length;
        let end = null, skipCount = 0, i = 0;
        for (i; i < nodeCount; i++) {
            const node = nodes[i];
            skipCount += 1;
            if (node instanceof LogicGroupEnd) {
                end = node;
                break;
            }
            else if (node instanceof LogicGroupBegin) {
                if (i + 1 >= nodeCount) {
                    node.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_UNEXPECTED_EOI_WHILE_PARSING_LOGIC_GROUP_END, node, 'Unexpected end of input while parsing input text.'));
                    return {
                        found: false,
                        logicalGroup: null,
                        skipCount: skipCount
                    };
                }
                const subGroup = this.findLogicalGroupEnd(node, nodes.slice(i + 1));
                if (subGroup.found) {
                    if (subGroup.logicalGroup != null) {
                        subNodes.push(subGroup.logicalGroup);
                    }
                    skipCount += subGroup.skipCount;
                    i += subGroup.skipCount;
                    continue;
                }
                else {
                    continue;
                }
            }
            else {
                subNodes.push(node);
            }
        }
        if (end == null) {
            root.pushError(AntlersError.makeSyntaxError(AntlersErrorCodes.TYPE_LOGIC_GROUP_NO_END, root, 'Unexpected end of input while parsing logic group.'));
            return {
                found: false,
                logicalGroup: null,
                skipCount: skipCount
            };
        }
        const parser = new LanguageParser();
        parser.setIsRoot(false);
        let logicalGroup = new LogicGroup();
        if (subNodes.length >= 2 && subNodes[1] instanceof ScopeAssignmentOperator) {
            logicalGroup = new ScopedLogicGroup();
            logicalGroup.scopeOperator = subNodes[1];
            if (i + 2 < nodeCount && nodes[i + 1] instanceof VariableNode && nodes[i + 2] instanceof StringValueNode) {
                const candidateVarNode = nodes[i + 1];
                if (candidateVarNode.name == LanguageKeywords.ScopeAs) {
                    const aliasNode = nodes[i + 2];
                    logicalGroup = new AliasedScopeLogicGroup();
                    if (logicalGroup instanceof AliasedScopeLogicGroup) {
                        logicalGroup.alias = aliasNode;
                    }
                    skipCount += 2;
                }
            }
            const scopedNodes = subNodes.splice(0, 2);
            if (logicalGroup instanceof ScopedLogicGroup || logicalGroup instanceof AliasedScopeLogicGroup) {
                logicalGroup.scope = scopedNodes[0];
            }
        }
        logicalGroup.nodes = parser.parse(subNodes);
        logicalGroup.start = root;
        logicalGroup.end = end;
        logicalGroup.startPosition = root.startPosition;
        logicalGroup.endPosition = end.endPosition;
        if (logicalGroup instanceof AliasedScopeLogicGroup) {
            if (logicalGroup.alias != null) {
                logicalGroup.endPosition = logicalGroup.alias.endPosition;
            }
        }
        return {
            found: true,
            logicalGroup: logicalGroup,
            skipCount: skipCount
        };
    }
    unpack(tokens) {
        if (tokens.length == 0) {
            return [];
        }
        if (tokens[0] instanceof SemanticGroup) {
            const wSg = tokens[0];
            return this.unpack(wSg.nodes);
        }
        return tokens;
    }
    collectUntil(tokens) {
        const len = tokens.length;
        let collectedTokens = [];
        for (let i = len - 1; i >= 0; i--) {
            if (LanguageParser.isAssignmentOperator(tokens[i])) {
                break;
            }
            const token = tokens.pop();
            collectedTokens.push(token);
        }
        collectedTokens = collectedTokens.reverse();
        if (collectedTokens.length >= 3) {
            const parser = new LanguageParser();
            parser.setIsRoot(false);
            return {
                collectedTokens: this.unpack(parser.parse(collectedTokens)),
                tokens: tokens
            };
        }
        return {
            collectedTokens: collectedTokens,
            tokens: tokens
        };
    }
    seek(type, startingAt) {
        for (let i = startingAt; i < this.tokens.length; i++) {
            if (this.tokens[i] instanceof type) {
                return {
                    found: true,
                    foundAt: i,
                    node: this.tokens[i]
                };
            }
        }
        return {
            found: false,
            foundAt: null,
            node: null
        };
    }
}
