// td-antlers-parser/out/antlers/tags/core/sections/yield.js
import { makeTagDoc } from '../../../../documentation/utils.js';
export class YieldContext {
    constructor(node) {
        this.node = node;
    }
}
const Yields = {
    tagName: "yields",
    hideFromCompletions: false,
    injectParentScope: false,
    allowsArbitraryParameters: false,
    allowsContentClose: true,
    parameters: [],
    introducedIn: null,
    requiresClose: false,
    resolveSpecialType: (node, project) => {
        const context = new YieldContext(node);
        return {
            context: context,
            issues: [],
        };
    },
};
const Yield = {
    tagName: "yield",
    hideFromCompletions: false,
    injectParentScope: false,
    allowsArbitraryParameters: false,
    allowsContentClose: true,
    parameters: [],
    introducedIn: null,
    requiresClose: false,
    resolveSpecialType: (node, project) => {
        const context = new YieldContext(node);
        return {
            context: context,
            issues: [],
        };
    },
    resolveDocumentation: (params) => {
        return makeTagDoc('yield Tag', 'The `yield` tag may be used to push rendered content to named regions within a template\'s structure.', 'https://statamic.dev/tags/yield');
    }
};
export { Yield, Yields };
