// td-antlers-parser/out/references/referenceManager.js
class ReferenceManager {
    constructor() {
        this.partialReferences = new Map();
        this.cacheReferences = new Map();
        this.removesPageScope = new Map();
    }
    clearAllReferences(documentUri) {
        this.clearPartialReferences(documentUri);
        this.clearCacheReferences(documentUri);
    }
    registerPartialReferences(documentUri, nodes) {
        if (this.partialReferences.has(documentUri) == false) {
            this.partialReferences.set(documentUri, []);
        }
        this.partialReferences.set(documentUri, nodes);
    }
    registerCacheReferences(documentUri, nodes) {
        if (this.cacheReferences.has(documentUri) == false) {
            this.cacheReferences.set(documentUri, []);
        }
        this.cacheReferences.set(documentUri, nodes);
    }
    clearPartialReferences(documentUri) {
        if (this.partialReferences.has(documentUri)) {
            this.partialReferences.set(documentUri, []);
        }
    }
    clearRemovesPageScope(documentUri) {
        if (this.removesPageScope.has(documentUri)) {
            this.removesPageScope.delete(documentUri);
        }
    }
    pageScopeDisabled(documentUri) {
        return this.removesPageScope.has(documentUri);
    }
    clearCacheReferences(documentUri) {
        if (this.cacheReferences.has(documentUri)) {
            this.cacheReferences.set(documentUri, []);
        }
    }
    setRemovesPageScope(documentUri, node) {
        this.removesPageScope.set(documentUri, node);
    }
    getPartialReferences(documentUri) {
        if (this.partialReferences.has(documentUri) == false) {
            return [];
        }
        return this.partialReferences.get(documentUri);
    }
    getCacheReferencnes(documentUri) {
        if (this.cacheReferences.has(documentUri) == false) {
            return [];
        }
        return this.cacheReferences.get(documentUri);
    }
    hasPartialReferences(documentPath) {
        return this.partialReferences.has(documentPath);
    }
    hasCacheReferences(documentPath) {
        return this.cacheReferences.has(documentPath);
    }
}
ReferenceManager.instance = null;
if (typeof ReferenceManager.instance == "undefined" ||
    ReferenceManager.instance == null) {
    ReferenceManager.instance = new ReferenceManager();
}
export default ReferenceManager;
