// td-antlers-parser/out/antlers/tags/core/user/userDeletePasskeyForm.js
import { makeTagDocWithCodeSample } from '../../../../documentation/utils.js';
const UserDeletePasskeyForm = {
    tagName: 'user:delete_passkey_form',
    hideFromCompletions: false,
    requiresClose: true,
    injectParentScope: false,
    allowsContentClose: false,
    allowsArbitraryParameters: false,
    introducedIn: null,
    parameters: [
        {
            isRequired: true,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'id',
            description: 'The ID of the passkey to delete.',
            expectsTypes: ['string'],
            isDynamic: false
        },
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'redirect',
            description: 'Where to redirect to after deletion succeeds.',
            expectsTypes: ['string'],
            isDynamic: false
        }
    ],
    resolveDocumentation: (params) => {
        return makeTagDocWithCodeSample('User Delete Passkey Form Tag', 'The `user:delete_passkey_form` tag renders a form for removing an existing passkey from the authenticated user.', `{{ user:delete_passkey_form id="{passkey_id}" }}
    <button>Delete passkey</button>
{{ /user:delete_passkey_form }}`, 'https://statamic.dev/tags/user-delete_passkey_form');
    }
};
export default UserDeletePasskeyForm;
