// td-antlers-parser/out/antlers/tags/core/user/userPasskeys.js
import { makeTagDocWithCodeSample } from '../../../../documentation/utils.js';
const UserPasskeys = {
    tagName: 'user:passkeys',
    hideFromCompletions: false,
    requiresClose: true,
    injectParentScope: false,
    allowsContentClose: false,
    allowsArbitraryParameters: false,
    introducedIn: null,
    parameters: [
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: false,
            name: 'as',
            description: 'Alias the looped passkeys under the given variable name.',
            expectsTypes: ['string'],
            isDynamic: false
        }
    ],
    resolveDocumentation: (params) => {
        return makeTagDocWithCodeSample('User Passkeys Tag', 'The `user:passkeys` tag iterates over the currently authenticated user\'s registered passkeys.', `{{ user:passkeys }}
    <li>{{ name }} - registered {{ created_at }}</li>
{{ /user:passkeys }}`, 'https://statamic.dev/tags/user-passkeys');
    }
};
export default UserPasskeys;
