// td-antlers-parser/out/runtime/document/scanners/nodeQueries.js
import { EscapedContentNode, LiteralNode } from '../../nodes/abstractNode.js';
import { LanguageParser } from '../../parser/languageParser.js';
export class NodeQueries {
    static findAssignmentNodes(nodes) {
        const returnNodes = [];
        for (let i = 0; i < nodes.length; i++) {
            if (LanguageParser.isAssignmentOperatorNode(nodes[i])) {
                returnNodes.push(nodes[i]);
            }
        }
        return returnNodes;
    }
    static findAbstractNodesBefore(node, nodes) {
        return this.findAbstractNodesBeforePosition(node.startPosition, nodes);
    }
    static findAbstractNodesBeforePosition(position, nodes) {
        if (position == null) {
            return [];
        }
        const subNodes = [];
        for (let i = 0; i < nodes.length; i++) {
            const checkNode = nodes[i];
            if (checkNode.startPosition != null) {
                if (checkNode.startPosition.index > position.index) {
                    break;
                }
                subNodes.push(checkNode);
            }
        }
        return subNodes;
    }
    static isBefore(node, position) {
        if (node.startPosition == null) {
            return false;
        }
        return node.startPosition.index < position.index;
    }
    static isAfter(node, position) {
        if (node.startPosition == null) {
            return false;
        }
        return node.startPosition.index > position.index;
    }
    static findNodesBeforePosition(position, nodes) {
        const beforeNodes = [];
        if (position == null) {
            return beforeNodes;
        }
        for (let i = 0; i < nodes.length; i++) {
            const node = nodes[i];
            if (NodeQueries.isLiteralType(node)) {
                continue;
            }
            if (node.startPosition != null && node.endPosition != null) {
                if (node.startPosition.index > position.index) {
                    break;
                }
                beforeNodes.push(node);
            }
        }
        return beforeNodes;
    }
    static findNodesAfterPosition(position, nodes) {
        const afterNodes = [];
        if (position == null) {
            return afterNodes;
        }
        for (let i = 0; i < nodes.length; i++) {
            const node = nodes[i];
            if (NodeQueries.isLiteralType(node)) {
                continue;
            }
            if (node.startPosition != null) {
                if (node.startPosition.index > position.index) {
                    afterNodes.push(node);
                }
            }
        }
        return afterNodes;
    }
    static findNodeBeforePosition(position, nodes) {
        if (position == null) {
            return null;
        }
        let lastNode = null;
        for (let i = 0; i < nodes.length; i++) {
            const node = nodes[i];
            if (NodeQueries.isLiteralType(node)) {
                continue;
            }
            if (node.startPosition != null) {
                if (node.startPosition.index > position.index) {
                    break;
                }
                lastNode = node;
            }
        }
        return lastNode;
    }
    static findNodeAfterPosition(position, nodes) {
        if (position == null) {
            return null;
        }
        for (let i = 0; i < nodes.length; i++) {
            const node = nodes[i];
            if (NodeQueries.isLiteralType(node)) {
                continue;
            }
            if (node.startPosition != null) {
                if (node.startPosition.index > position.index) {
                    return node;
                }
            }
        }
        return null;
    }
    /**
     * Locates a node at the provided position.
     *
     * @param position
     * @param nodes
     * @returns
     */
    static findNodeAtPosition(position, nodes) {
        if (position == null) {
            return null;
        }
        for (let i = 0; i < nodes.length; i++) {
            const node = nodes[i];
            if (NodeQueries.isLiteralType(node)) {
                continue;
            }
            if (node.startPosition != null && node.endPosition != null) {
                if (position.index >= node.startPosition.index && position.index <= node.endPosition.index) {
                    return node;
                }
            }
        }
        return null;
    }
    static isLiteralType(node) {
        if (node instanceof EscapedContentNode || node instanceof LiteralNode) {
            return true;
        }
        return false;
    }
}
