// td-antlers-parser/out/runtime/document/scanners/nodeScanner.js
import { AntlersNode, ConditionNode, EscapedContentNode, LiteralNode, ParserFailNode } from '../../nodes/abstractNode.js';
import { AntlersNodeQueries } from './antlersNodeQueries.js';
import { NodeQueries } from './nodeQueries.js';
export class NodeScanner {
    constructor(doc) {
        this.doc = doc;
    }
    getAncestors(position) {
        return AntlersNodeQueries.findAncestorsBeforePosition(position, this.doc.getAllAntlersNodes());
    }
    getNodeAt(position) {
        return NodeQueries.findNodeAtPosition(position, this.doc.getDocumentParser().getNodes());
    }
    getNodesBefore(position) {
        return NodeQueries.findNodesBeforePosition(position, this.doc.getDocumentParser().getNodes());
    }
    getNodeBefore(position) {
        return NodeQueries.findNodeBeforePosition(position, this.doc.getDocumentParser().getNodes());
    }
    getNodeAfter(position) {
        return NodeQueries.findNodeAfterPosition(position, this.doc.getDocumentParser().getNodes());
    }
    getNodesAfter(position) {
        return NodeQueries.findNodesAfterPosition(position, this.doc.getDocumentParser().getNodes());
    }
    getIsWithinPairedNode(position) {
        const node = this.getNodeBefore(position);
        if (node == null) {
            return false;
        }
        return AntlersNodeQueries.isPairedNode(node);
    }
    filter(predicate) {
        return this.doc.getDocumentParser().getNodes().filter(predicate);
    }
    getComments() {
        const nodes = [];
        this.doc.getAllNodes().forEach((node) => {
            if (node instanceof AntlersNode && node.isComment) {
                nodes.push(node);
            }
        });
        return nodes;
    }
    getMultilineNodes() {
        const nodes = [];
        this.doc.getAllNodes().forEach((node) => {
            if (node instanceof AntlersNode && !node.isComment && (node.startPosition?.line != node.endPosition?.line)) {
                nodes.push(node);
            }
        });
        return nodes;
    }
    getMultiLineComments() {
        const nodes = [];
        this.getComments().forEach((comment) => {
            if (comment.startPosition?.line != comment.endPosition?.line) {
                nodes.push(comment);
            }
        });
        return nodes;
    }
    /**
     * Retrieves all literal and escaped content nodes in the document.
     */
    getAllLiteralNodes() {
        const nodes = [];
        this.doc.getAllNodes().forEach((node) => {
            if (node instanceof LiteralNode || node instanceof EscapedContentNode) {
                nodes.push(node);
            }
        });
        return nodes;
    }
    getStructuralNodes() {
        const nodes = [];
        this.doc.getDocumentParser().getRenderNodes().forEach((node) => {
            if (node instanceof AntlersNode) {
                nodes.push(node);
            }
            else if (node instanceof ConditionNode) {
                if (node.logicBranches.length > 0) {
                    node.logicBranches.forEach((branch) => {
                        if (branch.head != null) {
                            const tBrancHead = branch.head;
                            if (tBrancHead.originalNode != null) {
                                nodes.push(tBrancHead.originalNode);
                            }
                            else {
                                nodes.push(tBrancHead);
                            }
                        }
                    });
                }
            }
        });
        return nodes;
    }
    getAllRuntimeNodes() {
        const allNodes = [];
        this.doc.getAllNodes().forEach((node) => {
            if (node instanceof ParserFailNode || node instanceof AntlersNode) {
                if (node.runtimeNodes.length > 0) {
                    node.runtimeNodes.forEach((runtimeNode) => {
                        allNodes.push(runtimeNode);
                    });
                }
            }
        });
        return allNodes;
    }
    getAllParameterNodes() {
        const allParameters = [];
        this.doc.getAllNodes().forEach((node) => {
            if (node instanceof ParserFailNode || node instanceof AntlersNode) {
                if (node.hasParameters) {
                    node.parameters.forEach((parameter) => {
                        allParameters.push(parameter);
                    });
                }
            }
        });
        return allParameters;
    }
    getAssignmentNodes() {
        return NodeQueries.findAssignmentNodes(this.getAllRuntimeNodes());
    }
}
