// td-antlers-parser/out/antlers/tags/core/getErrors/getError.js
import { makeTagDoc } from '../../../../documentation/utils.js';
import { GetErrorsParameters } from './getErrorsParameters.js';
const GetError = {
    tagName: 'get_error',
    hideFromCompletions: false,
    requiresClose: true,
    injectParentScope: false,
    allowsArbitraryParameters: false,
    allowsContentClose: false,
    parameters: GetErrorsParameters,
    introducedIn: null,
    augmentScope: (node, scope) => {
        if (node.isClosingTag) {
            return scope;
        }
        scope.addVariable({
            dataType: 'string',
            introducedBy: node,
            name: 'message',
            sourceName: '*internal.tag.get_error',
            sourceField: {
                blueprintName: '*internal.tag.get_error',
                displayName: 'get_error',
                import: null,
                instructionText: null,
                maxItems: null,
                name: 'message',
                refFieldSetField: null,
                sets: null,
                type: 'string'
            }
        });
        return scope;
    },
    resolveDocumentation: (params) => {
        return makeTagDoc('get_error Tag', 'The `get_error` tag can be used to retrieve the first error message for a specific field.', 'https://statamic.dev/tags/get_errors#get-the-first-error-for-a-specific-field');
    }
};
export default GetError;
