// td-antlers-parser/out/runtime/document/contexts/positionContext.js
import { getId } from '../../../utils/simpleIds.js';
export class PositionContext {
    constructor() {
        this.refId = '';
        this.node = null;
        this.leftPunctuation = null;
        this.rightPunctuation = null;
        this.leftChar = null;
        this.rightChar = null;
        this.word = null;
        this.char = null;
        this.interpolatedContext = false;
        this.leftWord = null;
        this.rightWord = null;
        this.isCursorInIdentifier = false;
        this.root = null;
        this.identifierContext = null;
        this.position = null;
        this.feature = null;
        this.isInParameter = false;
        this.subContext = null;
        this.parameterContext = null;
        this.modifierContext = null;
        this.generalContext = null;
        this.variableContext = null;
        this.cursorContext = CursorContext.Unknown;
        this.refId = getId();
    }
}
export var CursorContext;
(function (CursorContext) {
    CursorContext[CursorContext["Unknown"] = 0] = "Unknown";
    CursorContext[CursorContext["Variable"] = 1] = "Variable";
    CursorContext[CursorContext["Modifier"] = 2] = "Modifier";
    CursorContext[CursorContext["Parameter"] = 3] = "Parameter";
    CursorContext[CursorContext["General"] = 4] = "General";
    CursorContext[CursorContext["LanguageOperator"] = 5] = "LanguageOperator";
})(CursorContext || (CursorContext = {}));
