// td-antlers-parser/out/runtime/utilities/stringUtilities.js
// ctype regexes adapted from: https://github.com/locutusjs/locutus/blob/master/src/php/strings/setlocale.js
import { is_numeric } from './isNumeric.js';
export class StringUtilities {
    static boolLabel(value) {
        return value ? 'true' : 'false';
    }
    static normalizeLineEndings(string, to = "\n") {
        return string.replace(/\r?\n/g, to);
    }
    static breakByNewLine(value) {
        return value.replace(/(\r\n|\n|\r)/gm, "\n").split("\n");
    }
    static split(text) {
        return text.split('');
    }
    static substring(text, start, length) {
        return text.substr(start, length);
    }
    static ctypeSpace(char) {
        return char?.search(/^[\f\n\r\t\v ]+$/g) !== -1;
    }
    static ctypeAlpha(char) {
        return char?.search(/^[A-Za-z]+$/g) !== -1;
    }
    static ctypeDigit(char) {
        return char?.search(/^[\d]+$/g) !== -1;
    }
    static ctypePunct(char) {
        return char?.search(/^[\u0021-\u002F\u003A-\u0040\u005B-\u0060\u007B-\u007E]+$/g) !== -1;
    }
    static trimLeft(value, charList = "\\s") {
        return value.replace(new RegExp("^[" + charList + "]+"), '');
    }
    static trimRight(value, charList = "\\s") {
        return value.replace(new RegExp("[" + charList + "]+$"), '');
    }
    static isNumeric(string) {
        return is_numeric(string);
    }
    static substringCount(haystack, needle) {
        const subStr = needle.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
        return (haystack.match(new RegExp(subStr, 'gi')) || []).length;
    }
    static getLastLine(content) {
        const lines = StringUtilities.breakByNewLine(content);
        if (lines.length == 0) {
            return '';
        }
        return lines[lines.length - 1];
    }
    static getFirstLine(content) {
        const lines = StringUtilities.breakByNewLine(content);
        if (lines.length == 0) {
            return '';
        }
        return lines[0];
    }
    static snakeCase(string) {
        return string.replace(/([A-Z])/g, ($1) => `_${$1.toLowerCase()}`).replace(/^_/, '');
    }
    static makeSlug(length) {
        if (length <= 2) {
            length = 7;
        }
        let result = '';
        const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789', charactersLength = characters.length;
        for (let i = 0; i < length - 1; i++) {
            result += characters.charAt(Math.floor(Math.random() *
                charactersLength));
        }
        let slug = 'A' + result + 'A';
        return slug.toLowerCase();
    }
}
