// td-antlers-parser/out/runtime/runtime/environmentDetails.js
import ModifierManager from "../../antlers/modifierManager.js";
import TagManager from "../../antlers/tagManagerInstance.js";
class EnvironmentDetails {
    constructor() {
        this.tagNames = [];
        this.modifierNames = [];
    }
    getModifierNames() {
        return this.modifierNames.concat(ModifierManager.instance?.getModifierNames() ?? []);
    }
    getTagNames() {
        return this.tagNames.concat(TagManager.instance?.getTagNames() ?? []);
    }
    setTagNames(tagNames) {
        this.tagNames = tagNames;
    }
    setModifierNames(modifierNames) {
        this.modifierNames = modifierNames;
    }
    isTag(value) {
        if (EnvironmentDetails.alwaysLikeTag.includes(value)) {
            return true;
        }
        if (TagManager.instance?.isKnownTag(value)) {
            return true;
        }
        return this.tagNames.includes(value);
    }
    isModifier(value) {
        if (ModifierManager.instance?.hasModifier(value)) {
            return true;
        }
        return this.modifierNames.includes(value);
    }
}
EnvironmentDetails.alwaysLikeTag = [
    "if",
    "endif",
    "elseif",
    "else",
    "unless",
    "elseunless",
];
export default EnvironmentDetails;
