// td-antlers-parser/out/antlers/scope/injections.js
import { trimLeft } from "../../utils/strings.js";
import { Scope } from './scope.js';
class InjectionManager {
    constructor() {
        this.fileInjections = new Map();
        this.project = null;
    }
    updateProject(project) {
        this.project = project;
    }
    registerInjections(fileName, nodes) {
        if (nodes.length == 0) {
            return;
        }
        // Only register nodes that have items in the parameter cache.
        const newNodes = [];
        for (let i = 0; i < nodes.length; i++) {
            const thisNode = nodes[i];
            if (thisNode.hasParameters) {
                newNodes.push(thisNode);
            }
        }
        if (newNodes.length > 0) {
            const partialName = newNodes[0].getMethodNameValue();
            if (partialName != null && this.project != null) {
                const projectPartial = this.project.findPartial(partialName);
                if (projectPartial != null) {
                    if (!this.fileInjections.has(projectPartial.documentUri)) {
                        this.fileInjections.set(projectPartial.documentUri, new Map());
                    }
                    const partialFiles = this.fileInjections.get(projectPartial.documentUri);
                    partialFiles.set(fileName, newNodes);
                }
            }
        }
    }
    hasAvailableInjections(documentUri) {
        return this.fileInjections.has(documentUri);
    }
    getScopeInjection(documentUri, project) {
        const newScope = new Scope(project);
        newScope.name = "*injection*";
        const injections = this.fileInjections.get(documentUri);
        injections.forEach((nodes, file) => {
            for (let i = 0; i < nodes.length; i++) {
                const node = nodes[i];
                if (node.hasParameters) {
                    for (let j = 0; j < node.parameters.length; j++) {
                        const thisParam = node.parameters[j];
                        if (thisParam.containsSimpleValue()) {
                            newScope.addVariable({
                                dataType: "string",
                                name: thisParam.name,
                                sourceField: null,
                                sourceName: "scope.injection",
                                introducedBy: node,
                            });
                        }
                        else if (thisParam.isVariableReference) {
                            if (node.currentScope != null) {
                                if (node.currentScope.hasListInHistory(thisParam.value)) {
                                    const injectList = node.currentScope.findNestedScope(thisParam.value);
                                    if (injectList != null) {
                                        const adjustedName = trimLeft(thisParam.name, ":"), nestedInjection = injectList.copy();
                                        nestedInjection.name = adjustedName;
                                        newScope.addScopeList(adjustedName, injectList);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        });
        return newScope;
    }
}
InjectionManager.instance = null;
if (typeof InjectionManager.instance == 'undefined' || InjectionManager.instance == null) {
    InjectionManager.instance = new InjectionManager();
}
export default InjectionManager;
