// td-antlers-parser/out/antlers/tags/alias.js
export function createDefinitionAlias(tag, alias) {
    const tagCopy = copyTagDefinition(tag);
    tagCopy.tagName = alias;
    return tagCopy;
}
function copyTagDefinition(tag) {
    return {
        allowsArbitraryParameters: tag.allowsArbitraryParameters,
        allowsContentClose: tag.allowsContentClose,
        injectParentScope: tag.injectParentScope,
        parameters: tag.parameters,
        requiresClose: tag.requiresClose,
        tagName: tag.tagName,
        introducedIn: tag.introducedIn,
        hideFromCompletions: tag.hideFromCompletions,
        augmentScope: tag.augmentScope,
        requiresCloseResolver: tag.requiresCloseResolver,
        resolveCompletionItems: tag.resolveCompletionItems,
        resolveDynamicParameter: tag.resolveDynamicParameter,
        resolveSpecialType: tag.resolveSpecialType,
        resovleParameterCompletionItems: tag.resovleParameterCompletionItems,
        suggestAlternativeParams: tag.suggestAlternativeParams,
        resolveDocumentation: tag.resolveDocumentation
    };
}
