// td-antlers-parser/out/suggestions/genericTypesSuggestions.js
import { InsertTextFormat, CompletionItemKind, TextEdit, } from "vscode-languageserver-types";
import { MarkupKind } from "vscode-languageserver-types";
import ModifierManager from '../antlers/modifierManager.js';
export class GenericTypesSuggestions {
    static getCompletions(params, scopeVariable) {
        const items = [];
        let replaceOffset = 0;
        if (params.leftWord != params.leftMeaningfulWord) {
            replaceOffset = params.leftWord.length;
        }
        const range = {
            start: {
                line: params.position.line,
                character: params.position.character - replaceOffset,
            },
            end: params.position,
        };
        ModifierManager.instance?.getModifiersForType(scopeVariable.dataType)
            .filter((m) => {
            return m.canBeParameter;
        })
            .forEach((modifier) => {
            items.push({
                label: modifier.name,
                documentation: {
                    kind: MarkupKind.Markdown,
                    value: modifier.description,
                },
                insertTextFormat: InsertTextFormat.Snippet,
                kind: CompletionItemKind.Value,
                textEdit: TextEdit.replace(range, modifier.name + '="$1" '),
                command: {
                    title: "Suggest",
                    command: "editor.action.triggerSuggest",
                },
            });
        });
        return items;
    }
}
