// td-antlers-parser/out/antlers/tags/core/dictionary.js
import { makeTagDocWithCodeSample } from '../../../documentation/utils.js';
const Dictionary = {
    tagName: 'dictionary',
    hideFromCompletions: false,
    requiresClose: true,
    injectParentScope: false,
    allowsContentClose: false,
    allowsArbitraryParameters: true,
    introducedIn: null,
    parameters: [
        {
            isRequired: true,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'handle',
            description: 'The dictionary handle to load.',
            expectsTypes: ['string'],
            isDynamic: false
        },
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'paginate',
            description: 'Paginate the results.',
            expectsTypes: ['number', 'boolean'],
            isDynamic: false
        },
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'sort',
            description: 'The field(s) to sort results by.',
            expectsTypes: ['string'],
            isDynamic: false
        },
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'limit',
            description: 'Limit the number of results.',
            expectsTypes: ['number'],
            isDynamic: false
        }
    ],
    resolveDocumentation: (params) => {
        return makeTagDocWithCodeSample('Dictionary Tag', 'The `dictionary` tag loads values from a dictionary, such as a list of countries or currencies.', `{{ dictionary:countries }}
    <option value="{{ value }}">{{ label }}</option>
{{ /dictionary:countries }}`, 'https://statamic.dev/tags/dictionary');
    }
};
export default Dictionary;
