// td-antlers-parser/out/antlers/tags/core/getSite.js
import { makeTagDocWithCodeSample } from '../../../documentation/utils.js';
const GetSite = {
    tagName: 'get_site',
    hideFromCompletions: false,
    requiresClose: false,
    injectParentScope: false,
    allowsContentClose: true,
    allowsArbitraryParameters: false,
    introducedIn: null,
    parameters: [
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'handle',
            description: 'The handle of the site to retrieve. Defaults to the current site.',
            expectsTypes: ['string'],
            isDynamic: false
        }
    ],
    resolveDocumentation: (params) => {
        return makeTagDocWithCodeSample('Get Site Tag', 'The `get_site` tag retrieves information about a specific site by its handle.', `{{ get_site handle="default" }}
    {{ name }} ({{ locale }})
{{ /get_site }}`, 'https://statamic.dev/tags/get_site');
    }
};
export default GetSite;
