// td-antlers-parser/out/antlers/tags/core/glide.js
import { makeTagDoc } from '../../../documentation/utils.js';
import { tagToCompletionItem } from '../../documentedLabel.js';
import { EmptyCompletionResult, exclusiveResult, } from "../../tagManager.js";
import { makeGlideVariables } from "../../variables/glideVariables.js";
import GlideBatch from './glideBatch.js';
import GlideDataUrl from './glideDataUrl.js';
import { GlideParameters, resolveGlideParameterCompletions } from './glideParameters.js';
const GlideCompletionItems = [
    tagToCompletionItem(GlideBatch),
    tagToCompletionItem(GlideDataUrl)
];
const Glide = {
    tagName: "glide",
    hideFromCompletions: false,
    allowsArbitraryParameters: false,
    allowsContentClose: false,
    injectParentScope: false,
    requiresClose: false,
    parameters: GlideParameters,
    introducedIn: null,
    augmentScope: (node, scope) => {
        scope.addVariables(makeGlideVariables(node));
        return scope;
    },
    resovleParameterCompletionItems: resolveGlideParameterCompletions,
    resolveCompletionItems: (params) => {
        if (params.isPastTagPart == false &&
            (params.leftWord == "glide" || params.leftWord == "/glide") &&
            params.leftChar == ":") {
            return exclusiveResult(GlideCompletionItems);
        }
        return EmptyCompletionResult;
    },
    resolveDocumentation: (params) => {
        return makeTagDoc('glide Tag', 'The `glide` tag is used to manipulate images on the fly. Image manipulations may include resizing, cropping, or adjusting sharpness and constrast, amongst many others.', 'https://statamic.dev/tags/glide');
    }
};
export default Glide;
