// td-antlers-parser/out/antlers/tags/core/increment.js
;
import { makeTagDoc } from '../../../documentation/utils.js';
import { tagToCompletionItem } from '../../documentedLabel.js';
import { EmptyCompletionResult, exclusiveResult } from '../../tagManager.js';
import IncrementReset from './incrementReset.js';
const IncrementCompletionItems = [
    tagToCompletionItem(IncrementReset)
];
const Increment = {
    tagName: 'increment',
    hideFromCompletions: false,
    requiresClose: false,
    injectParentScope: false,
    allowsContentClose: false,
    allowsArbitraryParameters: false,
    introducedIn: null,
    parameters: [
        {
            isRequired: false,
            name: 'from',
            aliases: [],
            description: 'The number to start incrementing by',
            acceptsVariableInterpolation: false,
            allowsVariableReference: false,
            expectsTypes: ['number'],
            isDynamic: false
        },
        {
            isRequired: false,
            name: 'by',
            aliases: [],
            description: 'The number to increment by',
            acceptsVariableInterpolation: false,
            allowsVariableReference: false,
            expectsTypes: ['number'],
            isDynamic: false
        }
    ],
    resolveCompletionItems: (params) => {
        if (params.leftWord == 'increment' || params.leftWord == '/increment' && params.leftChar == ':') {
            return exclusiveResult(IncrementCompletionItems);
        }
        return EmptyCompletionResult;
    },
    resolveDocumentation: (params) => {
        return makeTagDoc('increment Tag', 'The `increment` tag can be used to increment a value each time the tag is encountered.', 'https://statamic.dev/tags/increment');
    }
};
export default Increment;
