// td-antlers-parser/out/antlers/tags/core/installed.js
import { CompletionItemKind } from 'vscode-languageserver-types';
;
import { makeTagDocWithCodeSample } from '../../../documentation/utils.js';
import { EmptyCompletionResult, exclusiveResult } from '../../tagManager.js';
const InstalledTag = {
    tagName: 'installed',
    hideFromCompletions: false,
    requiresClose: false,
    injectParentScope: false,
    allowsContentClose: true,
    allowsArbitraryParameters: true,
    introducedIn: '3.1.19',
    parameters: [],
    resolveCompletionItems: (params) => {
        if (params.project != null) {
            const projectPackages = params.project.getComposerPackages(), packageCompletions = [];
            projectPackages.forEach((composerPackage, packageName) => {
                packageCompletions.push({
                    label: composerPackage.name,
                    kind: CompletionItemKind.Text
                });
            });
            return exclusiveResult(packageCompletions);
        }
        return EmptyCompletionResult;
    },
    resolveDocumentation: (params) => {
        return makeTagDocWithCodeSample('installed Tag', 'The `installed` tag is used to check if the site has a specific Composer package available.', `{{ if {installed:composer/package-name} }}
    {{# Package is availabl.e #}}
{{ else }}
    {{# Something else. #}}
{{ /if }}`, 'https://statamic.dev/tags/installed');
    }
};
export { InstalledTag };
