// td-antlers-parser/out/antlers/tags/core/locales.js
;
import { makeTagDoc } from '../../../documentation/utils.js';
import { tagToCompletionItem } from '../../documentedLabel.js';
import { EmptyCompletionResult, exclusiveResult } from "../../tagManager.js";
import { makeLocaleVariables } from "../../variables/localeVariables.js";
import LocalesCount from './localeCount.js';
import { LocaleParameters } from './localeParameters.js';
const LocalesCompletionItems = [
    tagToCompletionItem(LocalesCount)
];
const Locales = {
    tagName: "locales",
    hideFromCompletions: false,
    allowsArbitraryParameters: false,
    allowsContentClose: false,
    injectParentScope: false,
    requiresClose: true,
    parameters: LocaleParameters,
    introducedIn: null,
    augmentScope: (node, scope) => {
        scope.addVariableArray("locale", makeLocaleVariables(node));
        scope.addVariable({
            name: "current",
            dataType: "string",
            sourceField: null,
            sourceName: "*internal.locale",
            introducedBy: node,
        });
        scope.addVariable({
            name: "is_current",
            dataType: "boolean",
            sourceField: null,
            sourceName: "*internal.locale",
            introducedBy: node,
        });
        scope.addVariable({
            name: "exists",
            dataType: "boolean",
            sourceField: null,
            sourceName: "*internal.locale",
            introducedBy: node,
        });
        scope.addVariable({
            name: "permalink",
            dataType: "string",
            sourceField: null,
            sourceName: "*internal.locale",
            introducedBy: node,
        });
        return scope;
    },
    resolveDocumentation: (params) => {
        return makeTagDoc('locales Tag', 'The `locales` tag can be used to access all the locales the current content is available in.', 'https://statamic.dev/tags/locales');
    },
    resolveCompletionItems: (params) => {
        if (params.isPastTagPart == false &&
            (params.leftWord == "locale" || params.leftWord == "/locale") &&
            params.leftChar == ":") {
            return exclusiveResult(LocalesCompletionItems);
        }
        return EmptyCompletionResult;
    }
};
export default Locales;
