// td-antlers-parser/out/antlers/tags/core/markdown.js
import { makeTagDoc } from '../../../documentation/utils.js';
import { tagToCompletionItem } from '../../documentedLabel.js';
import { EmptyCompletionResult, exclusiveResult } from '../../tagManager.js';
import MarkdownIndent from './markdownIndent.js';
const MarkdownCompletionItems = [
    tagToCompletionItem(MarkdownIndent)
];
const Markdown = {
    tagName: 'markdown',
    hideFromCompletions: false,
    injectParentScope: false,
    requiresClose: true,
    allowsArbitraryParameters: false,
    allowsContentClose: false,
    introducedIn: null,
    parameters: [],
    resolveCompletionItems: (params) => {
        if (params.isPastTagPart == false && (params.leftWord == 'markdown' || params.leftWord == '/markdown') && params.leftChar == ':') {
            return exclusiveResult(MarkdownCompletionItems);
        }
        return EmptyCompletionResult;
    },
    resolveDocumentation: (params) => {
        return makeTagDoc('markdown Tag', 'The `markdown` tag can be used to render the tags contents as markdown.', 'https://statamic.dev/tags/markdown');
    }
};
export default Markdown;
