// td-antlers-parser/out/antlers/tags/core/mountUrl.js
import { CompletionItemKind } from 'vscode-languageserver-types';
;
import { makeTagDoc } from '../../../documentation/utils.js';
import { exclusiveResult } from '../../tagManager.js';
import { makeCollectionNameSuggestions } from './collection/utils.js';
const MountUrlTag = {
    tagName: 'mount_url',
    hideFromCompletions: false,
    allowsArbitraryParameters: true,
    allowsContentClose: false,
    requiresClose: false,
    injectParentScope: false,
    introducedIn: '3.3.15',
    parameters: [
        {
            isRequired: false,
            acceptsVariableInterpolation: true,
            aliases: [],
            allowsVariableReference: true,
            description: 'The collection handle.',
            name: 'handle',
            isDynamic: false,
            expectsTypes: ['string']
        }
    ],
    resolveDocumentation: (params) => {
        return makeTagDoc('mount Tag', 'The `mount` tag can be used to retrieve the mount URL for a given collection.', null);
    },
    resovleParameterCompletionItems: (parameter, params) => {
        if (parameter.name == 'handle') {
            if (params.context?.parameterContext != null) {
                return exclusiveResult(makeCollectionNameSuggestions(params.context?.parameterContext.parameter?.getArrayValue() ?? [], params.project));
            }
        }
        return null;
    },
    resolveCompletionItems: (params) => {
        const items = [];
        if (params.isPastTagPart == false && (params.leftWord == 'mount_url' || params.leftWord == '/mount_url') && params.leftChar == ':') {
            const collectionNames = params.project.getCollectionNames();
            for (let i = 0; i < collectionNames.length; i++) {
                items.push({
                    label: collectionNames[i],
                    kind: CompletionItemKind.Field,
                    sortText: '0'
                });
            }
        }
        return {
            items: items,
            analyzeDefaults: false,
            isExclusiveResult: false
        };
    }
};
export default MountUrlTag;
