// td-antlers-parser/out/antlers/tags/core/sessionFlash.js
import { makeTagDoc } from '../../../documentation/utils.js';
import { returnDynamicParameter } from '../dynamicParameterResolver.js';
import { SessionVariableContext } from './contexts/sessionContext.js';
const SessionFlash = {
    tagName: 'session:flash',
    hideFromCompletions: false,
    injectParentScope: false,
    requiresClose: false,
    allowsArbitraryParameters: true,
    allowsContentClose: false,
    parameters: [],
    introducedIn: null,
    resolveDynamicParameter: returnDynamicParameter,
    resolveSpecialType: (node, project) => {
        const context = new SessionVariableContext(node);
        return {
            context: context,
            issues: []
        };
    },
    resolveDocumentation: (params) => {
        return makeTagDoc('session:flash Tag', 'The `session:flash` tag can be used to add a value to the user\'s session. Flashed values only persist for a single request, and are then removed.', 'https://statamic.dev/tags/session-flash');
    }
};
export default SessionFlash;
