// td-antlers-parser/out/antlers/tags/core/session.js
import { CompletionItemKind } from "vscode-languageserver-types";
import SessionVariableManager from "../../../references/sessionVariableManager.js";
import { tagToCompletionItem } from '../../documentedLabel.js';
import { EmptyCompletionResult, exclusiveResult } from "../../tagManager.js";
import SessionDump from './sessionDump.js';
import SessionFlash from './sessionFlash.js';
import SessionFlush from './sessionFlush.js';
import SessionForget from './sessionForget.js';
import SessionHas from './sessionHas.js';
import SessionSet from './sessionSet.js';
const SessionTagCompletionItems = [
    tagToCompletionItem(SessionSet),
    tagToCompletionItem(SessionFlash),
    tagToCompletionItem(SessionForget),
    tagToCompletionItem(SessionFlush),
    tagToCompletionItem(SessionHas),
    tagToCompletionItem(SessionDump)
];
const SessionTag = {
    tagName: "session",
    requiresClose: false,
    allowsContentClose: true,
    allowsArbitraryParameters: false,
    hideFromCompletions: false,
    injectParentScope: false,
    introducedIn: null,
    parameters: [
        {
            name: "as",
            description: "An optional alias for the session data",
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: false,
            expectsTypes: ["string"],
            isDynamic: false,
            isRequired: false,
        },
    ],
    resolveCompletionItems: (params) => {
        if (params.isPastTagPart == false &&
            (params.leftWord == "session" || params.leftWord == "/session") &&
            params.leftChar == ":") {
            const knownSessionVars = SessionVariableManager.instance?.getKnownSessionVariableNames() ?? [];
            let sessionCompletions = [];
            sessionCompletions = sessionCompletions.concat(SessionTagCompletionItems);
            for (let i = 0; i < knownSessionVars.length; i++) {
                sessionCompletions.push({
                    label: knownSessionVars[i],
                    kind: CompletionItemKind.Variable,
                });
            }
            return exclusiveResult(sessionCompletions);
        }
        return EmptyCompletionResult;
    },
    augmentScope: (node, scope) => {
        const asParam = node.findParameter("as"), knownParams = SessionVariableManager.instance?.getKnownSessionVariableNames() ?? [], scopeVariables = [];
        for (let i = 0; i < knownParams.length; i++) {
            scopeVariables.push({
                dataType: "*",
                name: knownParams[i],
                sourceField: null,
                sourceName: "project.session",
                introducedBy: node,
            });
        }
        if (asParam == null) {
            scope.addVariables(scopeVariables);
        }
        else {
            scope.addVariableArray(asParam.value, scopeVariables);
        }
        return scope;
    },
};
export default SessionTag;
