// td-antlers-parser/out/antlers/tags/core/userProfile.js
import { makeTagDoc } from '../../../documentation/utils.js';
import { formatSuggestionList } from "../../../suggestions/fieldFormatter.js";
import { EmptyCompletionResult, exclusiveResult, } from "../../tagManager.js";
import { UserProfileParameters } from './userProfileParameters.js';
const UserProfile = {
    tagName: "user:profile",
    hideFromCompletions: false,
    requiresClose: true,
    injectParentScope: false,
    allowsContentClose: false,
    allowsArbitraryParameters: false,
    parameters: UserProfileParameters,
    introducedIn: null,
    augmentScope: (node, scope) => {
        scope.injectBlueprint(node, "user");
        return scope;
    },
    resovleParameterCompletionItems: (parameter, params) => {
        if (parameter.name == "field") {
            return exclusiveResult(formatSuggestionList(params.project.getUserFields()));
        }
        return EmptyCompletionResult;
    },
    resolveDocumentation: (params) => {
        return makeTagDoc('user:profile Tag', 'The `user:profile` tag provides access to the currently logged in user information, or for a specific user when using the `id`, `email`, or `field` parameter.', null);
    }
};
export default UserProfile;
