// td-antlers-parser/out/antlers/tags/core/users.js
import { makeTagDocWithCodeSample } from '../../../documentation/utils.js';
const Users = {
    tagName: 'users',
    hideFromCompletions: false,
    requiresClose: true,
    injectParentScope: false,
    allowsContentClose: false,
    allowsArbitraryParameters: true,
    introducedIn: null,
    parameters: [
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'group',
            description: 'Limit results to users in the specified group(s).',
            expectsTypes: ['string', 'array'],
            isDynamic: false
        },
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'role',
            description: 'Limit results to users with the specified role(s).',
            expectsTypes: ['string', 'array'],
            isDynamic: false
        },
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'sort',
            description: 'The field(s) to sort the results by.',
            expectsTypes: ['string'],
            isDynamic: false
        },
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'limit',
            description: 'Limit the number of users returned.',
            expectsTypes: ['number'],
            isDynamic: false
        },
        {
            isRequired: false,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: true,
            name: 'offset',
            description: 'The number of users to skip before returning results.',
            expectsTypes: ['number'],
            isDynamic: false
        }
    ],
    resolveDocumentation: (params) => {
        return makeTagDocWithCodeSample('Users Tag', 'The `users` tag fetches a list of users from the site.', `{{ users }}
    <li>{{ name }} - {{ email }}</li>
{{ /users }}`, 'https://statamic.dev/tags/users');
    }
};
export default Users;
