// td-antlers-parser/out/antlers/tags/core/vite.js
;
import { makeTagDocWithCodeSample } from '../../../documentation/utils.js';
import { EmptyCompletionResult, exclusiveResult } from '../../tagManager.js';
import { tagToCompletionItem } from '../../documentedLabel.js';
import ViteAsset from './vite/viteAsset.js';
const ViteTagCompletionItems = [
    tagToCompletionItem(ViteAsset)
];
const Vite = {
    tagName: 'vite',
    hideFromCompletions: false,
    requiresClose: false,
    injectParentScope: false,
    allowsContentClose: false,
    allowsArbitraryParameters: false,
    introducedIn: null,
    parameters: [
        {
            isRequired: true,
            acceptsVariableInterpolation: false,
            aliases: [],
            allowsVariableReference: false,
            name: 'src',
            description: 'The entry point',
            expectsTypes: ['string'],
            isDynamic: false
        }
    ],
    resolveCompletionItems: (params) => {
        if ((params.leftWord == "vite" ||
            params.leftWord == "/vite") &&
            params.leftChar == ":") {
            return exclusiveResult(ViteTagCompletionItems);
        }
        return EmptyCompletionResult;
    },
    resolveDocumentation: (params) => {
        return makeTagDocWithCodeSample('vite Tag', 'Generates Vite tags for an entrypoint.', `{{ vite src="entry/point" }}`, null);
    }
};
export default Vite;
