// td-antlers-parser/out/antlers/tags/core/theme/themeOutput.js
import { makeTagDoc } from '../../../../documentation/utils.js';
const ThemeOutput = {
    tagName: "theme:output",
    hideFromCompletions: false,
    allowsArbitraryParameters: false,
    allowsContentClose: true,
    requiresClose: false,
    injectParentScope: false,
    introducedIn: null,
    parameters: [
        {
            isRequired: true,
            allowsVariableReference: true,
            acceptsVariableInterpolation: false,
            aliases: [],
            name: "src",
            description: "The path to the file",
            expectsTypes: ["string"],
            isDynamic: false,
        },
        {
            isRequired: false,
            allowsVariableReference: true,
            acceptsVariableInterpolation: false,
            aliases: [],
            name: "as",
            description: "An optional alias for the contents variable",
            expectsTypes: ["string"],
            isDynamic: false,
        },
    ],
    augmentScope: (node, scope) => {
        if (node.isClosedBy != null) {
            const param = node.findParameter("as");
            if (param != null) {
                scope.addVariable({
                    name: param.value,
                    dataType: "string",
                    sourceField: null,
                    sourceName: "*internal.theme.context",
                    introducedBy: node,
                });
            }
            else {
                scope.addVariable({
                    name: "output_contents",
                    dataType: "string",
                    sourceField: null,
                    sourceName: "*internal.theme.context",
                    introducedBy: node,
                });
            }
        }
        return scope;
    },
    resolveDocumentation: (params) => {
        return makeTagDoc('theme:output Tag', 'The `theme:output` tag may be used to output the contents of any arbitrary file, relative to the site\'s `resources/` directory.', null);
    }
};
export default ThemeOutput;
