// td-antlers-parser/out/projects/projectManager.js
class ProjectManager {
    constructor() {
        this.currentStructure = null;
        this.isReloading = false;
        this.isDirtyState = true;
        this.structuredProject = null;
    }
    setStructuredProject(fields) {
        this.structuredProject = fields;
    }
    hasStructuredBlueprints() {
        return this.structuredProject != null;
    }
    getAllStructuredBlueprints() {
        if (this.structuredProject == null) {
            return [];
        }
        return this.structuredProject.assets
            .concat(this.structuredProject.collections)
            .concat(this.structuredProject.forms)
            .concat(this.structuredProject.general)
            .concat(this.structuredProject.navigations)
            .concat(this.structuredProject.taxonomies);
    }
    getStructuredProject() {
        return this.structuredProject;
    }
    setActiveProject(project) {
        this.currentStructure = project;
    }
    getStructure() {
        return this.currentStructure;
    }
    hasStructure() {
        return this.currentStructure != null;
    }
    setDirtyState(isDirty) {
        this.isDirtyState = isDirty;
    }
    /**
     * Requests the project structure implementation to reload its information.
     */
    reloadDetails() {
        if (this.isReloading) {
            return;
        }
        if (this.isDirtyState == false) {
            return;
        }
        if (this.currentStructure != null) {
            this.isReloading = true;
            this.currentStructure = this.currentStructure.reloadDetails();
            this.isReloading = false;
            this.isDirtyState = false;
        }
    }
}
ProjectManager.instance = null;
if (typeof ProjectManager.instance == 'undefined' || ProjectManager.instance == null) {
    ProjectManager.instance = new ProjectManager();
}
export default ProjectManager;
