// td-antlers-parser/out/projects/statamicStructure.js
const MockStructure = {
    isMocked: true,
    rootPath: "",
    workingDirectory: "",
    assets: new Map(),
    assetFields: new Map(),
    globalBlueprintsPath: "",
    oauthProviders: [],
    fieldsetsPath: "",
    collections: new Map(),
    collectionScopes: [],
    blueprints: [],
    viewsPath: "",
    collectionBlueprintsPath: "",
    formBlueprintsPath: "",
    contentDirectory: "",
    taxonomyContentDirectory: "",
    taxonomiesBlueprintsPath: "",
    taxonomyPluralizedMapping: new Map(),
    taxonomyTerms: new Map(),
    globalFiles: [],
    fieldsets: new Map(),
    miscBlueprintsPath: "",
    blueprintFiles: [],
    collectionNames: [],
    forms: new Map(),
    globals: new Map(),
    taxonomies: new Map(),
    taxonomyNames: [],
    fields: new Map(),
    miscFields: new Map(),
    templates: [],
    views: [],
    partialCache: [],
    partialViewNames: [],
    hasMacrosFile: false,
    macroFilePath: "",
    composerPackages: [],
    statamicPackage: null,
    navigationMenus: new Map(),
    searchIndexes: [],
    userPermissionsPath: "",
    userRoles: new Map(),
    userRoleNames: [],
    userGroups: new Map(),
    userGroupNames: [],
    internalFieldReference: [],
    restoreProperties: null,
    namedBluePrintFields: new Map(),
    customModifierNames: [],
    customTags: [],
};
export {};
