// td-antlers-parser/out/runtime/errors/antlersError.js
import { Md5 } from 'ts-md5';
;
;
export var ErrorLevel;
(function (ErrorLevel) {
    ErrorLevel[ErrorLevel["Error"] = 0] = "Error";
    ErrorLevel[ErrorLevel["Warning"] = 1] = "Warning";
})(ErrorLevel || (ErrorLevel = {}));
export class AntlersError {
    constructor() {
        this.node = null;
        this.errorCode = '';
        this.message = '';
        this.level = ErrorLevel.Error;
        this.range = null;
        this.actions = [];
        this.lsRange = null;
        this.data = null;
    }
    hash() {
        let positionSlug = '';
        if (this.node != null) {
            positionSlug = (this.node.endPosition?.offset ?? 0).toString() + "|" +
                (this.node.startPosition?.offset ?? 0).toString() + "|";
        }
        return Md5.hashStr(positionSlug + "|" + this.errorCode);
    }
    static makeSyntaxError(errorCode, node, message, level) {
        if (level == null) {
            level = ErrorLevel.Error;
        }
        const error = new AntlersError();
        error.errorCode = errorCode;
        error.node = node;
        error.message = message;
        error.level = level;
        return error;
    }
}
